{
 "cells": [
  {
   "cell_type": "raw",
   "id": "512454c2",
   "metadata": {},
   "source": [
    "---\n",
    "title: \"PhenoCycler Processing Workflows with Voyager\"\n",
    "output: html_document\n",
    "vignette: >\n",
    "  %\\VignetteIndexEntry{cosmx_landing}\n",
    "  %\\VignetteEngine{knitr::rmarkdown}\n",
    "  %\\VignetteEncoding{UTF-8}\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6db98058",
   "metadata": {
    "tags": [
     "remove_cell"
    ]
   },
   "outputs": [],
   "source": [
    "knitr::opts_chunk$set(\n",
    "  collapse = TRUE,\n",
    "  comment = \"#>\",\n",
    "  eval = FALSE\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12fbb746",
   "metadata": {},
   "source": [
    "The commercially available [PhenoCycler Solution](https://www.akoyabio.com/phenocycler/) by Akoya Biosciences relies on Co-Detection by Indexing (CODEX), a technique enables highly multiplexed single-cell quantification of membrane protein expression in fresh-frozed or formalin fixed paraffin embedded (FFPE) tissues. In CODEX, antigens are stained simultaneously with unique oligonucleotide-conjugated antibodies. Antibody binding events are visualized iteratively via successive rounds of fluorescent microscopy. At each imaging round, cells are exposed to a mix of two non-fluorescent index nucleotides and two fluorescent labeling nucleotides. The DNA barcodes are designed such that only two antibodies can be labeled in a single imaging cycle, and only if the indexing nucleotide has been incorporated. Notably, CODEX completes a 30-antibody visualization in approximately 3.5 hours and is adaptable to most standard three-color fluorescence microscope platforms. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1910ed4",
   "metadata": {},
   "source": [
    "## Getting Started\n",
    "\n",
    "Several CODEX datasets were generated by the HuBMAP Consortium are are available for download from their [data  portal](https://portal.hubmapconsortium.org). Raw and processed data are typically avaiable for several fields of view and can be readily combined into a single [`SpatialFeatureExperiment`](https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html)(SFE) object. A tutorial for processing the output of various spatial transcriptomics technologies into a SFE object for use with `Voyager` is available on the [Getting Started]() page and in the vignette linked below. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c542150b",
   "metadata": {},
   "source": [
    "| Vignette | Colab Notebook | Description |\n",
    "|--------------------|:---------------:|---------------------------------|\n",
    "| [Create a `SFE` object](https://pachterlab.github.io/voyager/articles/create_sfe.html) |                | Download data, create `SFE` object|"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "903fd3c0",
   "metadata": {},
   "source": [
    "## Analysis Workflows\n",
    "\n",
    "The vignettes below demonstrate workflows that can be implemented with `Voyager` using data generated with CODEX technology. \n",
    "\n",
    "| Vignette | Colab Notebook | Description |\n",
    "|--------------------|:---------------:|---------------------------------|\n",
    "| [CODEX colon analysis](https://pachterlab.github.io/voyager/dev/articles/vig8_codex.html) | | Perform basic QC, data normalization, dimension reduction, spatial autocorrelation metrics, local spatial statistics, non-spatial clustering, differential expression |"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "tags,-all",
   "main_language": "R",
   "notebook_metadata_filter": "-all"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
