{
 "cells": [
  {
   "cell_type": "raw",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "dc5665cc",
||||||| 6a30bf1
   "id": "f4099258",
=======
   "id": "6a078ff7",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "6a078ff7",
=======
   "id": "96bce545",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "dc5665cc",
||||||| 6a30bf1
   "id": "f4099258",
=======
   "id": "6a078ff7",
>>>>>>> documentation-devel
=======
   "id": "401b3907",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {},
   "source": [
    "---\n",
    "title: \"How to build count matrices with cellatlas\"\n",
    "author: \"Kayla Jackson and A. Sina Booeshaghi\"\n",
    "date: \"`r Sys.Date()`\"\n",
    "output: \n",
    "  html_document:\n",
    "  toc: true\n",
    "  toc_float:\n",
    "    collapsed: false\n",
    "  toc_depth: 3\n",
    "bibliography: ref.bib\n",
    "vignette: >\n",
    "  %\\VignetteIndexEntry{create_mtx_ca}\n",
    "  %\\VignetteEngine{knitr::rmarkdown}\n",
    "  %\\VignetteEncoding{UTF-8}\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "b1ab5634",
||||||| 6a30bf1
   "id": "087ecfa4",
=======
   "id": "95fe1e6b",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "95fe1e6b",
=======
   "id": "a3bbd362",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "b1ab5634",
||||||| 6a30bf1
   "id": "087ecfa4",
=======
   "id": "95fe1e6b",
>>>>>>> documentation-devel
=======
   "id": "5741c31c",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "tags": [
     "remove_cell"
    ]
   },
   "outputs": [],
   "source": [
    "knitr::opts_chunk$set(\n",
    "  collapse = TRUE,\n",
    "  comment = \"#>\",\n",
    "  eval = FALSE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "20a55792",
||||||| 6a30bf1
   "id": "cf4d2147",
=======
   "id": "9c77a8ff",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "9c77a8ff",
=======
   "id": "9503228d",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "20a55792",
||||||| 6a30bf1
   "id": "cf4d2147",
=======
   "id": "9c77a8ff",
>>>>>>> documentation-devel
=======
   "id": "34fd6351",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "tags": [
     "remove_cell"
    ]
   },
   "outputs": [],
   "source": [
    "# Install Google Colab dependencies\n",
    "# Note: this can take 30+ minutes (many of the dependencies include C++ code, which needs to be compiled)\n",
    "\n",
    "# First install `sf`, `ragg` and `textshaping` and their system dependencies:\n",
    "system(\"apt-get -y update && apt-get install -y  libudunits2-dev libgdal-dev libgeos-dev libproj-dev libharfbuzz-dev libfribidi-dev\")\n",
    "\n",
    "install.packages(\"stringr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "6c7e1631",
||||||| 6a30bf1
   "id": "31231e4c",
=======
   "id": "6da5f822",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "6da5f822",
=======
   "id": "194df3c6",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "6c7e1631",
||||||| 6a30bf1
   "id": "31231e4c",
=======
   "id": "6da5f822",
>>>>>>> documentation-devel
=======
   "id": "9385d89c",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "name": "setup"
   },
   "outputs": [],
   "source": [
    "library(stringr)"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "17dd4849",
||||||| 6a30bf1
   "id": "cde30552",
=======
   "id": "f8f242d8",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "f8f242d8",
=======
   "id": "c726fdb7",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "17dd4849",
||||||| 6a30bf1
   "id": "cde30552",
=======
   "id": "f8f242d8",
>>>>>>> documentation-devel
=======
   "id": "e253f008",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Building Count Matrices with `cellatlas`\n",
    "A major challenge in uniformly preprocessing large amounts of single-cell genomics data from a variety of different assays is identifying and handling sequenced elements in a coherent and consistent fashion. Cell barcodes in reads from RNAseq data from 10x Multiome, for example, must be extracted and error corrected in the manner as cell barcodes in reads from ATACseq data from 10x Multiome so that barcode-barcode registration can occur. Uniform processing in this way minimzes computational variability and enables cross-assay comparisons.\n",
    "\n",
    "In this notebook we demonstrate how single-cell genomics data can be preprocessed to generate a cell by feature count matrix. This requires:\n",
    "\n",
    "1. FASTQ files\n",
    "2. `seqspec` specification for the FASTQ files\n",
    "3. Genome Sequence FASTA\n",
    "4. Genome Annotation GTF\n",
    "5. (optional) Feature barcode list\n",
    "\n",
    "# Install Packages\n",
    "The vignette makes use of two non-standard command line tools, [`jq`](https://jqlang.github.io/jq/) and [`tree`](https://mama.indstate.edu/users/ice/tree/). The code cell below installs these tools on a Linux operating system and should be updated for Mac and Windows users."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "23ba771d",
||||||| 6a30bf1
   "id": "d417a998",
=======
   "id": "ac672723",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "ac672723",
=======
   "id": "80f9e935",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "23ba771d",
||||||| 6a30bf1
   "id": "d417a998",
=======
   "id": "ac672723",
>>>>>>> documentation-devel
=======
   "id": "2c5d6a80",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": true,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "# Install `tree` to view files\n",
    "system(\"wget --quiet --show-progress ftp://mama.indstate.edu/linux/tree/tree-2.1.0.tgz\")\n",
    "\n",
    "# Install `jq`, a command-line tool for extracting key value pairs from JSON files \n",
    "system(\"wget --quiet --show-progress https://github.com/stedolan/jq/releases/download/jq-1.6/jq-osx-amd64\")\n",
    "system(\"chmod +x jq-linux64 && mv jq-linux64 /usr/local/bin/jq\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "acdc0a5e",
||||||| 6a30bf1
   "id": "02484fda",
=======
   "id": "ed131ae0",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "ed131ae0",
=======
   "id": "bfacb007",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "acdc0a5e",
||||||| 6a30bf1
   "id": "02484fda",
=======
   "id": "ed131ae0",
>>>>>>> documentation-devel
=======
   "id": "aa475624",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "We will continue with other dependencies that can be installed on any operating system. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "6ff7d297",
||||||| 6a30bf1
   "id": "9d8415e9",
=======
   "id": "45df717e",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "45df717e",
=======
   "id": "f053e7b3",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "6ff7d297",
||||||| 6a30bf1
   "id": "9d8415e9",
=======
   "id": "45df717e",
>>>>>>> documentation-devel
=======
   "id": "dbda44cb",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": true
   },
   "outputs": [],
   "source": [
    "# Clone the cellatlas repo and install the package\n",
    "system(\"git clone https://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@github.com/cellatlas/cellatlas.git\")\n",
    "system(\"cd cellatlas && pip install .\")\n",
    "\n",
    "# Install dependencies\n",
    "system(\"yes | pip uninstall --quiet seqspec\")\n",
    "system(\"pip install --quiet git+https://github.com/IGVF/seqspec.git\")\n",
    "system(\"pip install --quiet gget kb-python\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "f69179f8",
||||||| 6a30bf1
   "id": "43a1060c",
=======
   "id": "7aae81d5",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "7aae81d5",
=======
   "id": "0a40e557",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "f69179f8",
||||||| 6a30bf1
   "id": "43a1060c",
=======
   "id": "7aae81d5",
>>>>>>> documentation-devel
=======
   "id": "cd3d7580",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Preprocessing Workflows\n",
    "Below, we have generated technology specific pre-processing workflows that utilize `seqspec` to format genomic library sequence and structure and `cellatlas` to generate pipelines that process raw data into a count matrix. Currently, example workflows exist for data generated using [SPLiT-seq](#rna_10xsplitseq), [ClickTags](#clicktag), [Chromium V3 Chemistry](#rna_10xv3), [Chromium Single Cell ATAC-seq](#atac_10x), [Chromium Single Cell ATAC Multiome ATAC + Gene Expression](#atac_10xmulti), [Chromium Single Cell CRISPR Screening](#crispr_10x), and [Chromium Nuclei Isolation](#nuclei).\n",
    "\n",
    "Currently, [Visium](#rna_10xvisium) is the only spatial technology with an available `seqspec`.  \n",
    "\n",
    "More information on how to generate a `seqspec` for technologies not listed here is available on [GitHub](https://github.com/IGVF/seqspec). \n",
    "\n",
    "## <a id=\"rna_10xvisium\"></a>Preprocessing for Visium\n",
    "\n",
    "### Examine the spec\n",
    "**Note:** We move the relevant data to the working directory and `gunzip` the barcode onlist."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "4a6a188b",
||||||| 6a30bf1
   "id": "3d6a2537",
=======
   "id": "4574241c",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "4574241c",
=======
   "id": "4591d697",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "4a6a188b",
||||||| 6a30bf1
   "id": "3d6a2537",
=======
   "id": "4574241c",
>>>>>>> documentation-devel
=======
   "id": "d2a69701",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"mv cellatlas/examples/rna-visium-spatial/* .\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "b88938f9",
||||||| 6a30bf1
   "id": "fae68a61",
=======
   "id": "c926fdfa",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "c926fdfa",
=======
   "id": "5b6e9f9d",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "b88938f9",
||||||| 6a30bf1
   "id": "fae68a61",
=======
   "id": "c926fdfa",
>>>>>>> documentation-devel
=======
   "id": "3300fa84",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "We first use `seqspec print` to check that the read structure matches what we expect. This command prints out an ordered tree representation of the sequenced elements contained in the FASTQ files. Note that the names of the nodes in the `seqspec` must match the names of the FASTQ files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "59c9c998",
||||||| 6a30bf1
   "id": "835b5022",
=======
   "id": "c983b7e1",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "c983b7e1",
=======
   "id": "8ca35f27",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "59c9c998",
||||||| 6a30bf1
   "id": "835b5022",
=======
   "id": "c983b7e1",
>>>>>>> documentation-devel
=======
   "id": "a3595103",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"seqspec print spec.yaml\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "487f2301",
||||||| 6a30bf1
   "id": "15e98a66",
=======
   "id": "d42e73e5",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "d42e73e5",
=======
   "id": "b6e38d91",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "487f2301",
||||||| 6a30bf1
   "id": "15e98a66",
=======
   "id": "d42e73e5",
>>>>>>> documentation-devel
=======
   "id": "d1405b99",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Fetch the references\n",
    "This step is only necessary if the modality that we are processing uses a transcriptome reference-based alignment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "7e2db9c7",
||||||| 6a30bf1
   "id": "968c0050",
=======
   "id": "127ebcfd",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "127ebcfd",
=======
   "id": "303aa4e9",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "7e2db9c7",
||||||| 6a30bf1
   "id": "968c0050",
=======
   "id": "127ebcfd",
>>>>>>> documentation-devel
=======
   "id": "3664b331",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"gget ref -o ref.json -w dna,gtf mus_musculus\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "38980239",
||||||| 6a30bf1
   "id": "981eabad",
=======
   "id": "5e4b5f6f",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "5e4b5f6f",
=======
   "id": "d2692f4a",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "38980239",
||||||| 6a30bf1
   "id": "981eabad",
=======
   "id": "5e4b5f6f",
>>>>>>> documentation-devel
=======
   "id": "b9120b8b",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Build the pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "4d226676",
||||||| 6a30bf1
   "id": "779544cb",
=======
   "id": "05e53f4c",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "05e53f4c",
=======
   "id": "adf69ca4",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "4d226676",
||||||| 6a30bf1
   "id": "779544cb",
=======
   "id": "05e53f4c",
>>>>>>> documentation-devel
=======
   "id": "afaf631c",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "FA <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.mus_musculus.genome_dna.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)\n",
    "\n",
    "GTF <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.mus_musculus.annotation_gtf.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "40afd747",
||||||| 6a30bf1
   "id": "920df142",
=======
   "id": "789b1a19",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "789b1a19",
=======
   "id": "c79acb7e",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "40afd747",
||||||| 6a30bf1
   "id": "920df142",
=======
   "id": "789b1a19",
>>>>>>> documentation-devel
=======
   "id": "349be1fa",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "We now supply all of the relevant objects to `cellatlas build` to produce the appropriate commands to be run to build the pipeline. This includes a reference building step and a read counting and quantification step both of which are performed with `kallisto` and `bustools` as part of the `kb-python` package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "fe72a724",
||||||| 6a30bf1
   "id": "dcd4362f",
=======
   "id": "5a96f4f9",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "5a96f4f9",
=======
   "id": "3b833451",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "fe72a724",
||||||| 6a30bf1
   "id": "dcd4362f",
=======
   "id": "5a96f4f9",
>>>>>>> documentation-devel
=======
   "id": "777dff0b",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "args <- c(\n",
    "  \"-o out\", \n",
    "  \"-s spec.yaml\",\n",
    "  \"-m rna\",  \n",
    "  \"-fa\", FA,\n",
    "  \"-g\", GTF,\n",
    "  \"fastqs/R1.fastq.gz fastqs/R2.fastq.gz\")\n",
    "\n",
    "system2(command = \"cellatlas\", args = c(\"build\", args))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "b7680369",
||||||| 6a30bf1
   "id": "fcaf593f",
=======
   "id": "f837c9ac",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "f837c9ac",
=======
   "id": "4eca1f10",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "b7680369",
||||||| 6a30bf1
   "id": "fcaf593f",
=======
   "id": "f837c9ac",
>>>>>>> documentation-devel
=======
   "id": "e50dbbe6",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Run the pipeline\n",
    "We can extract and view the commands in the pipeline using `jq`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "67589a09",
||||||| 6a30bf1
   "id": "04f51eb5",
=======
   "id": "987ef38b",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "987ef38b",
=======
   "id": "09b610e6",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "67589a09",
||||||| 6a30bf1
   "id": "04f51eb5",
=======
   "id": "987ef38b",
>>>>>>> documentation-devel
=======
   "id": "8d9b50f4",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "cmds <- system2(\"jq\", \"-r '.commands[] | values[]' out/cellatlas_info.json\", stdout=TRUE)\n",
    "cmds <- str_subset(cmds, \"[\\\\[\\\\]]\", negate=TRUE)\n",
    "cmds <- str_extract(cmds, \"kb.*(txt|gz)\")\n",
    "\n",
    "cmds"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "1f88dd44",
||||||| 6a30bf1
   "id": "66833387",
=======
   "id": "dc076850",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "dc076850",
=======
   "id": "94a03205",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "1f88dd44",
||||||| 6a30bf1
   "id": "66833387",
=======
   "id": "dc076850",
>>>>>>> documentation-devel
=======
   "id": "41a19741",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "Now we can run the commands from `out/cellatlas_info.json` on the command line. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "dd5b75d7",
||||||| 6a30bf1
   "id": "1d99fedc",
=======
   "id": "d5139cf4",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "d5139cf4",
=======
   "id": "4d21280c",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "dd5b75d7",
||||||| 6a30bf1
   "id": "1d99fedc",
=======
   "id": "d5139cf4",
>>>>>>> documentation-devel
=======
   "id": "e9a7c375",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "lapply(cmds, function(cmd) system(cmd))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "d229e465",
||||||| 6a30bf1
   "id": "3a611770",
=======
   "id": "deec20af",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "deec20af",
=======
   "id": "542b106d",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "d229e465",
||||||| 6a30bf1
   "id": "3a611770",
=======
   "id": "deec20af",
>>>>>>> documentation-devel
=======
   "id": "454260da",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Inspect the output \n",
    "We inspect the `out/run_info.json` and `out/kb_info.json` as a simple QC on the pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "86eda638",
||||||| 6a30bf1
   "id": "a74af770",
=======
   "id": "05f17a2e",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "05f17a2e",
=======
   "id": "c5764cc3",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "86eda638",
||||||| 6a30bf1
   "id": "a74af770",
=======
   "id": "05f17a2e",
>>>>>>> documentation-devel
=======
   "id": "f48783c8",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {},
   "outputs": [],
   "source": [
    "system(\"tree out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "5d87ffc8",
||||||| 6a30bf1
   "id": "5357b30f",
=======
   "id": "f6e5f8f8",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "f6e5f8f8",
=======
   "id": "2339c2d5",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "5d87ffc8",
||||||| 6a30bf1
   "id": "5357b30f",
=======
   "id": "f6e5f8f8",
>>>>>>> documentation-devel
=======
   "id": "e04081ee",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"cat out/run_info.json\")\n",
    "# system(\"cat out/kb_info.json\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "12f424bb",
||||||| 6a30bf1
   "id": "49e4a0af",
=======
   "id": "5a1934ba",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "5a1934ba",
=======
   "id": "67457194",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "12f424bb",
||||||| 6a30bf1
   "id": "49e4a0af",
=======
   "id": "5a1934ba",
>>>>>>> documentation-devel
=======
   "id": "14d10854",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## <a id=\"rna_10xsplitseq\"></a>Preprocessing for SPLiT-seq\n",
    "**Note:** We move the relevant data to the working directory and `gunzip` the barcode onlist. The data for this example are located in the `cellatlas/examples/rna-splitseq/` directory. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "efcaeb79",
||||||| 6a30bf1
   "id": "c80c4306",
=======
   "id": "ccb0d8a8",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "ccb0d8a8",
=======
   "id": "6723b26d",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "efcaeb79",
||||||| 6a30bf1
   "id": "c80c4306",
=======
   "id": "ccb0d8a8",
>>>>>>> documentation-devel
=======
   "id": "2155fa3a",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"mv cellatlas/examples/rna-splitseq/* .\")\n",
    "system(\"gunzip barcode*\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "b5fe5586",
||||||| 6a30bf1
   "id": "2fef13fd",
=======
   "id": "b321d5cc",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "b321d5cc",
=======
   "id": "0370bf4f",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "b5fe5586",
||||||| 6a30bf1
   "id": "2fef13fd",
=======
   "id": "b321d5cc",
>>>>>>> documentation-devel
=======
   "id": "23648ec7",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "The `seqspec print` command prints out an ordered tree representation of the sequenced elements contained in the FASTQ files. Note that the names of the nodes in the `seqspec` must match the names of the FASTQ files. The `seqspec` for SPLiT-seq contains the specification for multiple split-pool rounds. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "18e642e9",
||||||| 6a30bf1
   "id": "09b831bc",
=======
   "id": "77f72262",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "77f72262",
=======
   "id": "0680bfc5",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "18e642e9",
||||||| 6a30bf1
   "id": "09b831bc",
=======
   "id": "77f72262",
>>>>>>> documentation-devel
=======
   "id": "f31bb794",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"seqspec print spec.yaml\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "9bfb2b17",
||||||| 6a30bf1
   "id": "df3485d9",
=======
   "id": "1d212fe3",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "1d212fe3",
=======
   "id": "52930fb6",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "9bfb2b17",
||||||| 6a30bf1
   "id": "df3485d9",
=======
   "id": "1d212fe3",
>>>>>>> documentation-devel
=======
   "id": "d9135547",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Fetch the references\n",
    "This step is only necessary if the modality that we are processing uses a transcriptome reference-based alignment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "5cea090f",
||||||| 6a30bf1
   "id": "1482a3a7",
=======
   "id": "e7480917",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "e7480917",
=======
   "id": "5d040a58",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "5cea090f",
||||||| 6a30bf1
   "id": "1482a3a7",
=======
   "id": "e7480917",
>>>>>>> documentation-devel
=======
   "id": "e8e6c6c0",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"gget ref -o ref.json -w dna,gtf mus_musculus\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "9a875c0c",
||||||| 6a30bf1
   "id": "659305d4",
=======
   "id": "7196f059",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "7196f059",
=======
   "id": "24758370",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "9a875c0c",
||||||| 6a30bf1
   "id": "659305d4",
=======
   "id": "7196f059",
>>>>>>> documentation-devel
=======
   "id": "3b3ff067",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Build the pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "a44958b2",
||||||| 6a30bf1
   "id": "d786305d",
=======
   "id": "ad13ec25",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "ad13ec25",
=======
   "id": "a443fb07",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "a44958b2",
||||||| 6a30bf1
   "id": "d786305d",
=======
   "id": "ad13ec25",
>>>>>>> documentation-devel
=======
   "id": "78458e81",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "FA <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.mus_musculus.genome_dna.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)\n",
    "\n",
    "GTF <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.mus_musculus.annotation_gtf.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "3db4062f",
||||||| 6a30bf1
   "id": "d47a637f",
=======
   "id": "ce21316f",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "ce21316f",
=======
   "id": "dd6f0427",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "3db4062f",
||||||| 6a30bf1
   "id": "d47a637f",
=======
   "id": "ce21316f",
>>>>>>> documentation-devel
=======
   "id": "db88c1f9",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "args <- c(\n",
    "  \"-o out\", \n",
    "  \"-s spec.yaml\",\n",
    "  \"-m rna\",  \n",
    "  \"-fa\", FA,\n",
    "  \"-g\", GTF,\n",
    "  \"fastqs/R1.fastq.gz fastqs/R2.fastq.gz\")\n",
    "\n",
    "system2(command = \"cellatlas\", args = c(\"build\", args))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "fbad18de",
||||||| 6a30bf1
   "id": "00716eca",
=======
   "id": "f881e6ce",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "f881e6ce",
=======
   "id": "b6dccc5d",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "fbad18de",
||||||| 6a30bf1
   "id": "00716eca",
=======
   "id": "f881e6ce",
>>>>>>> documentation-devel
=======
   "id": "9410b025",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Run the pipeline\n",
    "To run the pipeline we simply extract the commands from `out/cellatlas_info.json` and run them on the command line. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "3880c740",
||||||| 6a30bf1
   "id": "c71b0ad1",
=======
   "id": "004486a2",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "004486a2",
=======
   "id": "6d09f65c",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "3880c740",
||||||| 6a30bf1
   "id": "c71b0ad1",
=======
   "id": "004486a2",
>>>>>>> documentation-devel
=======
   "id": "e871959b",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "cmds <- system2(\"jq\", \"-r '.commands[] | values[]' out/cellatlas_info.json\", stdout=TRUE)\n",
    "cmds <- str_subset(cmds, \"[\\\\[\\\\]]\", negate=TRUE)\n",
    "cmds <- str_extract(cmds, \"kb.*(txt|gz)\")\n",
    "\n",
    "lapply(cmds, function(cmd) system(cmd))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "c6b09af0",
||||||| 6a30bf1
   "id": "c68d07b4",
=======
   "id": "b116e692",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "b116e692",
=======
   "id": "0ca76f79",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "c6b09af0",
||||||| 6a30bf1
   "id": "c68d07b4",
=======
   "id": "b116e692",
>>>>>>> documentation-devel
=======
   "id": "d2dc002d",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Inspect the output \n",
    "We inspect the `out/run_info.json` and `out/kb_info.json` as a simple QC on the pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "a5cfc714",
||||||| 6a30bf1
   "id": "0ff37d4c",
=======
   "id": "460f61fb",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "460f61fb",
=======
   "id": "2f20d45d",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "a5cfc714",
||||||| 6a30bf1
   "id": "0ff37d4c",
=======
   "id": "460f61fb",
>>>>>>> documentation-devel
=======
   "id": "8106d81a",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "system(\"tree out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "991ea9c2",
||||||| 6a30bf1
   "id": "3bc0ac61",
=======
   "id": "15d2ef5d",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "15d2ef5d",
=======
   "id": "0c3bf954",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "991ea9c2",
||||||| 6a30bf1
   "id": "3bc0ac61",
=======
   "id": "15d2ef5d",
>>>>>>> documentation-devel
=======
   "id": "3db2950a",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"cat out/run_info.json\")\n",
    "# system(\"cat out/kb_info.json\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "8d639f0c",
||||||| 6a30bf1
   "id": "3eeffe30",
=======
   "id": "5b123f36",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "5b123f36",
=======
   "id": "67c65439",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "8d639f0c",
||||||| 6a30bf1
   "id": "3eeffe30",
=======
   "id": "5b123f36",
>>>>>>> documentation-devel
=======
   "id": "904cc31c",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## <a id=\"clicktag\"></a>Preprocessing for ClickTags\n",
    "The data for this example are located in the `cellatlas/examples/tag-clicktag/*` directory. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "2471b342",
||||||| 6a30bf1
   "id": "61bd644c",
=======
   "id": "cc5bf77f",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "cc5bf77f",
=======
   "id": "b25af9d8",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "2471b342",
||||||| 6a30bf1
   "id": "61bd644c",
=======
   "id": "cc5bf77f",
>>>>>>> documentation-devel
=======
   "id": "51383734",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"mv cellatlas/examples/tag-clicktag/* .\")\n",
    "system(\"gunzip 737K-august-2016.txt.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "412e1050",
||||||| 6a30bf1
   "id": "aee58fda",
=======
   "id": "5ba08969",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "5ba08969",
=======
   "id": "8451a412",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "412e1050",
||||||| 6a30bf1
   "id": "aee58fda",
=======
   "id": "5ba08969",
>>>>>>> documentation-devel
=======
   "id": "4d7cd009",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "The `seqspec print` command prints out an ordered tree representation of the sequenced elements contained in the FASTQ files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "589a94da",
||||||| 6a30bf1
   "id": "d1c89bcf",
=======
   "id": "0abf10d8",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "0abf10d8",
=======
   "id": "da085644",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "589a94da",
||||||| 6a30bf1
   "id": "d1c89bcf",
=======
   "id": "0abf10d8",
>>>>>>> documentation-devel
=======
   "id": "acf45c67",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"seqspec print spec.yaml\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "f84e37ce",
||||||| 6a30bf1
   "id": "2ba52896",
=======
   "id": "b34cb77b",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "b34cb77b",
=======
   "id": "2acf9621",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "f84e37ce",
||||||| 6a30bf1
   "id": "2ba52896",
=======
   "id": "b34cb77b",
>>>>>>> documentation-devel
=======
   "id": "ee70be82",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Fetch the references\n",
    "This step is only necessary if the modality that we are processing uses a transcriptome reference-based alignment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "2f30847a",
||||||| 6a30bf1
   "id": "2601fd31",
=======
   "id": "c1dc9a3c",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "c1dc9a3c",
=======
   "id": "fe578da6",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "2f30847a",
||||||| 6a30bf1
   "id": "2601fd31",
=======
   "id": "c1dc9a3c",
>>>>>>> documentation-devel
=======
   "id": "acd36964",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"gget ref -o ref.json -w dna,gtf homo_sapiens\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "4b41a5bf",
||||||| 6a30bf1
   "id": "5dc8fb44",
=======
   "id": "129cd5fa",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "129cd5fa",
=======
   "id": "67d9157d",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "4b41a5bf",
||||||| 6a30bf1
   "id": "5dc8fb44",
=======
   "id": "129cd5fa",
>>>>>>> documentation-devel
=======
   "id": "90f06ba2",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Build the pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "024071ac",
||||||| 6a30bf1
   "id": "da7e1210",
=======
   "id": "abc9e7d9",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "abc9e7d9",
=======
   "id": "af2f408d",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "024071ac",
||||||| 6a30bf1
   "id": "da7e1210",
=======
   "id": "abc9e7d9",
>>>>>>> documentation-devel
=======
   "id": "a72f86c7",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "FA <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.homo_sapiens.genome_dna.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)\n",
    "\n",
    "GTF <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.homo_sapiens.annotation_gtf.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "e8a6af65",
||||||| 6a30bf1
   "id": "04c0c7bf",
=======
   "id": "6a4299f2",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "6a4299f2",
=======
   "id": "e9f84bce",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "e8a6af65",
||||||| 6a30bf1
   "id": "04c0c7bf",
=======
   "id": "6a4299f2",
>>>>>>> documentation-devel
=======
   "id": "e18b5934",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "args <- c(\n",
    "  \"-o out\", \n",
    "  \"-s spec.yaml\",\n",
    "  \"-m tag\",  \n",
    "  \"-fa\", FA,\n",
    "  \"-g\", GTF,\n",
    "  \"-fb\", \"feature_barcodes.txt\",\n",
    "  \"fastqs/R1.fastq.gz fastqs/R2.fastq.gz\")\n",
    "\n",
    "system2(command = \"cellatlas\", args = c(\"build\", args))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "f049293b",
||||||| 6a30bf1
   "id": "9a14f6d8",
=======
   "id": "a13b4fa5",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "a13b4fa5",
=======
   "id": "2e1e042a",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "f049293b",
||||||| 6a30bf1
   "id": "9a14f6d8",
=======
   "id": "a13b4fa5",
>>>>>>> documentation-devel
=======
   "id": "83c62614",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Run the pipeline\n",
    "To run the pipeline we simply extract the commands from `out/cellatlas_info.json` and run them on the command line. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "0be2dae6",
||||||| 6a30bf1
   "id": "f8c52e95",
=======
   "id": "ceb30c87",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "ceb30c87",
=======
   "id": "f7c86acb",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "0be2dae6",
||||||| 6a30bf1
   "id": "f8c52e95",
=======
   "id": "ceb30c87",
>>>>>>> documentation-devel
=======
   "id": "6143a572",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "cmds <- system2(\"jq\", \"-r '.commands[] | values[]' out/cellatlas_info.json\", stdout=TRUE)\n",
    "cmds <- str_subset(cmds, \"[\\\\[\\\\]]\", negate=TRUE)\n",
    "cmds <- str_extract(cmds, \"kb.*(txt|gz)\")\n",
    "\n",
    "lapply(cmds, function(cmd) system(cmd))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "ebcdd79c",
||||||| 6a30bf1
   "id": "96368852",
=======
   "id": "39ff98a8",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "39ff98a8",
=======
   "id": "3177f01e",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "ebcdd79c",
||||||| 6a30bf1
   "id": "96368852",
=======
   "id": "39ff98a8",
>>>>>>> documentation-devel
=======
   "id": "25ab405b",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Inspect the output \n",
    "We inspect the `out/run_info.json` and `out/kb_info.json` as a simple QC on the pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "f2f13fe0",
||||||| 6a30bf1
   "id": "1abe631a",
=======
   "id": "2343c8b7",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "2343c8b7",
=======
   "id": "1599ca7d",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "f2f13fe0",
||||||| 6a30bf1
   "id": "1abe631a",
=======
   "id": "2343c8b7",
>>>>>>> documentation-devel
=======
   "id": "0f81de7b",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "system(\"tree out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "a2fd6586",
||||||| 6a30bf1
   "id": "0a02e63e",
=======
   "id": "29488fe2",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "29488fe2",
=======
   "id": "2b48dc12",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "a2fd6586",
||||||| 6a30bf1
   "id": "0a02e63e",
=======
   "id": "29488fe2",
>>>>>>> documentation-devel
=======
   "id": "02f300c2",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"cat out/run_info.json\")\n",
    "# system(\"cat out/kb_info.json\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "36913b72",
||||||| 6a30bf1
   "id": "710dbf18",
=======
   "id": "8b78672e",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "8b78672e",
=======
   "id": "eee321e2",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "36913b72",
||||||| 6a30bf1
   "id": "710dbf18",
=======
   "id": "8b78672e",
>>>>>>> documentation-devel
=======
   "id": "0e9ae649",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## <a id=\"rna_10xv3\"></a>Preprocessing for Chromium V3 Chemistry\n",
    "The data for this example are located in the `cellatlas/examples/rna-10xv3/` directory. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "73372844",
||||||| 6a30bf1
   "id": "35db72a6",
=======
   "id": "c9054989",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "c9054989",
=======
   "id": "a60f060c",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "73372844",
||||||| 6a30bf1
   "id": "35db72a6",
=======
   "id": "c9054989",
>>>>>>> documentation-devel
=======
   "id": "8a1ea529",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"mv cellatlas/examples/rna-10xv3/* .\")\n",
    "system(\"gunzip 3M-february-2018.txt.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "47150f1e",
||||||| 6a30bf1
   "id": "96051c5e",
=======
   "id": "5762883d",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "5762883d",
=======
   "id": "ffe8ca50",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "47150f1e",
||||||| 6a30bf1
   "id": "96051c5e",
=======
   "id": "5762883d",
>>>>>>> documentation-devel
=======
   "id": "3ca1521b",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "The `seqspec print` command prints out an ordered tree representation of the sequenced elements contained in the FASTQ files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "9e9f9d53",
||||||| 6a30bf1
   "id": "dfcd02f6",
=======
   "id": "4d20fb77",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "4d20fb77",
=======
   "id": "e753d04c",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "9e9f9d53",
||||||| 6a30bf1
   "id": "dfcd02f6",
=======
   "id": "4d20fb77",
>>>>>>> documentation-devel
=======
   "id": "88d13cd5",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"seqspec print spec.yaml\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "7dc85aa4",
||||||| 6a30bf1
   "id": "2ae18133",
=======
   "id": "ffc544ae",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "ffc544ae",
=======
   "id": "66ff5a83",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "7dc85aa4",
||||||| 6a30bf1
   "id": "2ae18133",
=======
   "id": "ffc544ae",
>>>>>>> documentation-devel
=======
   "id": "bb357995",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Fetch the references\n",
    "This step is only necessary if the modality that we are processing uses a transcriptome reference-based alignment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "2314f0bb",
||||||| 6a30bf1
   "id": "ebde192a",
=======
   "id": "9161c8a8",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "9161c8a8",
=======
   "id": "3b847561",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "2314f0bb",
||||||| 6a30bf1
   "id": "ebde192a",
=======
   "id": "9161c8a8",
>>>>>>> documentation-devel
=======
   "id": "57d3c49f",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"gget ref -o ref.json -w dna,gtf homo_sapiens\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "e99f1dd0",
||||||| 6a30bf1
   "id": "4a30fe52",
=======
   "id": "0911628b",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "0911628b",
=======
   "id": "9e9343dc",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "e99f1dd0",
||||||| 6a30bf1
   "id": "4a30fe52",
=======
   "id": "0911628b",
>>>>>>> documentation-devel
=======
   "id": "a6ec2b0f",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Build the pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "b01c40bf",
||||||| 6a30bf1
   "id": "e7d17b1d",
=======
   "id": "df712296",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "df712296",
=======
   "id": "6a3bb7c1",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "b01c40bf",
||||||| 6a30bf1
   "id": "e7d17b1d",
=======
   "id": "df712296",
>>>>>>> documentation-devel
=======
   "id": "48abe3ed",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "FA <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.homo_sapiens.genome_dna.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)\n",
    "\n",
    "GTF <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.homo_sapiens.annotation_gtf.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "f9f6b501",
||||||| 6a30bf1
   "id": "41be6efb",
=======
   "id": "6cbde5d4",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "6cbde5d4",
=======
   "id": "736d69ff",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "f9f6b501",
||||||| 6a30bf1
   "id": "41be6efb",
=======
   "id": "6cbde5d4",
>>>>>>> documentation-devel
=======
   "id": "2366f431",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "args <- c(\n",
    "  \"-o out\", \n",
    "  \"-s spec.yaml\",\n",
    "  \"-m rna\",  \n",
    "  \"-fa\", FA,\n",
    "  \"-g\", GTF,\n",
    "  \"-fb\", \"feature_barcodes.txt\",\n",
    "  \"fastqs/R1.fastq.gz fastqs/R2.fastq.gz\")\n",
    "\n",
    "system2(command = \"cellatlas\", args = c(\"build\", args))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "86b3cb51",
||||||| 6a30bf1
   "id": "51e80a01",
=======
   "id": "ac80114f",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "ac80114f",
=======
   "id": "39b5a573",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "86b3cb51",
||||||| 6a30bf1
   "id": "51e80a01",
=======
   "id": "ac80114f",
>>>>>>> documentation-devel
=======
   "id": "61b4c584",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Run the pipeline\n",
    "To run the pipeline we simply extract the commands from `out/cellatlas_info.json` and run them on the command line. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "e6c1df84",
||||||| 6a30bf1
   "id": "ed497c57",
=======
   "id": "4ce5b2b7",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "4ce5b2b7",
=======
   "id": "a11807d8",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "e6c1df84",
||||||| 6a30bf1
   "id": "ed497c57",
=======
   "id": "4ce5b2b7",
>>>>>>> documentation-devel
=======
   "id": "4e22680d",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "cmds <- system2(\"jq\", \"-r '.commands[] | values[]' out/cellatlas_info.json\", stdout=TRUE)\n",
    "cmds <- str_subset(cmds, \"[\\\\[\\\\]]\", negate=TRUE)\n",
    "cmds <- str_extract(cmds, \"kb.*(txt|gz)\")\n",
    "\n",
    "lapply(cmds, function(cmd) system(cmd))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "3d5e5d29",
||||||| 6a30bf1
   "id": "9c771515",
=======
   "id": "9983e885",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "9983e885",
=======
   "id": "3b42429f",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "3d5e5d29",
||||||| 6a30bf1
   "id": "9c771515",
=======
   "id": "9983e885",
>>>>>>> documentation-devel
=======
   "id": "f302e71a",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Inspect the output \n",
    "We inspect the `out/run_info.json` and `out/kb_info.json` as a simple QC on the pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "33099171",
||||||| 6a30bf1
   "id": "065b79eb",
=======
   "id": "e11874a5",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "e11874a5",
=======
   "id": "963943f2",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "33099171",
||||||| 6a30bf1
   "id": "065b79eb",
=======
   "id": "e11874a5",
>>>>>>> documentation-devel
=======
   "id": "ffa12c2c",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "system(\"tree out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "c52b8dbf",
||||||| 6a30bf1
   "id": "cb04b50e",
=======
   "id": "1ca201af",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "1ca201af",
=======
   "id": "0beb7c78",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "c52b8dbf",
||||||| 6a30bf1
   "id": "cb04b50e",
=======
   "id": "1ca201af",
>>>>>>> documentation-devel
=======
   "id": "78be5610",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"cat out/run_info.json\")\n",
    "# system(\"cat out/kb_info.json\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "701f6cce",
||||||| 6a30bf1
   "id": "b3dcb069",
=======
   "id": "7179f200",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "7179f200",
=======
   "id": "4e37b0f1",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "701f6cce",
||||||| 6a30bf1
   "id": "b3dcb069",
=======
   "id": "7179f200",
>>>>>>> documentation-devel
=======
   "id": "c7d3e1a2",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## <a id=\"atac_10x\"></a>Preprocessing for Chromium Single Cell ATAC-seq\n",
    "The data for this example are located in the `cellatlas/examples/atac-10xatac/` directory. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "3e0f1738",
||||||| 6a30bf1
   "id": "71a77f5b",
=======
   "id": "8061e2d4",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "8061e2d4",
=======
   "id": "8f52459a",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "3e0f1738",
||||||| 6a30bf1
   "id": "71a77f5b",
=======
   "id": "8061e2d4",
>>>>>>> documentation-devel
=======
   "id": "b3f77d6b",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"mv cellatlas/examples/atac-10xatac/* .\")\n",
    "system(\"gunzip *.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "5cc03e26",
||||||| 6a30bf1
   "id": "e80a2203",
=======
   "id": "fb07fc2e",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "fb07fc2e",
=======
   "id": "c10dc459",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "5cc03e26",
||||||| 6a30bf1
   "id": "e80a2203",
=======
   "id": "fb07fc2e",
>>>>>>> documentation-devel
=======
   "id": "17744402",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "The `seqspec print` command prints out an ordered tree representation of the sequenced elements contained in the FASTQ files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "2231ebce",
||||||| 6a30bf1
   "id": "24f260b6",
=======
   "id": "95e3ac8f",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "95e3ac8f",
=======
   "id": "60469724",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "2231ebce",
||||||| 6a30bf1
   "id": "24f260b6",
=======
   "id": "95e3ac8f",
>>>>>>> documentation-devel
=======
   "id": "34dd83ca",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"seqspec print spec.yaml\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "debb35b9",
||||||| 6a30bf1
   "id": "f1258f76",
=======
   "id": "040cba8a",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "040cba8a",
=======
   "id": "a3869c92",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "debb35b9",
||||||| 6a30bf1
   "id": "f1258f76",
=======
   "id": "040cba8a",
>>>>>>> documentation-devel
=======
   "id": "119fcf83",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Fetch the references\n",
    "This step is only necessary if the modality that we are processing uses a transcriptome reference-based alignment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "ac913bab",
||||||| 6a30bf1
   "id": "dc08385b",
=======
   "id": "4ed31637",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "4ed31637",
=======
   "id": "2c252a24",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "ac913bab",
||||||| 6a30bf1
   "id": "dc08385b",
=======
   "id": "4ed31637",
>>>>>>> documentation-devel
=======
   "id": "8098c48f",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"gget ref -o ref.json -w dna,gtf homo_sapiens\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "2904c2f7",
||||||| 6a30bf1
   "id": "47541360",
=======
   "id": "528093b8",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "528093b8",
=======
   "id": "97b4639e",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "2904c2f7",
||||||| 6a30bf1
   "id": "47541360",
=======
   "id": "528093b8",
>>>>>>> documentation-devel
=======
   "id": "6e869d4c",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Build the pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "aacfe652",
||||||| 6a30bf1
   "id": "60bf3601",
=======
   "id": "606e2983",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "606e2983",
=======
   "id": "70e59d16",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "aacfe652",
||||||| 6a30bf1
   "id": "60bf3601",
=======
   "id": "606e2983",
>>>>>>> documentation-devel
=======
   "id": "71b46b0a",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "FA <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.homo_sapiens.genome_dna.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)\n",
    "\n",
    "GTF <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.homo_sapiens.annotation_gtf.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "e57b3b95",
||||||| 6a30bf1
   "id": "1358100b",
=======
   "id": "0c9fb9f3",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "0c9fb9f3",
=======
   "id": "6bebcc0d",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "e57b3b95",
||||||| 6a30bf1
   "id": "1358100b",
=======
   "id": "0c9fb9f3",
>>>>>>> documentation-devel
=======
   "id": "6aa44cfe",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "args <- c(\n",
    "  \"-o out\", \n",
    "  \"-s spec.yaml\",\n",
    "  \"-m atac\",  \n",
    "  \"-fa\", FA,\n",
    "  \"-g\", GTF,\n",
    "  \"fastqs/R1.fastq.gz fastqs/R2.fastq.gz fastqs/I2.fastq.gz\")\n",
    "\n",
    "system2(command = \"cellatlas\", args = c(\"build\", args))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "9cfd5370",
||||||| 6a30bf1
   "id": "a52250ea",
=======
   "id": "66324f47",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "66324f47",
=======
   "id": "2688a9af",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "9cfd5370",
||||||| 6a30bf1
   "id": "a52250ea",
=======
   "id": "66324f47",
>>>>>>> documentation-devel
=======
   "id": "69941832",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Run the pipeline\n",
    "To run the pipeline we extract the commands from `out/cellatlas_info.json` and run them on the command line. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "54f7e908",
||||||| 6a30bf1
   "id": "326d6371",
=======
   "id": "6e52603a",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "6e52603a",
=======
   "id": "ffee5af0",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "54f7e908",
||||||| 6a30bf1
   "id": "326d6371",
=======
   "id": "6e52603a",
>>>>>>> documentation-devel
=======
   "id": "e39a1be8",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "cmds <- system2(\"jq\", \"-r '.commands[] | values[]' out/cellatlas_info.json\", stdout=TRUE)\n",
    "cmds <- str_subset(cmds, \"[\\\\[\\\\]]\", negate=TRUE)\n",
    "cmds <- str_trim(cmds)\n",
    "cmds <- str_remove_all(cmds, '\\\\\\\",$|\\\\\\\"$|^\\\\\\\"')\n",
    "cmds <- str_replace_all(cmds, fixed(\"\\\\\\\"\"), \"\\\"\")\n",
    "cmds <- str_replace_all(cmds, fixed(\"\\\\t\"), \"\\t\")\n",
    "\n",
    "cmds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "914832e7",
||||||| 6a30bf1
   "id": "528ba5c6",
=======
   "id": "1320bfd1",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "1320bfd1",
=======
   "id": "1c10e27b",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "914832e7",
||||||| 6a30bf1
   "id": "528ba5c6",
=======
   "id": "1320bfd1",
>>>>>>> documentation-devel
=======
   "id": "1cf14fc8",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "lapply(cmds, function(cmd) system(cmd))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "2ebdc78e",
||||||| 6a30bf1
   "id": "6363149e",
=======
   "id": "58fd2a77",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "58fd2a77",
=======
   "id": "d68fd5c0",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "2ebdc78e",
||||||| 6a30bf1
   "id": "6363149e",
=======
   "id": "58fd2a77",
>>>>>>> documentation-devel
=======
   "id": "4c3ac838",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Inspect the output \n",
    "We inspect the `out/run_info.json` and `out/kb_info.json` as a simple QC on the pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "1eeec679",
||||||| 6a30bf1
   "id": "a1bce11f",
=======
   "id": "bcdac4d9",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "bcdac4d9",
=======
   "id": "03a6eb6a",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "1eeec679",
||||||| 6a30bf1
   "id": "a1bce11f",
=======
   "id": "bcdac4d9",
>>>>>>> documentation-devel
=======
   "id": "52b4fc3b",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "system(\"tree out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "fd1bf2d1",
||||||| 6a30bf1
   "id": "63dc55fb",
=======
   "id": "0a748917",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "0a748917",
=======
   "id": "c6aca22f",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "fd1bf2d1",
||||||| 6a30bf1
   "id": "63dc55fb",
=======
   "id": "0a748917",
>>>>>>> documentation-devel
=======
   "id": "9aeb9d02",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"cat out/run_info.json\")\n",
    "# system(\"cat out/kb_info.json\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "c102a33f",
||||||| 6a30bf1
   "id": "1b886df4",
=======
   "id": "34acd2e8",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "34acd2e8",
=======
   "id": "52c316b6",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "c102a33f",
||||||| 6a30bf1
   "id": "1b886df4",
=======
   "id": "34acd2e8",
>>>>>>> documentation-devel
=======
   "id": "0c3d5191",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## <a id=\"atac_10xmulti\"></a>Preprocessing for Chromium Single Cell ATAC Multiome ATAC\n",
    "The data for this example are located in the `cellatlas/examples/atac-10xmultiome/` directory. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "a63c350b",
||||||| 6a30bf1
   "id": "4395ccaa",
=======
   "id": "c8d9c498",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "c8d9c498",
=======
   "id": "8a980513",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "a63c350b",
||||||| 6a30bf1
   "id": "4395ccaa",
=======
   "id": "c8d9c498",
>>>>>>> documentation-devel
=======
   "id": "83887f07",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"mv cellatlas/examples/atac-10xmultiome/* .\")\n",
    "system(\"gunzip *.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "f08562ad",
||||||| 6a30bf1
   "id": "958ce5d8",
=======
   "id": "6342b1f0",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "6342b1f0",
=======
   "id": "5c1a137f",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "f08562ad",
||||||| 6a30bf1
   "id": "958ce5d8",
=======
   "id": "6342b1f0",
>>>>>>> documentation-devel
=======
   "id": "ab954c4f",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "The `seqspec print` command prints out an ordered tree representation of the sequenced elements contained in the FASTQ files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "3a032c0c",
||||||| 6a30bf1
   "id": "a3081bb5",
=======
   "id": "2a4ce92a",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "2a4ce92a",
=======
   "id": "12694fd2",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "3a032c0c",
||||||| 6a30bf1
   "id": "a3081bb5",
=======
   "id": "2a4ce92a",
>>>>>>> documentation-devel
=======
   "id": "ab9e06d2",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"seqspec print spec.yaml\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "b81189a1",
||||||| 6a30bf1
   "id": "ffba2925",
=======
   "id": "249ba152",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "249ba152",
=======
   "id": "388e15e8",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "b81189a1",
||||||| 6a30bf1
   "id": "ffba2925",
=======
   "id": "249ba152",
>>>>>>> documentation-devel
=======
   "id": "9a299c49",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Fetch the references\n",
    "This step is only necessary if the modality that we are processing uses a transcriptome reference-based alignment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "fea48029",
||||||| 6a30bf1
   "id": "f0dd4d2f",
=======
   "id": "9ac3b008",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "9ac3b008",
=======
   "id": "98697010",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "fea48029",
||||||| 6a30bf1
   "id": "f0dd4d2f",
=======
   "id": "9ac3b008",
>>>>>>> documentation-devel
=======
   "id": "040955f2",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"gget ref -o ref.json -w dna,gtf mus_musculus\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "79145991",
||||||| 6a30bf1
   "id": "6b1646b0",
=======
   "id": "b362ec60",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "b362ec60",
=======
   "id": "1398a809",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "79145991",
||||||| 6a30bf1
   "id": "6b1646b0",
=======
   "id": "b362ec60",
>>>>>>> documentation-devel
=======
   "id": "2c072e8a",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Build the pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "bb79119f",
||||||| 6a30bf1
   "id": "2e420ec7",
=======
   "id": "80af1338",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "80af1338",
=======
   "id": "25d34bd4",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "bb79119f",
||||||| 6a30bf1
   "id": "2e420ec7",
=======
   "id": "80af1338",
>>>>>>> documentation-devel
=======
   "id": "77bca7c3",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "FA <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.mus_musculus.genome_dna.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)\n",
    "\n",
    "GTF <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.mus_musculus.annotation_gtf.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "f7ff0fc4",
||||||| 6a30bf1
   "id": "e0bd88d5",
=======
   "id": "8270d7e8",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "8270d7e8",
=======
   "id": "05bf2a7c",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "f7ff0fc4",
||||||| 6a30bf1
   "id": "e0bd88d5",
=======
   "id": "8270d7e8",
>>>>>>> documentation-devel
=======
   "id": "6e81566f",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "args <- c(\n",
    "  \"-o out\", \n",
    "  \"-s spec.yaml\",\n",
    "  \"-m atac\",  \n",
    "  \"-fa\", FA,\n",
    "  \"-g\", GTF,\n",
    "  \"fastqs/atac_R1.fastq.gz fastqs/atac_R2.fastq.gz fastqs/atac_I2.fastq.gz\")\n",
    "\n",
    "system2(command = \"cellatlas\", args = c(\"build\", args))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "3f6e07f3",
||||||| 6a30bf1
   "id": "8bfc13d2",
=======
   "id": "fbd73766",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "fbd73766",
=======
   "id": "b4a82170",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "3f6e07f3",
||||||| 6a30bf1
   "id": "8bfc13d2",
=======
   "id": "fbd73766",
>>>>>>> documentation-devel
=======
   "id": "cee5ed2b",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Run the pipeline\n",
    "To run the pipeline we extract the commands from `out/cellatlas_info.json` and run them on the command line. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "27c02fa8",
||||||| 6a30bf1
   "id": "35310769",
=======
   "id": "1abedbe7",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "1abedbe7",
=======
   "id": "3eab2042",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "27c02fa8",
||||||| 6a30bf1
   "id": "35310769",
=======
   "id": "1abedbe7",
>>>>>>> documentation-devel
=======
   "id": "2d40a22c",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "cmds <- system2(\"jq\", \"-r '.commands[] | values[]' out/cellatlas_info.json\", stdout=TRUE)\n",
    "cmds <- str_subset(cmds, \"[\\\\[\\\\]]\", negate=TRUE)\n",
    "cmds <- str_trim(cmds)\n",
    "cmds <- str_remove_all(cmds, '\\\\\\\",$|\\\\\\\"$|^\\\\\\\"')\n",
    "cmds <- str_replace_all(cmds, fixed(\"\\\\\\\"\"), \"\\\"\")\n",
    "cmds <- str_replace_all(cmds, fixed(\"\\\\t\"), \"\\t\")\n",
    "\n",
    "cmds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "346e9bf6",
||||||| 6a30bf1
   "id": "4b5e0aa2",
=======
   "id": "f8d79313",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "f8d79313",
=======
   "id": "4327bcbf",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "346e9bf6",
||||||| 6a30bf1
   "id": "4b5e0aa2",
=======
   "id": "f8d79313",
>>>>>>> documentation-devel
=======
   "id": "d3dc846b",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "lapply(cmds, function(cmd) system(cmd))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "190291c1",
||||||| 6a30bf1
   "id": "a9e62e79",
=======
   "id": "1b946b38",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "1b946b38",
=======
   "id": "97e48b6f",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "190291c1",
||||||| 6a30bf1
   "id": "a9e62e79",
=======
   "id": "1b946b38",
>>>>>>> documentation-devel
=======
   "id": "860c7d8c",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Inspect the output \n",
    "We inspect the `out/run_info.json` and `out/kb_info.json` as a simple QC on the pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "57967ae0",
||||||| 6a30bf1
   "id": "13f6e530",
=======
   "id": "a27b4b05",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "a27b4b05",
=======
   "id": "3122a77a",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "57967ae0",
||||||| 6a30bf1
   "id": "13f6e530",
=======
   "id": "a27b4b05",
>>>>>>> documentation-devel
=======
   "id": "b37e2930",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "system(\"tree out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "19218f1c",
||||||| 6a30bf1
   "id": "31c7a899",
=======
   "id": "60015ada",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "60015ada",
=======
   "id": "486a137e",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "19218f1c",
||||||| 6a30bf1
   "id": "31c7a899",
=======
   "id": "60015ada",
>>>>>>> documentation-devel
=======
   "id": "6a51e462",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"cat out/run_info.json\")\n",
    "# system(\"cat out/kb_info.json\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "756ef0aa",
||||||| 6a30bf1
   "id": "39864044",
=======
   "id": "0543d7d1",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "0543d7d1",
=======
   "id": "2405251d",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "756ef0aa",
||||||| 6a30bf1
   "id": "39864044",
=======
   "id": "0543d7d1",
>>>>>>> documentation-devel
=======
   "id": "c8b2a3ee",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## <a id=\"crispr_10x\"></a>Preprocessing for Chromium Single Cell CRISPR Screening \n",
    "The data for this example are located in the `cellatlas/examples/crispr-10xcrispr/` directory. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "4e48067c",
||||||| 6a30bf1
   "id": "5da15242",
=======
   "id": "e507210d",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "e507210d",
=======
   "id": "50186cc3",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "4e48067c",
||||||| 6a30bf1
   "id": "5da15242",
=======
   "id": "e507210d",
>>>>>>> documentation-devel
=======
   "id": "9b134408",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"mv cellatlas/examples/crispr-10xcrispr/* .\")\n",
    "system(\"gunzip *.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "1b6c01e0",
||||||| 6a30bf1
   "id": "b074fc91",
=======
   "id": "735a8752",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "735a8752",
=======
   "id": "9719b016",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "1b6c01e0",
||||||| 6a30bf1
   "id": "b074fc91",
=======
   "id": "735a8752",
>>>>>>> documentation-devel
=======
   "id": "5323da94",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "The `seqspec print` command prints out an ordered tree representation of the sequenced elements contained in the FASTQ files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "273a26f9",
||||||| 6a30bf1
   "id": "83d5812d",
=======
   "id": "a992535f",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "a992535f",
=======
   "id": "353fa04d",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "273a26f9",
||||||| 6a30bf1
   "id": "83d5812d",
=======
   "id": "a992535f",
>>>>>>> documentation-devel
=======
   "id": "34fe1935",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"seqspec print spec.yaml\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "193edb24",
||||||| 6a30bf1
   "id": "f4c8ca53",
=======
   "id": "b3a92a6f",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "b3a92a6f",
=======
   "id": "1e66e939",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "193edb24",
||||||| 6a30bf1
   "id": "f4c8ca53",
=======
   "id": "b3a92a6f",
>>>>>>> documentation-devel
=======
   "id": "5ce04d09",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Fetch the references\n",
    "This step is only necessary if the modality that we are processing uses a transcriptome reference-based alignment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "01a04c0b",
||||||| 6a30bf1
   "id": "15e15dbe",
=======
   "id": "84c1b64e",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "84c1b64e",
=======
   "id": "67b0a487",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "01a04c0b",
||||||| 6a30bf1
   "id": "15e15dbe",
=======
   "id": "84c1b64e",
>>>>>>> documentation-devel
=======
   "id": "97cfd934",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"gget ref -o ref.json -w dna,gtf homo_sapiens\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "cb19000d",
||||||| 6a30bf1
   "id": "74cbde9c",
=======
   "id": "0375a096",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "0375a096",
=======
   "id": "733c1496",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "cb19000d",
||||||| 6a30bf1
   "id": "74cbde9c",
=======
   "id": "0375a096",
>>>>>>> documentation-devel
=======
   "id": "dc1d08b7",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Build the pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "e9a033c6",
||||||| 6a30bf1
   "id": "5e4d58b7",
=======
   "id": "8b8313c5",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "8b8313c5",
=======
   "id": "3299df18",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "e9a033c6",
||||||| 6a30bf1
   "id": "5e4d58b7",
=======
   "id": "8b8313c5",
>>>>>>> documentation-devel
=======
   "id": "189489ce",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "FA <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.homo_sapiens.genome_dna.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)\n",
    "\n",
    "GTF <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.homo_sapiens.annotation_gtf.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "5df11cce",
||||||| 6a30bf1
   "id": "4ad148d6",
=======
   "id": "18c66e26",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "18c66e26",
=======
   "id": "8c706539",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "5df11cce",
||||||| 6a30bf1
   "id": "4ad148d6",
=======
   "id": "18c66e26",
>>>>>>> documentation-devel
=======
   "id": "9a1aaf3c",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "args <- c(\n",
    "  \"-o out\", \n",
    "  \"-s spec.yaml\",\n",
    "  \"-m crispr\",  \n",
    "  \"-fa\", FA,\n",
    "  \"-g\", GTF,\n",
    "  \"-fb\", \"feature_barcodes.txt\",\n",
    "  \"fastqs/R1.fastq.gz fastqs/R2.fastq.gz\")\n",
    "\n",
    "system2(command = \"cellatlas\", args = c(\"build\", args))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "77958983",
||||||| 6a30bf1
   "id": "5aa976d5",
=======
   "id": "47f457eb",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "47f457eb",
=======
   "id": "7b1ef3d0",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "77958983",
||||||| 6a30bf1
   "id": "5aa976d5",
=======
   "id": "47f457eb",
>>>>>>> documentation-devel
=======
   "id": "788c7f93",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Run the pipeline\n",
    "To run the pipeline we extract the commands from `out/cellatlas_info.json` and run them on the command line. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "70f724db",
||||||| 6a30bf1
   "id": "dc3749d6",
=======
   "id": "c3f54d30",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "c3f54d30",
=======
   "id": "1fb48bc3",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "70f724db",
||||||| 6a30bf1
   "id": "dc3749d6",
=======
   "id": "c3f54d30",
>>>>>>> documentation-devel
=======
   "id": "9b813856",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "cmds <- system2(\"jq\", \"-r '.commands[] | values[]' out/cellatlas_info.json\", stdout=TRUE)\n",
    "cmds <- str_subset(cmds, \"[\\\\[\\\\]]\", negate=TRUE)\n",
    "cmds <- str_extract(cmds, \"kb.*(txt|gz)\")\n",
    "\n",
    "cmds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "a63bb88d",
||||||| 6a30bf1
   "id": "04b771c6",
=======
   "id": "aa77b956",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "aa77b956",
=======
   "id": "4a60a836",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "a63bb88d",
||||||| 6a30bf1
   "id": "04b771c6",
=======
   "id": "aa77b956",
>>>>>>> documentation-devel
=======
   "id": "2dd6f920",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "lapply(cmds, function(cmd) system(cmd))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "8c460985",
||||||| 6a30bf1
   "id": "e271f462",
=======
   "id": "9496aeae",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "9496aeae",
=======
   "id": "536dc5fc",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "8c460985",
||||||| 6a30bf1
   "id": "e271f462",
=======
   "id": "9496aeae",
>>>>>>> documentation-devel
=======
   "id": "8b402901",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Inspect the output \n",
    "We inspect the `out/run_info.json` and `out/kb_info.json` as a simple QC on the pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "6cd43675",
||||||| 6a30bf1
   "id": "bfecf96b",
=======
   "id": "82bb4ef9",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "82bb4ef9",
=======
   "id": "e644b63f",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "6cd43675",
||||||| 6a30bf1
   "id": "bfecf96b",
=======
   "id": "82bb4ef9",
>>>>>>> documentation-devel
=======
   "id": "edb90b79",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "system(\"tree out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "d93e6c22",
||||||| 6a30bf1
   "id": "7209758d",
=======
   "id": "388e9fe9",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "388e9fe9",
=======
   "id": "a1023893",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "d93e6c22",
||||||| 6a30bf1
   "id": "7209758d",
=======
   "id": "388e9fe9",
>>>>>>> documentation-devel
=======
   "id": "9b21da03",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"cat out/run_info.json\")\n",
    "# system(\"cat out/kb_info.json\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "9c75170a",
||||||| 6a30bf1
   "id": "a32ee301",
=======
   "id": "97ff891c",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "97ff891c",
=======
   "id": "873d470c",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "9c75170a",
||||||| 6a30bf1
   "id": "a32ee301",
=======
   "id": "97ff891c",
>>>>>>> documentation-devel
=======
   "id": "d4ec5886",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## <a id=\"#nuclei\"></a>Preprocessing for Chromium Nuclei Isolation\n",
    "The data for this example are located in the `cellatlas/examples/rna-10xv3-nuclei/` directory. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "6adb411a",
||||||| 6a30bf1
   "id": "02c7e4e7",
=======
   "id": "84b8288a",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "84b8288a",
=======
   "id": "73e5cdbf",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "6adb411a",
||||||| 6a30bf1
   "id": "02c7e4e7",
=======
   "id": "84b8288a",
>>>>>>> documentation-devel
=======
   "id": "236fd499",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"mv cellatlas/examples/rna-10xv3-nuclei/* .\")\n",
    "system(\"gunzip *.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "f17e2743",
||||||| 6a30bf1
   "id": "6e65cc5f",
=======
   "id": "4634c9d1",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "4634c9d1",
=======
   "id": "dedbdfe7",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "f17e2743",
||||||| 6a30bf1
   "id": "6e65cc5f",
=======
   "id": "4634c9d1",
>>>>>>> documentation-devel
=======
   "id": "a77d7533",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "The `seqspec print` command prints out an ordered tree representation of the sequenced elements contained in the FASTQ files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "abd5a2fa",
||||||| 6a30bf1
   "id": "57d49470",
=======
   "id": "32d10bea",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "32d10bea",
=======
   "id": "ebe4267d",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "abd5a2fa",
||||||| 6a30bf1
   "id": "57d49470",
=======
   "id": "32d10bea",
>>>>>>> documentation-devel
=======
   "id": "91e95a27",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"seqspec print spec.yaml\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "0375559b",
||||||| 6a30bf1
   "id": "3c9b70a5",
=======
   "id": "82d39ba0",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "82d39ba0",
=======
   "id": "1527a25a",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "0375559b",
||||||| 6a30bf1
   "id": "3c9b70a5",
=======
   "id": "82d39ba0",
>>>>>>> documentation-devel
=======
   "id": "86d17f09",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Fetch the references\n",
    "This step is only necessary if the modality that we are processing uses a transcriptome reference-based alignment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "81491726",
||||||| 6a30bf1
   "id": "54b0706f",
=======
   "id": "78e3a24c",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "78e3a24c",
=======
   "id": "ac85202a",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "81491726",
||||||| 6a30bf1
   "id": "54b0706f",
=======
   "id": "78e3a24c",
>>>>>>> documentation-devel
=======
   "id": "e09799e9",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "system(\"gget ref -o ref.json -w dna,gtf homo_sapiens\")"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "e7f61b28",
||||||| 6a30bf1
   "id": "255ea092",
=======
   "id": "74f98b98",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "74f98b98",
=======
   "id": "5b309e1c",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "e7f61b28",
||||||| 6a30bf1
   "id": "255ea092",
=======
   "id": "74f98b98",
>>>>>>> documentation-devel
=======
   "id": "4d3a210c",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Build the pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "315714f5",
||||||| 6a30bf1
   "id": "c2b0191f",
=======
   "id": "701048f2",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "701048f2",
=======
   "id": "21bb36a8",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "315714f5",
||||||| 6a30bf1
   "id": "c2b0191f",
=======
   "id": "701048f2",
>>>>>>> documentation-devel
=======
   "id": "f08bf2bf",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "FA <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.homo_sapiens.genome_dna.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)\n",
    "\n",
    "GTF <- system2(\"jq\",\n",
    "  args = c(\"-r\", \"'.homo_sapiens.annotation_gtf.ftp'\", \"ref.json\"),\n",
    "  stdout = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "b095e01f",
||||||| 6a30bf1
   "id": "39cdf4d2",
=======
   "id": "404a1ac6",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "404a1ac6",
=======
   "id": "d24fa602",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "b095e01f",
||||||| 6a30bf1
   "id": "39cdf4d2",
=======
   "id": "404a1ac6",
>>>>>>> documentation-devel
=======
   "id": "4152d4ce",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "args <- c(\n",
    "  \"-o out\", \n",
    "  \"-s spec.yaml\",\n",
    "  \"-m rna\",  \n",
    "  \"-fb feature_barcodes.txt\",\n",
    "  \"-fa\", FA,\n",
    "  \"-g\", GTF,\n",
    "  \"fastqs/R1.fastq.gz fastqs/R2.fastq.gz\")\n",
    "\n",
    "system2(command = \"cellatlas\", args = c(\"build\", args))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "76e750cd",
||||||| 6a30bf1
   "id": "0a1e963f",
=======
   "id": "33f4a14c",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "33f4a14c",
=======
   "id": "0e4590af",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "76e750cd",
||||||| 6a30bf1
   "id": "0a1e963f",
=======
   "id": "33f4a14c",
>>>>>>> documentation-devel
=======
   "id": "2f5e5f58",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Run the pipeline\n",
    "To run the pipeline we extract the commands from `out/cellatlas_info.json` and run them on the command line. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "759d1355",
||||||| 6a30bf1
   "id": "02f208a2",
=======
   "id": "1fc722ec",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "1fc722ec",
=======
   "id": "f355b56f",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "759d1355",
||||||| 6a30bf1
   "id": "02f208a2",
=======
   "id": "1fc722ec",
>>>>>>> documentation-devel
=======
   "id": "1d51bcc6",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "cmds <- system2(\"jq\", \"-r '.commands[] | values[]' out/cellatlas_info.json\", stdout=TRUE)\n",
    "cmds <- str_subset(cmds, \"[\\\\[\\\\]]\", negate=TRUE)\n",
    "cmds <- str_extract(cmds, \"kb.*(txt|gz)\")\n",
    "\n",
    "cmds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "98ad8b75",
||||||| 6a30bf1
   "id": "ca6b68bf",
=======
   "id": "03940d2b",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "03940d2b",
=======
   "id": "3f256cd4",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "98ad8b75",
||||||| 6a30bf1
   "id": "ca6b68bf",
=======
   "id": "03940d2b",
>>>>>>> documentation-devel
=======
   "id": "a946229f",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "lapply(cmds, function(cmd) system(cmd))"
   ]
  },
  {
   "cell_type": "markdown",
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "7a2f3eae",
||||||| 6a30bf1
   "id": "052b0cee",
=======
   "id": "c9f98382",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "c9f98382",
=======
   "id": "36b04d4a",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "7a2f3eae",
||||||| 6a30bf1
   "id": "052b0cee",
=======
   "id": "c9f98382",
>>>>>>> documentation-devel
=======
   "id": "a34afe31",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Inspect the output \n",
    "We inspect the `out/run_info.json` and `out/kb_info.json` as a simple QC on the pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "fad73690",
||||||| 6a30bf1
   "id": "342e6d6e",
=======
   "id": "9ae1e4b6",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "9ae1e4b6",
=======
   "id": "914e9890",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "fad73690",
||||||| 6a30bf1
   "id": "342e6d6e",
=======
   "id": "9ae1e4b6",
>>>>>>> documentation-devel
=======
   "id": "97998d4a",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "system(\"tree out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
   "id": "f4d55ce2",
||||||| 6a30bf1
   "id": "4f2067da",
=======
   "id": "ef74bfff",
>>>>>>> documentation-devel
||||||| d28b18c
   "id": "ef74bfff",
=======
   "id": "bf0cb631",
>>>>>>> documentation-devel
||||||| 6893f63
<<<<<<< HEAD
   "id": "f4d55ce2",
||||||| 6a30bf1
   "id": "4f2067da",
=======
   "id": "ef74bfff",
>>>>>>> documentation-devel
=======
   "id": "480b43ae",
>>>>>>> c74c98c2450c2a1ba922ed9442f136238d6bc88b
   "metadata": {
    "eval": false
   },
   "outputs": [],
   "source": [
    "system(\"cat out/run_info.json\")\n",
    "# system(\"cat out/kb_info.json\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
