---
title: "ClickTags preprocessing with cellatlas"
author: "Kayla Jackson and A. Sina Booeshaghi"
date: "`r Sys.Date()`"
output: 
  html_document:
  toc: true
  toc_float:
    collapsed: false
  toc_depth: 3
jupyter:
  kernelspec:
    display_name: R
    language: R
    name: ir
bibliography: ref.bib
vignette: >
  %\VignetteIndexEntry{clicktag_ca}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

```{r, include = FALSE, eval = FALSE}
# Install Google Colab dependencies
# Note: this can take 30+ minutes (many of the dependencies include C++ code, which needs to be compiled)

# First install `sf`, `ragg` and `textshaping` and their system dependencies:
system("apt-get -y update && apt-get install -y  libudunits2-dev libgdal-dev libgeos-dev libproj-dev libharfbuzz-dev libfribidi-dev")

install.packages("stringr")
```

```{r setup}
library(stringr)
```

# Building Count Matrices with `cellatlas`
A major challenge in uniformly preprocessing large amounts of single-cell genomics data from a variety of different assays is identifying and handling sequenced elements in a coherent and consistent fashion. Cell barcodes in reads from RNAseq data from 10x Multiome, for example, must be extracted and error corrected in the manner as cell barcodes in reads from ATACseq data from 10x Multiome so that barcode-barcode registration can occur. Uniform processing in this way minimzes computational variability and enables cross-assay comparisons.

In this notebook we demonstrate how single-cell genomics data can be preprocessed to generate a cell by feature count matrix. This requires:

1. FASTQ files
2. `seqspec` specification for the FASTQ files
3. Genome Sequence FASTA
4. Genome Annotation GTF
5. (optional) Feature barcode list

# Install Packages
The vignette makes use of two non-standard command line tools, [`jq`](https://jqlang.github.io/jq/) and [`tree`](https://mama.indstate.edu/users/ice/tree/). The code cell below installs these tools on a Linux operating system and should be updated for Mac and Windows users.
```{r}
# Install `tree` to view files
system("wget --quiet --show-progress ftp://mama.indstate.edu/linux/tree/tree-2.1.0.tgz")

# Install `jq`, a command-line tool for extracting key value pairs from JSON files 
system("wget --quiet --show-progress https://github.com/stedolan/jq/releases/download/jq-1.6/jq-osx-amd64")
system("chmod +x jq-linux64 && mv jq-linux64 /usr/local/bin/jq")
```
We will continue with other dependencies that can be installed on any operating system. 
```{r}
# Clone the cellatlas repo and install the package
system("git clone https://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@github.com/cellatlas/cellatlas.git")
system("cd cellatlas && pip install .")

# Install dependencies
system("yes | pip uninstall --quiet seqspec")
system("pip install --quiet git+https://github.com/IGVF/seqspec.git")
system("pip install --quiet gget kb-python")
```

# Preprocessing for ClickTags
The data for this example are located in the `cellatlas/examples/tag-clicktag/*` directory. 
```{r, eval = FALSE}
system("mv cellatlas/examples/tag-clicktag/* .")
system("gunzip 737K-august-2016.txt.gz")
```
The `seqspec print` command prints out an ordered tree representation of the sequenced elements contained in the FASTQ files.
```{r, eval = FALSE}
system("seqspec print spec.yaml")
```
## Fetch the references
This step is only necessary if the modality that we are processing uses a transcriptome reference-based alignment. 
```{r, eval = FALSE}
system("gget ref -o ref.json -w dna,gtf homo_sapiens")
```
## Build the pipeline
```{r, eval = FALSE}
FA <- system2("jq",
  args = c("-r", "'.homo_sapiens.genome_dna.ftp'", "ref.json"),
  stdout = TRUE)

GTF <- system2("jq",
  args = c("-r", "'.homo_sapiens.annotation_gtf.ftp'", "ref.json"),
  stdout = TRUE)
```

```{r, eval = FALSE}
args <- c(
  "-o out", 
  "-s spec.yaml",
  "-m tag",  
  "-fa", FA,
  "-g", GTF,
  "-fb", "feature_barcodes.txt",
  "fastqs/R1.fastq.gz fastqs/R2.fastq.gz")

system2(command = "cellatlas", args = c("build", args))
```
## Run the pipeline
To run the pipeline we simply extract the commands from `out/cellatlas_info.json` and run them on the command line. 
```{r, eval = FALSE}
cmds <- system2("jq", "-r '.commands[] | values[]' out/cellatlas_info.json", stdout=TRUE)
cmds <- str_subset(cmds, "[\\[\\]]", negate=TRUE)
cmds <- str_extract(cmds, "kb.*(txt|gz)")

lapply(cmds, function(cmd) system(cmd))
```
## Inspect the output 
We inspect the `out/run_info.json` and `out/kb_info.json` as a simple QC on the pipeline.
```{r, eval = FALSE}
system("tree out")
```

```{r, eval = FALSE}
system("cat out/run_info.json")
# system("cat out/kb_info.json")
```
