---
title: "Xenium breast cancer dataset"
author: "Lambda Moses"
date: "`r Sys.Date()`"
output: 
  html_document:
  toc: true
  toc_float:
    collapsed: false
  toc_depth: 3
bibliography: ref.bib
vignette: >
  %\VignetteIndexEntry{xenium}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

Xenium is a new technology from 10X genomics for single cell resolution smFISH based spatial transcriptomics. The first Xenium dataset is for formalin fixed paraffin embedded (FFPE) human breast tumor, reported in [@Janesick2022-rp] and downloaded from the [10X website](https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast). This might not be representative of Xenium data after the formal release in late 2022.

The gene count matrix was downloaded as an HDF5 file and read into R as a `SingleCellExperiment` (SCE) object with `DropletUtils::read10xCounts()`. The cell metadata (including centroid coordinates) and cell segmentation polygons were downloaded as `parquet` files, a more compact way to store columnar data than CSV, and read into R as data frames with `arrow::read_parquet()`. The cell polygons were converted into `sf` data frame with `SpatialFeatureExperiment::df2sf()`. Then the SCE object was converted into `SpatialFeatureExperiment` (SFE) and the polygon geometry was added to the SFE object, which is in the `SFEData` package.

Here we load the packages used in this vignette.

```{r setup}
library(Voyager)
library(SFEData)
library(SingleCellExperiment)
library(SpatialExperiment)
library(ggplot2)
library(stringr)
library(scater) # devel version of plotExpression
library(scuttle)
library(ggforce)
library(tidyselect)
library(BiocParallel)
library(BiocSingular)
library(bluster)
library(scran)
library(dplyr)
library(tidyr)
library(ggnewscale)
theme_set(theme_void())
```

```{r}
sfe <- JanesickBreastData(dataset = "rep2")
sfe
```

The SFE object doesn't have column names (i.e. cell IDs). Here we assign cell IDs.
```{r}
colnames(sfe) <- seq_len(ncol(sfe))
```

This is what the tissue, with the cell outlines, looks like

```{r}
ggplot(cellSeg(sfe)) + geom_sf(size = 0.1) +
    theme_bw() +
    scale_x_continuous(expand = expansion()) +
    scale_y_continuous(expand = expansion())
```

Plot cell density in space
```{r}
as.data.frame(spatialCoords(sfe)) |> 
    ggplot(aes(x_centroid, y_centroid)) +
    geom_hex(bins = 200) +
    scale_fill_distiller(palette = "Blues", direction = 1) +
    theme_bw() + coord_equal() +
    scale_x_continuous(expand = expansion()) +
    scale_y_continuous(expand = expansion())
```

# Quality control
## Cells
Some QC metrics are precomputed and are stored in `colData`

```{r}
names(colData(sfe))
```

Since there're more cells, it would be better to plot the tissue larger, so we'll plot the histogram of QC metrics and the spatial plots separately, unlike in the CosMx vignette.

```{r}
n_panel <- 313
```

```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(nCounts/n_panel)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

Here we divided nCounts by the total number of genes probed, so this histogram is comparable to those from other smFISH-based datasets. 

```{r}
plotSpatialFeature(sfe, "nCounts", colGeometryName = "cellSeg")
```

There seems to be FOV artifacts. However, the cell ID and FOV information is unavailable. 

Since there are only a few hundred genes plus negative control probes, all row names of the SFE object can be printed out to find what the negative control probes are called.
```{r}
rownames(sfe)
```

According to the Xenium paper [@Janesick2022-rp], there are 3 types of controls:

> 1) probe controls to assess non-specific binding to RNA, 
2) decoding controls to assess misassigned genes, and 
3) genomic DNA (gDNA) controls to ensure the signal is from RNA.

The paper does not explain in detail how those control probes were designed, nor explain what the blank probes are. But the blank probes can be used as a negative control.

```{r}
is_blank <- str_detect(rownames(sfe), "^BLANK_")
sum(is_blank)
```

This should be number 1, the probe control
```{r}
is_neg <- str_detect(rownames(sfe), "^NegControlProbe")
sum(is_neg)
```

This should be number 2, the decoding control
```{r}
is_neg2 <- str_detect(rownames(sfe), "^NegControlCodeword")
sum(is_neg2)
```

This must be number 3, gDNA control
```{r}
is_anti <- str_detect(rownames(sfe), "^antisense")
sum(is_anti)
```

Also make an indicator of whether a feature is any sort of negative control
```{r}
is_any_neg <- is_blank | is_neg | is_neg2 | is_anti
```

Here we plot the number of genes detected per cell
```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(nGenes/n_panel)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

Compared to the [FFPE CosMX non-small cell lung cancer dataset](https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#cells), more transcripts per gene on average and a larger proportion of all genes are detected in this dataset, which is also FFPE. However, this should be interpreted with care, since these two datasets are from different tissues and have different gene panels, so this may or may not indicate that Xenium has better detection efficiency than CosMX. 

```{r}
plotSpatialFeature(sfe, "nGenes", colGeometryName = "cellSeg")
```

How does nCounts relate to nGenes?
```{r}
# I think the function to plot density2d rather than points should go into the package
as.data.frame(colData(sfe)) |> 
    ggplot(aes(nCounts, nGenes)) +
    geom_hex(bins = 150) +
    scale_fill_viridis_c() +
    theme_bw()
```

There appear to be two branches. I wonder if they're related to the sparse vs. dense regions. They probably are. 

Here we plot the distribution of cell area
```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(cell_area)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

That's should be in pixels. There's a very long tail.

How is cell area distributed in space?
```{r}
plotSpatialFeature(sfe, "cell_area", colGeometryName = "cellSeg")
```

Cells in the sparse region tend to be larger than those in the dense region. This may be biological or an artifact of the cell segmentation algorithm or both.

Here we plot the distribution of nucleus area

```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(nucleus_area)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

There seems to be an outlier. Here we locate the outlier in space.

```{r}
plotSpatialFeature(sfe, "nucleus_area", colGeometryName = "cellSeg")
```

That's the outlier near the right edge of the section, throwing off the dynamic range of this plot. Upon inspection of the H&E image, that outlier is a bit of tissue debris that doesn't look like a cell. But we can still that cells in the dense, gland like regions tend to have larger nuclei. This may be biological, or that nuclei are so densely packed in those regions that they are more likely to be undersegmented, i.e. when multiple nuclei are counted as one by the nuclei segmentation program, or both.

How does cell area relate to nuclei area?
```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(cell_area, nucleus_area)) +
    geom_hex(bins = 150) +
    scale_fill_viridis_c() +
    theme_bw()
```

Again, there are two branches, probably related to cell density and cell type. That nucleus outlier also has large cell area, though not as much an outlier in cell area. However, it is a spatial outlier as it's unusually large compared to its neighbors (scroll up two plots back). 

How about proportion of cell in this z-plane taken up by the nucleus? Here we calculate this proportion and then plot its distribution.
```{r}
colData(sfe)$prop_nuc <- sfe$nucleus_area / sfe$cell_area
```

```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(prop_nuc)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

This distribution could have been generated from two peaks added up. From the histogram, there seems not to be cells without nuclei or segmentation artifacts where the nucleus is larger than the cell. However, since there are so many cells in this dataset that just a few cells would not be visible on this histogram, here we double check:
```{r}
# No nucleus
sum(sfe$nucleus_area < 1)
# Nucleus larger than cell
sum(sfe$nucleus_area > sfe$cell_area)
```

So there are no cells without nuclei or nuclei larger than their cells. Here we plot the nuclei proportion in space:
```{r}
plotSpatialFeature(sfe, "prop_nuc", colGeometryName = "cellSeg")
```

Cells in some histological regions have larger proportions occupied by the nuclei. It would be interesting to see, controlling for cell type, how cell area, nucleus area, and proportion of cell occupied by nucleus relate to gene expression. However, a problem here is that cell segmentation is only available for one z-plane here and these areas also relate to where this z-plane intersects each cell. 

How does the proportion of cell area occupied by the nucleus relate to cell area? Here we plot a 2D histogram to better show the density of point on this plot and avoid overplotting.
```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(prop_nuc, cell_area)) +
    geom_hex(bins = 150) +
    scale_fill_viridis_c() +
    theme_bw()
```

Smaller cells tend to have higher proportion occupied by the nucleus. This can be related to cell type, or limitation to how small the nuclei can be in this tissue.

How does nucleus area relate the proportion of cell occupied by the nucleus?
```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(prop_nuc, nucleus_area)) +
    geom_hex(bins = 150) +
    scale_fill_viridis_c() +
    theme_bw()
```

This is all over the place. There are more cells with both small nuclei and low proportion of area occupied by the nucleus.

## Negative controls

The `addPerCellQCMetrics()` function in the `scuttle` package can conveniently add transcript counts, proportion of total counts, and number of features detected for any subset of features to the SCE object. Here we do this for the SFE object, as SFE inherits from SCE. 
```{r}
sfe <- addPerCellQCMetrics(sfe, subsets = list(blank = is_blank,
                                               negProbe = is_neg,
                                               negCodeword = is_neg2,
                                               anti = is_anti,
                                               any_neg = is_any_neg))
```

```{r}
names(colData(sfe))
```

Here we plot the proportion of transcript counts coming from any negative control. But first, we use Tidyverse to reshape the percentage columns for the different types of negative controls into the long form to plot them in different panels of the same plot.

```{r}
negs_long <- as.data.frame(colData(sfe)) |> 
    select(ends_with("_percent")) |> 
    pivot_longer(cols = ends_with("_percent"), names_to = "variable", 
                 values_to = "value")
```

```{r}
ggplot(negs_long, aes(value)) +
    geom_histogram(bins = 100) +
    theme_bw() +
    facet_wrap(~ variable)
```

The histogram is dominated by the bin at 0. We also plot the histogram only for cells with at least 1 count from a negative control. The NA's come from cells that got segmented but have no transcripts detected.
```{r}
negs_long |> 
    filter(value > 0, !is.na(value)) |> 
    ggplot(aes(value)) +
    geom_histogram(bins = 100) +
    theme_bw() +
    facet_wrap(~ variable)
```

The vast majority of these cells have a small percentage of transcript counts from negative controls, but there are outliers with up to 50%. 

Here we plot the distribution of the number of negative control counts per cell:
```{r}
as.data.frame(colData(sfe)) |> 
    select(ends_with("_detected")) |> 
    pivot_longer(cols = ends_with("_detected"), names_to = "variable", 
                 values_to = "value") |> 
    filter(value > 0) |> 
    ggplot(aes(value)) +
    geom_histogram(bins = 100) +
    scale_x_continuous(breaks = scales::breaks_width(1)) +
    theme_bw() +
    theme(panel.grid.minor.x = element_blank()) +
    facet_wrap(~ variable)
```

The counts are low, mostly only 1, but there are outliers with up to 10 counts. Then the outlier with 50% of counts from negative controls must have very low total real transcript counts to begin with.

The `scuttle` package can detect outliers, but by default it says anything above 0 is an outlier, since they are over 3 median absolute deviations (MADs) away from the median, which is 0, and the MAD is 0 since the vast majority of cells don't have any negative control count. But it makes sense to allow a small proportion of negative controls. Here we use the distribution just for cells with at least 1 negative control count to find outliers. This distribution has a very long tail and some definite outliers.

Are cells that are outliers in one type of negative control also outliers in other negative controls?

```{r}
ggplot(as.data.frame(colData(sfe))) +
    geom_hex(aes(x = .panel_x, y = .panel_y), bins = 50) +
    geom_autodensity() +
    scale_fill_viridis_c() +
    facet_matrix(vars(ends_with("_percent")), layer.diag = 2) +
    theme_bw()
```

It seems that sometimes yes, but largely no, hence the bins along x = 0 and y = 0 outside the last row and column (sum of all types of negative controls).

Here we get those outliers, based only on cells with at least 1 negative control count

```{r}
get_neg_ctrl_outliers <- function(col, sfe) {
    inds <- colData(sfe)$nCounts > 0 & colData(sfe)[[col]] > 0
    df <- colData(sfe)[inds,]
    outlier_inds <- isOutlier(df[[col]], type = "higher")
    outliers <- rownames(df)[outlier_inds]
    col2 <- str_remove(col, "^subsets_")
    col2 <- str_remove(col2, "_percent$")
    new_colname <- paste("is", col2, "outlier", sep = "_")
    colData(sfe)[[new_colname]] <- colnames(sfe) %in% outliers
    sfe
}
```

```{r}
cols_use <- names(colData(sfe))[str_detect(names(colData(sfe)), "_percent$")]
for (n in cols_use) {
    sfe <- get_neg_ctrl_outliers(n, sfe)
}
```

```{r}
names(colData(sfe))
```

See where the outliers are located in space
```{r}
plotSpatialFeature(sfe, "is_blank_outlier", colGeometryName = "cellSeg")
```

They're kind of hard to see. I wonder if they tend to be smaller.
```{r}
plotColData(sfe, y = "is_blank_outlier", x = "cell_area", 
            # to not to plot points, only in devel version of scater
            point_fun = function(...) list()) 
```

They do seem to be smaller. Outliers for negative probe control and negative codeword control are also hard to see on the plot, so their plots are skipped here. But one region in the tissue tends to have more counts from antisense controls. 
```{r}
plotSpatialFeature(sfe, "is_anti_outlier", colGeometryName = "cellSeg")
```

Now we have identified the outliers. We will remove the outliers and empty cells before proceeding to further analyses.

```{r}
inds_keep <- sfe$nCounts > 0 & sfe$nucleus_area < 400 & !sfe$is_anti_outlier &
    !sfe$is_blank_outlier & !sfe$is_negCodeword_outlier & !sfe$is_negProbe_outlier
(sfe <- sfe[,inds_keep])
```

Over 1000 cells were removed. Among non-outliers, how to counts from different types of negative controls relate to each other?
```{r}
ggplot(as.data.frame(colData(sfe))) +
    geom_hex(aes(x = .panel_x, y = .panel_y), bins = 50) +
    geom_autodensity() +
    scale_fill_viridis_c() +
    facet_matrix(vars(ends_with("_percent")), layer.diag = 2) +
    theme_bw()
```

The bin with the highest counts (yellow) is at the origin, the plot shows rays, possibly from the small number of counts.

How many negative control features are detected per cell?
```{r}
table(sfe$subsets_anti_detected)
```

```{r}
table(sfe$subsets_blank_detected)
```

```{r}
table(sfe$subsets_negCodeword_detected)
```

```{r}
table(sfe$subsets_negProbe_detected)
```

There are at most 3 count per cell per type. Hence those rays in the matrix plot. For the non-outliers, each type is at most around 1%, so this data looks like really good. 

## Genes

Real genes generally have higher mean expression across cells than negative controls.
```{r}
rowData(sfe)$is_neg <- is_any_neg
plotRowData(sfe, x = "means", y = "is_neg") +
    scale_y_log10() +
    annotation_logticks(sides = "b")
```

Here we look at the mean and variance of each gene
```{r}
rowData(sfe)$means <- rowMeans(counts(sfe))
rowData(sfe)$vars <- rowVars(counts(sfe))
```

Here the real genes and negative controls are plotted in different colors
```{r}
as.data.frame(rowData(sfe)) |> 
    ggplot(aes(means, vars)) +
    geom_bin2d(bins = 50, data = as.data.frame(rowData(sfe)[!is_any_neg,])) +
    scale_fill_distiller(palette = "Blues", direction = 1,
                         name = "Counts (real genes)") +
    new_scale_fill() +
    geom_bin2d(bins = 50, data = as.data.frame(rowData(sfe)[is_any_neg,])) +
    scale_fill_distiller(palette = "Reds", direction = 1,
                         name = "Counts (negative controls)") +
    geom_abline(slope = 1, intercept = 0, color = "red") +
    scale_x_log10() + scale_y_log10() +
    annotation_logticks() +
    coord_equal() +
    theme_bw()
```

The red line $y = x$ is expected if the data follows a Poisson distribution. Negative controls and real genes form mostly separate clusters. Negative controls stick close to the line, while real genes are overdispersed. Unlike in the [CosMX dataset](https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#genes), the negative controls don't seem overdispersed.

# Spatial autocorrelation of QC metrics

There's a sparse and a dense region. Then what kind of neighborhood graph shall I use? What does "neighbor" mean here? Should cells in the sparse region just be singletons? What's the length scale of their influence? It might depends on the cell type and how contact and secreted signals are used in the cell type, and length scale of influence. If k nearest neighbors are used, then the neighbors in the dense region are much closer together than those in the sparse region. If distance based neighbors are used, then cells in the dense region will have more neighbors than cells in the sparse region, and the sparse region can break into multiple compartments if the distance cutoff is not long enough. 

For the purpose of demonstration, we use k nearest neighbors with $k = 5$, with inverse distance weighting. Having more neighbors will lead to longer computation time of spatial autocorrelation metrics.

```{r}
system.time(
    colGraph(sfe, "knn5") <- findSpatialNeighbors(sfe, method = "knearneigh", 
                                                  dist_type = "idw", k = 5, 
                                                  style = "W")
)
```

```{r}
sfe <- colDataMoransI(sfe, c("nCounts", "nGenes", "cell_area", "nucleus_area"),
                      colGraphName = "knn5")
```

```{r}
colFeatureData(sfe)[c("nCounts", "nGenes", "cell_area", "nucleus_area"),]
```

Global Moran's I indicate positive spatial autocorrelation. As the strength of spatial autocorrelation can vary spatially, we also run local Moran's I.

```{r}
sfe <- colDataUnivariate(sfe, type = "localmoran", 
                         features = c("nCounts", "nGenes", "cell_area", 
                                      "nucleus_area"),
                         colGraphName = "knn5", BPPARAM = MulticoreParam(2))
```

```{r, fig.width=9, fig.height=6}
plotLocalResult(sfe, "localmoran",
                features = c("nCounts", "nGenes", "cell_area", "nucleus_area"),
                colGeometryName = "cellSeg",
                divergent = TRUE, diverge_center = 0)
```

Interestingly, nCounts is more homogenous in the interior of the dense region, while nGenes is more homogenous by the edge of the dense region. As expected, cell area is more homogenous in the sparse region. However, the nucleus area is more homogenous in the interior of the dense region. 

Moran plot for nCounts
```{r}
sfe <- colDataUnivariate(sfe, "moran.plot", "nCounts", colGraphName = "knn5")
```

```{r}
moranPlot(sfe, "nCounts") + theme_bw()
```

There are no obvious clusters here.

# Moran's I

By default, for gene experssion, the log normalized counts are used in spatial autocorrelation metrics, so before running Moran's I, we normalize the data.

```{r}
sfe <- logNormCounts(sfe)
```

Somehow while thre're fewer genes in this dataset compared to the CosMX dataset and not too many more cells, running Moran's I for all genes took much longer. Use more cores if available to speed this up.
```{r}
system.time(
    sfe <- runMoransI(sfe, colGraphName = "knn5", BPPARAM = MulticoreParam(2))
)
```

```{r}
rowData(sfe)$is_neg <- is_any_neg
```

```{r}
plotRowData(sfe, x = "moran_sample01", y = "is_neg")
```

As expected, generally the negative controls are tightly clustered around 0, while the real genes have positive Moran's I, which means there is generally no technical artifact spatial trend. No significantly negative Moran's I is observed. Why is negative spatial autocorrelation so rare in gene expression?

What are the two negative controls with a sizable Moran's I?
```{r, fig.width=9, fig.height=3}
ord <- order(rowData(sfe)$moran_sample01[is_any_neg], decreasing = TRUE)[1:2]
top_neg <- rownames(sfe)[is_any_neg][ord]
plotSpatialFeature(sfe, top_neg, colGeometryName = "cellSeg")
```

There is somewhat a spatial trend for that antisense probe, with more detected in the upper left. However, this might not significantly affect other results since there are at most 2 counts and at most about 1% of all counts in each cell. The negative control codeword has at most 1 count per cell and the cells with this negative control detected seem to be few and far between. 

These are the most detected negative controls, and the most detected one is also the one with the highest Moran's I among negative controls. However, the other negative control with higher Moran's I is not among the most detected.
```{r}
head(sort(rowData(sfe)$means[is_any_neg], decreasing = TRUE), 15)
```

What are the genes with the highest Moran's I?
```{r, fig.width=9, fig.height=6}
top_moran <- rownames(sfe)[order(rowData(sfe)$moran_sample01, decreasing = TRUE)[1:4]]
plotSpatialFeature(sfe, top_moran, colGeometryName = "cellSeg")
```

They all highlight the same histological regions, as in the [CosMX vignette](https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#morans-i). How does Moran's I relate to gene expression level?
```{r}
plotRowData(sfe, x = "means", y = "moran_sample01")
```

Very highly expressed genes have higher Moran's I, but there are some less expressed genes with higher Moran's I as well.

# Non-spatial dimension reduction and clustering

Here we run non-spatial PCA as for scRNA-seq data
```{r}
set.seed(29)
sfe <- runPCA(sfe, ncomponents = 30, scale = TRUE, BSPARAM = IrlbaParam())
```

```{r}
ElbowPlot(sfe, ndims = 30) + theme_bw()
```

```{r}
plotDimLoadings(sfe, dims = 1:6) + theme_bw()
```

```{r, fig.width=9, fig.height=9}
spatialReducedDim(sfe, "PCA", 6, colGeometryName = "cellSeg", divergent = TRUE,
                  diverge_center = 0, ncol = 2)
```

While spatial region is not explicitly used, the PC's highlight spatial regions due to spatial autocorrelation in gene expression and histological regions with different cell types.

```{r}
set.seed(29)
sfe <- runUMAP(sfe, dimred = "PCA", n_dimred = 15)
```

Non-spatial clustering and locating the clusters in space
```{r}
colData(sfe)$cluster <- clusterRows(reducedDim(sfe, "PCA")[,1:15],
                                    BLUSPARAM = SNNGraphParam(
                                        cluster.fun = "leiden",
                                        cluster.args = list(
                                            resolution_parameter = 0.5,
                                            objective_function = "modularity")))
```

```{r, fig.height=8, fig.width=8}
plotPCA(sfe, ncomponents = 4, colour_by = "cluster") +
    geom_density2d()
```

```{r}
plotUMAP(sfe, colour_by = "cluster")
```

Plot the location of the clusters in space
```{r}
plotSpatialFeature(sfe, "cluster", colGeometryName = "cellSeg")
```

# Differential expression
Cluster marker genes are found with Wilcoxon rank sum test as commonly done for scRNA-seq.
```{r}
markers <- findMarkers(sfe, groups = colData(sfe)$cluster,
                       test.type = "wilcox", pval.type = "all", direction = "up")
```

It's already sorted by p-values
```{r}
markers[[6]]
```

Get the the significant marker for each cluster to plot
```{r}
genes_use <- vapply(markers, function(x) rownames(x)[1], FUN.VALUE = character(1))
plotExpression(sfe, genes_use, x = "cluster", point_fun = function(...) list())
```

Plot more top marker genes in a heatmap
```{r, fig.width=6, fig.height=6}
genes_use2 <- unique(unlist(lapply(markers, function(x) rownames(x)[1:5])))
plotGroupedHeatmap(sfe, genes_use2, group = "cluster", colour = scales::viridis_pal()(100))
```

# Local spatial statistics of marker genes

First we plot those genes in space as a reference
```{r, fig.width=9, fig.height=9}
plotSpatialFeature(sfe, genes_use, colGeometryName = "centroids", ncol = 3,
                   size = 0.1)
```

Global Moran's I of these marker genes

```{r}
setNames(rowData(sfe)[genes_use, "moran_sample01"], genes_use)
```

All these marker genes have positive spatial autocorrelation, but some stronger than others.

Local Moran's I of these marker genes
```{r}
sfe <- runUnivariate(sfe, "localmoran", features = genes_use, colGraphName = "knn5",
                     BPPARAM = MulticoreParam(2))
```

```{r, fig.width=9, fig.height=9}
plotLocalResult(sfe, type = "localmoran", features = genes_use, 
                colGeometryName = "centroids", ncol = 3, divergent = TRUE,
                diverge_center = 0, size = 0)
```

It seems that some histological regions tend to be more spatially homogenous in gene expression than others. The epithelial region tends to be more homogenous. For some genes, regions with higher expression also have higher local Moran's I, such as FOXA1 and GATA3, while for some genes, this is not the case, such as FGL2 and LUM. 

Run local spatial heteroscdasticity (LOSH) for these marker genes to find local heterogeneity
```{r}
sfe <- runUnivariate(sfe, "LOSH", features = genes_use, colGraphName = "knn5")
```

```{r, fig.width=9, fig.height=9}
plotLocalResult(sfe, type = "LOSH", features = genes_use, 
                colGeometryName = "centroids", ncol = 3, size = 0.1)
```

Again, just like in the [CosMX dataset](https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#local-spatial-statistics-of-marker-genes), LOSH is higher where the gene is more highly expressed in some (e.g. CD3E, LUM, TENT5C) but not all cases (e.g. FOXA1, GATA3). This may be due to spatial distribution of different cell types.

# Session info
```{r}
sessionInfo()
```

# References
