---
title: "Xenium breast cancer dataset"
author: "Lambda Moses"
date: "`r Sys.Date()`"
output: 
  html_document:
  toc: true
  toc_float:
    collapsed: false
  toc_depth: 3
jupyter:
  kernelspec:
    display_name: R
    language: R
    name: ir
bibliography: ref.bib
vignette: >
  %\VignetteIndexEntry{xenium}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>", fig.align = "center"
)
```

```{r, include = FALSE, eval = FALSE}
# Install Google Colab dependencies
# Note: this can take 30+ minutes (many of the dependencies include C++ code, which needs to be compiled)

# First install `sf`, `ragg` and `textshaping` and their system dependencies:
system("apt-get -y update && apt-get install -y  libudunits2-dev libgdal-dev libgeos-dev libproj-dev libharfbuzz-dev libfribidi-dev")
install.packages("sf")
install.packages("textshaping")
install.packages("ragg")

# Install system dependencies of some other R packages that Voyager either imports or suggests:
system("apt-get install -y libfribidi-dev libcairo2-dev libmagick++-dev")

# Install Voyager from Bioconductor:
install.packages("BiocManager")
BiocManager::install(version = "3.17", ask = FALSE, update = FALSE, Ncpus = 2)
BiocManager::install("scater")
system.time(
  BiocManager::install("Voyager", dependencies = TRUE, Ncpus = 2, update = FALSE)
)

packageVersion("Voyager")
```

# Introduction

Xenium is a new technology from 10X genomics for single cell resolution smFISH based spatial transcriptomics. The dataset used here is from human pancreas and processed with Xenium Onboarding Analysis (XOA) v2. It was originally downloaded [here](https://cf.10xgenomics.com/samples/xenium/2.0.0/Xenium_V1_human_Pancreas_FFPE/Xenium_V1_human_Pancreas_FFPE_outs.zip). In the version of the data used here, the large Zarr files have been removed because they're not used here anyway and to speed up download.

Here we load the packages used in this vignette.

```{r setup, message=FALSE}
library(Voyager)
library(SFEData) 
library(SingleCellExperiment)
library(SpatialExperiment)
library(SpatialFeatureExperiment)
library(ggplot2)
library(ggforce)
library(stringr)
library(scater) 
library(scuttle)
library(BiocParallel)
library(BiocSingular)
library(bluster)
library(scran)
library(patchwork)
library(RBioFormats)
library(fs)
library(sf)
library(arrow)
theme_set(theme_bw())
options(timeout = Inf)
```

```{r}
base <- "/Volumes/Archives"
if (!dir.exists(file.path(base, "xenium2_pancreas"))) {
    download.file("https://caltech.box.com/shared/static/6yvpahb97dgp4y2gx7oqjktom9a7qp3o",
                  destfile = file.path(base, "xenium2_pancreas.tar.gz"))
    untar(file.path(base, "xenium2_pancreas.tar.gz"), exdir = base)
}
fn <- file.path(base, "xenium2_pancreas")
```

This should be the standard XOA v2 output file structure (excluding the Zarr files)
```{r}
dir_tree(fn)
```

The gene count matrix is in the `cell_feature_matrix` h5 or tar.gz file. The cell metadata is in the `cells.csv.gz` or `cells.parquet` file. The cell segmentation polygon vertices are in the `cell_boundaries.csv.gz` or `parquet` file. The nucleus segmentation polygon vertices are in the `nucleus_boundaries.csv.gz` or `parquet` files. The transcript spot coordinates are in the `transcripts.csv.gz` or `parquet` file. The images for different histological stains are in the `morphology_focus` directory. These are the files relevant to the SFE objects; see [10X documentation](https://www.10xgenomics.com/support/software/xenium-onboard-analysis/latest/analysis/xoa-output-at-a-glance) for what the other files are.

The `readXenium()` function in `SpatialFeatureExperiment` reads XOA v1 and v2 outputs into R.

```{r}
# Annoying RBioFormats bug
try(sfe <- readXenium(fn))
# Use gene symbols as row names but Ensembl IDs are still in rowData(sfe)
(sfe <- readXenium(fn, row.names = "symbol", add_molecules = TRUE))
```

There are 140702 cells in this dataset, a little more than in the CosMX dataset.

This is what the tissue, with the cell outlines, looks like

```{r, fig.width=6, fig.height=3}
plotGeometry(sfe, "cellSeg")
```

Plot the nuclei
```{r, fig.width=6, fig.height=3}
plotGeometry(sfe, "nucSeg")
```

Plot cell density in space
```{r, fig.width=6, fig.height=2.5}
plotCellBin2D(sfe, binwidth = 50)
```

There are clearly two different regions, one with high cell density, the other with lower density. They also seem to have very different spatial structures.

Plot total transcript density
```{r}
tx_coords <- st_coordinates(txSpots(sfe))
tx_points <- df2sf(tx_coords, spatialCoordsNames = c("X", "Y", "Z"), 
                   geometryType = "POINT")
tx_points <- st_geometry(tx_points)
grid <- st_make_grid(tx_points, cellsize = 50)
tx_int <- st_intersects(grid, tx_points)
tx_int <- lengths(tx_int)
grid <- st_sf(tx_count = tx_int, geometry = grid)
grid <- grid[grid$tx_count > 0,]
```

```{r, fig.width=6, fig.height=2.5}
ggplot(grid, aes(fill = tx_count)) +
    geom_sf(linewidth = 0) +
    scale_fill_distiller(palette = "Blues", direction = 1)
```

Cool, so the region with higher cell density also has higher transcript density.
```{r}
rm(list = c("tx_points", "tx_coords"))
gc()
```

Plot the images; the image has 4 channels, which are DAPI, ATP1A1/CD45/E-Cadherin, 18S, and AlphaSMA/Vimentin, in this order. The images are pyramids with multiple resolutions; some applications would not require the highest resolution. The file for each channel is around 370 MB. Only the metadata is read in R and only the relevant portion of the image at the highest resolution necessary is loaded into memory when needed, say when plotting. 

```{r}
getImg(sfe)
```

There are 4 image files, each for one channel, though the XML metadata of the files connect them together so BioFormats treats them as one file.
```{r}
dir_tree(dirname(imgSource(getImg(sfe))))
```

The image can be plotted with `plotImage()` in `Voyager`, but only up to 3 channels can be ploted at the same time. Say plot 18S (red), ATP1A1/CD45/E-Cadherin (green), and DAPI (blue). These are specified in the `channel` argument, using channel indices. When using 3 channels, by default the indices correspond to R, G, and B channels in this order. The `normalize_channels` argument determines if the channels should be normalized individually in case the different channels have different dynamic ranges. We recommend setting it to `TRUE` for fluorescent images where each channel is separated separately using different fluorophores but not the bright field images where the RGB channels are imaged simultaneously.
```{r, fig.width=6, fig.height=3}
plotImage(sfe, image_id = "morphology_focus", channel = 3:1, show_axes = TRUE,
          dark = TRUE, normalize_channels = TRUE)
```

Zoom into smaller regions, one in the dense region, another in the sparser region
```{r}
bbox1 <- c(xmin = 5500, xmax = 6000, ymin = -1250, ymax = -750)
bbox2 <- c(xmin = 1500, xmax = 2000, ymin = -2000, ymax = -1500)
```

Plot the bounding boxes in the full image
```{r, fig.width=6, fig.height=3}
bboxes_sf <- c(st_as_sfc(st_bbox(bbox1)), st_as_sfc(st_bbox(bbox2)))
plotImage(sfe, image_id = "morphology_focus", channel = 3:1, show_axes = TRUE,
          dark = TRUE, normalize_channels = TRUE) +
    geom_sf(data = bboxes_sf, fill = NA, color = "white")
```

```{r}
plotImage(sfe, image_id = "morphology_focus", channel = 3:1,
          bbox = bbox1, normalize_channels = TRUE)
```

```{r}
plotImage(sfe, image_id = "morphology_focus", channel = 3:1,
          bbox = bbox2, normalize_channels = TRUE)
```

Since we're using the RGB channels for color, this is not colorblind friendly. Plot the channels separately to be colorblind friendly:
```{r}
p1 <- plotImage(sfe, image_id = "morphology_focus", channel = 1,
          bbox = bbox2) +
    ggtitle("DAPI")
p2 <- plotImage(sfe, image_id = "morphology_focus", channel = 2,
          bbox = bbox2) +
    ggtitle("ATP1A1/CD45/E-Cadherin")
p1 + p2
```

```{r}
p3 <- plotImage(sfe, image_id = "morphology_focus", channel = 1,
          bbox = bbox1) +
    ggtitle("DAPI")
p4 <- plotImage(sfe, image_id = "morphology_focus", channel = 2,
          bbox = bbox1) +
    ggtitle("ATP1A1/CD45/E-Cadherin")
p3 + p4
```

We can also plot the cell and nuclei geometries on top of the images
```{r}
plotGeometry(sfe, "cellSeg", fill = FALSE, dark = TRUE,
             image_id = "morphology_focus", 
             channel = 3:1, bbox = bbox1, normalize_channels = TRUE)
```

```{r}
plotGeometry(sfe, "nucSeg", fill = FALSE, dark = TRUE,
             image_id = "morphology_focus", 
             channel = 3:1, bbox = bbox1, normalize_channels = TRUE)
```

Plot cells and nuclei together; since there're still quite a lot of cells in that bounding box and the plot looks busy, we can use a even smaller bounding box to plus both cells and nuclei
```{r}
bbox3 <- c(xmin = 5750, xmax = 6000, ymin = -1100, ymax = -850)
nuc <- st_intersection(st_as_sfc(st_bbox(bbox3)), nucSeg(sfe))
nuc <- nuc - bbox3[c("xmin", "ymin")]
```

```{r}
plotGeometry(sfe, "cellSeg", fill = FALSE, dark = TRUE,
             image_id = "morphology_focus", 
             channel = 3:1, bbox = bbox3, normalize_channels = TRUE) +
    geom_sf(data = nuc, fill = NA, color = "magenta")
```

# Quality control
## Negative controls
Some of the features are not genes but negative controls
```{r}
names(rowData(sfe))
```

```{r}
unique(rowData(sfe)$Type)
```

According to the Xenium paper [@Janesick2022-rp], there are 3 types of controls:

> 1) probe controls to assess non-specific binding to RNA, 
2) decoding controls to assess misassigned genes, and 
3) genomic DNA (gDNA) controls to ensure the signal is from RNA.

The paper does not explain in detail how those control probes were designed.

```{r}
is_blank <- rowData(sfe)$Type == "Unassigned Codeword"
sum(is_blank)
```

This should be number 1, the probe control
```{r}
is_neg <- rowData(sfe)$Type == "Negative Control Probe"
sum(is_neg)
```

This should be number 2, the decoding control
```{r}
is_neg2 <- rowData(sfe)$Type == "Negative Control Codeword"
sum(is_neg2)
```

Also make an indicator of whether a feature is any sort of negative control
```{r}
is_any_neg <- is_blank | is_neg | is_neg2
```

The `addPerCellQCMetrics()` function in the `scuttle` package can conveniently add transcript counts, proportion of total counts, and number of features detected for any subset of features to the SCE object. Here we do this for the SFE object, as SFE inherits from SCE. 
```{r}
sfe <- addPerCellQCMetrics(sfe, subsets = list(unassigned = is_blank,
                                               negProbe = is_neg,
                                               negCodeword = is_neg2,
                                               any_neg = is_any_neg))
```

```{r}
names(colData(sfe))
```

Next we plot the proportion of transcript counts coming from any negative control. 

```{r}
cols_use <- names(colData(sfe))[str_detect(names(colData(sfe)), "_percent$")]
plotColDataHistogram(sfe, cols_use, bins = 100)
```

The histogram is dominated by the bin at zero and there are some extreme outliers too few to be seen but evident from the scale of the x axis. We also plot the histogram only for cells with at least 1 count from a negative control. The NA's come from cells that got segmented but have no transcripts detected.
```{r}
plotColDataHistogram(sfe, cols_use, bins = 100) + 
    scale_x_log10() +
    annotation_logticks(sides = "b")
```

For the most part cells have less than 10% of spots assigned to the negative controls.

Next we plot the distribution of the number of negative control counts per cell:
```{r}
cols_use <- names(colData(sfe))[str_detect(names(colData(sfe)), "_sum$")]
plotColDataHistogram(sfe, cols_use, bins = 20, ncol = 2) +
    # Avoid decimal breaks on x axis unless there're too few breaks
    scale_x_continuous(breaks = scales::breaks_extended(Q = c(1,2,5))) +
    scale_y_log10() +
    annotation_logticks(sides = "l")
```

The counts are low, mostly zero, but there are some cells with up to 2 counts of all types aggregated. Then the outlier with 30% of counts from negative controls must have very low total real transcript counts to begin with. The negative controls indicates the prevalence of false positives. Here we have only a few hundred cells out of over 140,000 that have negative control probes detected. 

```{r}
names(colData(sfe))
```

Where are cells with higher proportion of negative control probes (i.e. low total counts and have negative control probes detected) located in space?
```{r}
data("ditto_colors")
```

```{r}
sfe$more_neg <- sfe$subsets_any_neg_percent > 10
plotSpatialFeature(sfe, "subsets_any_neg_sum", colGeometryName = "cellSeg",
                   show_axes = TRUE) +
    geom_sf(data = SpatialFeatureExperiment::centroids(sfe)[sfe$more_neg,],
            size = 0.5, color = ditto_colors[1]) +
    theme(legend.position = "bottom")
```

Cells that have any negative control probe appear to be randomly distributed in space, but cells that have higher proportion of negative control seem to be few and over-represented in the less dense area. Those could be segmentation artifacts and not actually cells.

Where are negative control spots located?
```{r}
tx_coords <- st_coordinates(txSpots(sfe)[is_any_neg,])
tx_points <- df2sf(tx_coords, spatialCoordsNames = c("X", "Y", "Z"), 
                   geometryType = "POINT")
tx_points <- st_geometry(tx_points)
grid <- st_make_grid(tx_points, cellsize = 100)
tx_int <- st_intersects(grid, tx_points)
tx_int <- lengths(tx_int)
grid <- st_sf(tx_count = tx_int, geometry = grid)
grid <- grid[grid$tx_count > 0,]
```

```{r, fig.width=6, fig.height=2.5}
ggplot(grid, aes(fill = tx_count)) +
    geom_sf(linewidth = 0) +
    scale_fill_distiller(palette = "Blues", direction = 1)
```

Generally there are more negative control spots in the region with higher cell and transcript density, which is not surprising.

Zoom into one of those areas to inspect
```{r}
bbox <- c(xmin = 1550, xmax = 1750, ymin = -1800, ymax = -1700)
```

```{r}
nuc <- st_intersection(st_as_sfc(st_bbox(bbox)), nucSeg(sfe))
nuc <- nuc - bbox[c("xmin", "ymin")]
more_neg <- st_intersection(st_as_sfc(st_bbox(bbox)),
                         SpatialFeatureExperiment::centroids(sfe)[sfe$more_neg,])
more_neg <- more_neg - bbox[c("xmin", "ymin")]
neg_spots <- st_intersection(st_as_sfc(st_bbox(bbox)), tx_points)
neg_spots <- neg_spots - bbox[c("xmin", "ymin")]
```

```{r}
plotSpatialFeature(sfe, "total_counts", colGeometryName = "cellSeg", 
                   aes_use = "color",
                   fill = NA, dark = TRUE, image_id = "morphology_focus", 
                   channel = 3:1, bbox = bbox, normalize_channels = TRUE) +
    geom_sf(data = nuc, fill = NA, color = "magenta") +
    geom_sf(data = more_neg, shape = 4,
            size = 4, color = ditto_colors[1]) +
    geom_sf(data = neg_spots, color = "white", size = 1, shape = 3) +
    theme(legend.position = "bottom")
```

The two cells in this bounding box with over 10% of counts from negative controls are indicated with organge crosses. In terms of total counts and morphology, they don't seem exceptional, so they might not be lower quality cells.

```{r}
sfe$has_any_neg <- sfe$subsets_any_neg_sum > 0
```

```{r}
plotColData(sfe, y = "has_any_neg", x = "cell_area", 
            # To not to plot the points in the violin plots, too many of them
            point_fun = function(...) list()) /
    
```

```{r}
plotColData(sfe, y = "has_any_neg", x = "total_counts", 
            point_fun = function(...) list()) 
```

It seems that cells that have negative controls tend to be larger and have more total counts. This is the opposite trend from the first Xenium preview data from 2022 which has a region in the top left corner with more negative control probes detected (see `JanesickBreastData()`). Given that cells with any negative control count only have 1 or 2 molecules from negative controls, this trend can probably be explained simply by the larger number of molecules, just like when tossing a fair coin 10 times and repeat that for 1000 times, the probability of getting 10 tails in at least one trial is large. The probability is even larger if done 10000 times.

## Cells
Some QC metrics are precomputed and are stored in `colData`

```{r}
names(colData(sfe))
```

Since there're more cells, it would be better to plot the tissue larger, so we'll plot the histogram of QC metrics and the spatial plots separately, unlike in the CosMx vignette.

```{r}
n_panel <- sum(!is_any_neg)
sfe$nCounts <- colSums(counts(sfe)[!is_any_neg,])
sfe$nGenes <- colSums(counts(sfe)[!is_any_neg,] > 0)
colData(sfe)$nCounts_normed <- sfe$nCounts/n_panel
colData(sfe)$nGenes_normed <- sfe$nGenes/n_panel
```

Here we divided the total number of transcript molecules detected from genes (`nCounts`) and the number of genes detected (`nGenes`) by the total number of genes probed, so this histogram is comparable to those from other smFISH-based datasets. 

```{r}
plotColDataHistogram(sfe, c("nCounts", "nGenes"))
```

There're weirdly some discrete values of the number of genes detected, which excludes negative controls.
```{r}
plotColDataHistogram(sfe, c("nCounts_normed", "nGenes_normed"))
```

Compared to the [FFPE CosMX non-small cell lung cancer dataset](https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#cells), fewer transcripts per gene on average and a smaller proportion of all genes are detected in this dataset, which is also FFPE. However, this should be interpreted with care, since these two datasets are from different tissues and have different gene panels, so this may or may not indicate that Xenium has better detection efficiency than CosMX. See (I know the two papers) for systematic comparisons of different imaging based technologies with the same tissue.

```{r, fig.width=6, fig.height=2.5}
plotSpatialFeature(sfe, "nCounts", colGeometryName = "cellSeg")
```

There seem to be FOV artifacts. However, the cell ID and FOV information were unavailable so we cannot examine them. 

```{r, fig.width=6, fig.height=2.5}
plotSpatialFeature(sfe, "nGenes", colGeometryName = "cellSeg")
```

A standard examination is to look at the relationship between nCounts and nGenes:
```{r}
plotColData(sfe, x="nCounts", y="nGenes", bins = 70) +
    scale_fill_distiller(palette = "Blues", direction = 1)
```

Here we plot the distribution of cell area
```{r}
plotColDataHistogram(sfe, c("cell_area", "nucleus_area"), scales = "free_y")
```

That should be in pixels. There's a very long tail. The nuclei are much smaller than the cells.

How is cell area distributed in space?
```{r, fig.width=6, fig.height=2.5}
plotSpatialFeature(sfe, "cell_area", colGeometryName = "cellSeg", 
                   show_axes = TRUE)
```

The largest cells tend to be in the sparse area. This may be biological or an artifact of the cell segmentation algorithm or both.

Here the nuclei segmentations are plotted instead of cell segmentation. The nuclei are much smaller to the extent that they are difficult to see.
```{r, fig.width=6, fig.height=2.5}
plotSpatialFeature(sfe, "nucleus_area", colGeometryName = "nucSeg",
                   show_axes = TRUE)
```

Zoom into smaller regions to see the nature of the very large cells
```{r}
bbox3 <- c(xmin = 1350, xmax = 1550, ymin = -1750, ymax = -1550)
bbox4 <- c(xmin = 3400, xmax = 3600, ymin = -2900, ymax = -2700)
```

```{r}
nuc <- st_intersection(st_as_sfc(st_bbox(bbox3)), nucSeg(sfe))
nuc <- nuc - bbox3[c("xmin", "ymin")]
```

```{r}
plotGeometry(sfe, "cellSeg", fill = FALSE, dark = TRUE,
             image_id = "morphology_focus", 
             channel = 3:1, bbox = bbox3, normalize_channels = TRUE) +
    geom_sf(data = nuc, fill = NA, color = "magenta")
```

```{r}
nuc <- st_intersection(st_as_sfc(st_bbox(bbox4)), nucSeg(sfe))
nuc <- nuc - bbox4[c("xmin", "ymin")]
```

```{r}
plotGeometry(sfe, "cellSeg", fill = FALSE, dark = TRUE,
             image_id = "morphology_focus", 
             channel = 3:1, bbox = bbox4, normalize_channels = TRUE) +
    geom_sf(data = nuc, fill = NA, color = "magenta")
```

In most cases, the large cells don't immediately look like artifacts of segmentation, so maybe don't remove them for QC purposes. Meanwhile we see different morphologies in the cells and nuclei which can be interesting to further explore.

Next we calculate the proportion of cell in this z-plane taken up by the nucleus, and examine the distribution:

```{r}
colData(sfe)$prop_nuc <- sfe$nucleus_area / sfe$cell_area
```

```{r}
summary(sfe$prop_nuc)
```

```{r}
plotColDataHistogram(sfe, "prop_nuc")
```

The NA's are cells that don't have nuclei detected. There are also nearly 2500 cells entirely taken up the their nuclei.

Here we plot the nuclei proportion in space:
```{r, fig.width=6, fig.height=2.5}
plotSpatialFeature(sfe, "prop_nuc", colGeometryName = "cellSeg")
```

Cells in some histological regions have larger proportions occupied by the nuclei. It is interesting to check, controlling for cell type, how cell area, nucleus area, and the proportion of cell occupied by nucleus relate to gene expression. However, a problem in performing such an analysis is that cell segmentation is only available for one z-plane here and these areas also relate to where this z-plane intersects each cell. 

Does cell area relate to `nCounts`, nucleus area, proportion of area taken up by nucleus, and so on? Here we plot how each pair of variables relate to each other in a matrix plot using the [`ggforce`](https://ggforce.data-imaginist.com/) package.

```{r}
cols_use <- c("nCounts", "nGenes", "cell_area", "nucleus_area", "prop_nuc")
df <- colData(sfe)[,cols_use] |> as.data.frame()
```

```{r, fig.width=8, fig.height=6}
ggplot(df) +
    geom_bin2d(aes(x = .panel_x, y = .panel_y), bins = 30) +
    geom_autodensity(fill = "gray90", color = "gray70", linewidth = 0.2) +
    facet_matrix(vars(tidyselect::everything()), layer.diag = 2) +
    scale_fill_distiller(palette = "Blues", direction = 1)
```

Cool, so, generally, `nCounts`, `nGenes`, cell area, and nucleus area positively correlate with each other, but the proportion of cell area taken up by the nucleus doesn't seem to correlate with the other variables.

```{r}
sfe$has_nuclei <- !is.na(sfe$nucleus_area)
plotColData(sfe, y = "has_nuclei", x = "cell_area", 
            point_fun = function(...) list())
```

It seems that cells that don't have nuclei detected tend to be smaller, probably because less of their nuclei are captured in this tissue section. Where are the cells without nuclei distributed in space?

```{r, fig.width=6, fig.height=2.5}
plotSpatialFeature(sfe, "has_nuclei", colGeometryName = "cellSeg",
                   show_axes = TRUE) +
    # To highlight the cells that don't have nuclei
    geom_sf(data = SpatialFeatureExperiment::centroids(sfe)[!sfe$has_nuclei,], 
            color = ditto_colors[1], size = 0.3)
```

It seems that there are more cells without nuclei in regions with higher cell density; would be interesting to do spatial point process analysis. Zoom into a smaller region to inspect, especially the cells outside the main piece of tissue:

```{r}
bbox5 <- c(xmin = 6400, xmax = 6800, ymin = -300, ymax = -50)
bbox6 <- c(xmin = 600, xmax = 1000, ymin = -600, ymax = -300)
bbox7 <- c(xmin = 6800, xmax = 7200, ymin = -2900, ymax = -2500)
```

```{r}
nuc <- st_intersection(st_as_sfc(st_bbox(bbox5)), nucSeg(sfe))
nuc <- nuc - bbox5[c("xmin", "ymin")]
```

```{r}
plotGeometry(sfe, "cellSeg", fill = FALSE, dark = TRUE,
             image_id = "morphology_focus", 
             channel = 3:1, bbox = bbox5, normalize_channels = TRUE) +
    geom_sf(data = nuc, fill = NA, color = "magenta")
```

```{r}
nuc <- st_intersection(st_as_sfc(st_bbox(bbox6)), nucSeg(sfe))
nuc <- nuc - bbox6[c("xmin", "ymin")]
```

```{r}
plotGeometry(sfe, "cellSeg", fill = FALSE, dark = TRUE,
             image_id = "morphology_focus", 
             channel = 3:1, bbox = bbox6, normalize_channels = TRUE) +
    geom_sf(data = nuc, fill = NA, color = "magenta")
```

```{r}
nuc <- st_intersection(st_as_sfc(st_bbox(bbox7)), nucSeg(sfe))
nuc <- nuc - bbox7[c("xmin", "ymin")]
```

```{r}
plotGeometry(sfe, "cellSeg", fill = FALSE, dark = TRUE,
             image_id = "morphology_focus", 
             channel = 3:1, bbox = bbox7, normalize_channels = TRUE) +
    geom_sf(data = nuc, fill = NA, color = "magenta")
```

Here in QC we are looking for low quality cells. From what we've explored so far, cells with negative control counts and exceptionally large cells might not be suspect, but cells far outside the tissue are suspect.

## Genes
Here we look at the mean and variance of each gene
```{r}
rowData(sfe)$means <- rowMeans(counts(sfe))
rowData(sfe)$vars <- rowVars(counts(sfe))
```

Real genes generally have higher mean expression across cells than negative controls.
```{r}
rowData(sfe)$is_neg <- is_any_neg
plotRowData(sfe, x = "means", y = "is_neg") +
    scale_y_log10() +
    annotation_logticks(sides = "b")
```

Here the real genes and negative controls are plotted in different colors
```{r}
plotRowData(sfe, x="means", y="vars", color_by = "is_neg") +
    geom_abline(slope = 1, intercept = 0, color = "red") +
    scale_x_log10() + scale_y_log10() +
    annotation_logticks() +
    coord_equal() +
    labs(color = "Negative control")
```

The red line $y = x$ is expected if the data follows a Poisson distribution. Negative controls and real genes form mostly separate clusters. Negative controls stick close to the line, while real genes are overdispersed. Unlike in the [CosMX dataset](https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#genes), the negative controls don't seem overdispersed.

# Spatial autocorrelation of QC metrics

There's a sparse and a dense region. This poses the question of what type of neighborhood graph to use, e.g. it is conceivable that cells in the sparse region should just be singletons. Furthermore, it is unclear what the length scale of their influence might be. It might depend on the cell type and how contact and secreted signals are used in the cell type, and length scale of the influence. If k nearest neighbors are used, then the neighbors in the dense region are much closer together than those in the sparse region. If distance based neighbors are used, then cells in the dense region will have more neighbors than cells in the sparse region, and the sparse region can break into multiple compartments if the distance cutoff is not long enough. 

For the purpose of demonstration, we use k nearest neighbors with $k = 5$, with inverse distance weighting. Note that using more neighbors leads to longer computation time of spatial autocorrelation metrics.

```{r}
system.time(
    colGraph(sfe, "knn5") <- findSpatialNeighbors(sfe, method = "knearneigh", 
                                                  dist_type = "idw", k = 5, 
                                                  style = "W")
)
```

```{r}
sfe <- colDataMoransI(sfe, c("nCounts", "nGenes", "cell_area", "nucleus_area"),
                      colGraphName = "knn5")
```

```{r}
colFeatureData(sfe)[c("nCounts", "nGenes", "cell_area", "nucleus_area"),]
```

Global Moran's I indicatse positive spatial autocorrelation. As the strength of spatial autocorrelation can vary spatially, we also run local Moran's I.

```{r}
sfe <- colDataUnivariate(sfe, type = "localmoran", 
                         features = c("nCounts", "nGenes", "cell_area", 
                                      "nucleus_area"),
                         colGraphName = "knn5", BPPARAM = MulticoreParam(2))
```

The `pointsize` argument adjusts the point size in `scattermore`. The default is 0, meaning single pixels, but since the cells in the sparse region are hard to see that way, we increase `pointsize`. We would still plot the polygons in larger single panel plots, but use `scattermore` in multi-panel plots where the polygons in each panel are invisible anyway due to the small size to save some time.
```{r, fig.width=9, fig.height=6}
plotLocalResult(sfe, "localmoran",
                features = c("nCounts", "nGenes", "cell_area", "nucleus_area"),
                colGeometryName = "centroids", scattermore = TRUE,
                divergent = TRUE, diverge_center = 0, pointsize = 1)
```

Interestingly, nCounts is more homogeneous in the interior of the dense region, while nGenes is more homogeneous by the edge of the dense region. As expected, cell area is more homogeneous in the sparse region. However, the nucleus area is more homogeneous in the interior of the dense region. 

Moran plot for nCounts
```{r}
sfe <- colDataUnivariate(sfe, "moran.plot", "nCounts", colGraphName = "knn5")
```

```{r, fig.width=8, fig.height=7}
p1 <- moranPlot(sfe, "nCounts", binned = TRUE, plot_influential = FALSE) 
p2 <- moranPlot(sfe, "nCounts", binned = TRUE)
p1 / p2 + plot_layout(guides = "collect")
```

There are no obvious clusters here. In the lower panel, the 2D histogram of influential points is plotted in red.

# Moran's I

By default, for gene expression, the log normalized counts are used in spatial autocorrelation metrics, so before running Moran's I, we normalize the data.

```{r}
sfe <- logNormCounts(sfe)
```

Use more cores if available to speed this up.
```{r}
system.time(
    sfe <- runMoransI(sfe, colGraphName = "knn5", BPPARAM = MulticoreParam(2))
)
```

```{r}
rowData(sfe)$is_neg <- is_any_neg
```

```{r}
plotRowData(sfe, x = "moran_sample01", y = "is_neg")
```

As expected, generally the negative controls are tightly clustered around 0, while the real genes have positive Moran's I, which means there is generally no technical artifact spatial trend. No significantly negative Moran's I is observed. Why is negative spatial autocorrelation so rare in gene expression?

What are the two negative controls with a sizable Moran's I?
```{r, fig.width=9, fig.height=3}
ord <- order(rowData(sfe)$moran_sample01[is_any_neg], decreasing = TRUE)[1:2]
top_neg <- rownames(sfe)[is_any_neg][ord]
plotSpatialFeature(sfe, top_neg, colGeometryName = "centroids",
                   scattermore = TRUE, pointsize = 1)
```

There is somewhat a spatial trend for that antisense probe, with more detected in the upper left. However, this might not significantly affect other results since there are at most 2 counts and at most about 1% of all counts in each cell. The negative control codeword has at most 1 count per cell and the cells with this negative control detected seem to be few and far between. 

These are the most detected negative controls, and the most detected one is also the one with the highest Moran's I among negative controls. However, the other negative control with higher Moran's I is not among the most detected.
```{r}
head(sort(rowData(sfe)$means[is_any_neg], decreasing = TRUE), 15)
```

What are the genes with the highest Moran's I?
```{r, fig.width=9, fig.height=9}
top_moran <- rownames(sfe)[order(rowData(sfe)$moran_sample01, decreasing = TRUE)[1:6]]
plotSpatialFeature(sfe, top_moran, colGeometryName = "centroids",
                   scattermore = TRUE, ncol = 2, pointsize = 0.5)
```

They all highlight the same histological regions, as in the [CosMX vignette](https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#morans-i). How does Moran's I relate to gene expression level?
```{r}
plotRowData(sfe, x = "means", y = "moran_sample01")
```

Very highly expressed genes have higher Moran's I, but there are some less expressed genes with higher Moran's I as well.

# Non-spatial dimension reduction and clustering

Here we run non-spatial PCA as for scRNA-seq data
```{r}
set.seed(29)
sfe <- runPCA(sfe, ncomponents = 30, scale = TRUE, BSPARAM = IrlbaParam())
```

```{r}
ElbowPlot(sfe, ndims = 30)
```

```{r}
plotDimLoadings(sfe, dims = 1:6)
```

```{r, fig.width=9, fig.height=9}
spatialReducedDim(sfe, "PCA", 6, colGeometryName = "centroids", divergent = TRUE,
                  diverge_center = 0, ncol = 2, scattermore = TRUE, pointsize = 0.5)
```

While spatial region is not explicitly used, the PC's highlight spatial regions due to spatial autocorrelation in gene expression and histological regions with different cell types.

Non-spatial clustering and locating the clusters in space
```{r}
colData(sfe)$cluster <- clusterRows(reducedDim(sfe, "PCA")[,1:15],
                                    BLUSPARAM = SNNGraphParam(
                                        cluster.fun = "leiden",
                                        cluster.args = list(
                                            resolution_parameter = 0.5,
                                            objective_function = "modularity")))
```

Now the `scater` can also rasterize the plots with lots of points with the `rasterise` argument, but with a different mechanism from `scattermore` that requires more system dependencies. 
```{r, fig.height=8, fig.width=8}
plotPCA(sfe, ncomponents = 4, colour_by = "cluster", rasterise = FALSE)
```

Plot the location of the clusters in space
```{r}
plotSpatialFeature(sfe, "cluster", colGeometryName = "cellSeg")
```

# Differential expression
Cluster marker genes are found with Wilcoxon rank sum test as commonly done for scRNA-seq.
```{r}
markers <- findMarkers(sfe, groups = colData(sfe)$cluster,
                       test.type = "wilcox", pval.type = "all", direction = "up")
```

It's already sorted by p-values:
```{r}
markers[[6]]
```

The code below extracts the significant markers for each cluster:
```{r}
genes_use <- vapply(markers, function(x) rownames(x)[1], FUN.VALUE = character(1))
plotExpression(sfe, genes_use, x = "cluster", point_fun = function(...) list())
```

This allows for plotting more top marker genes in a heatmap:
```{r, fig.width=6, fig.height=8}
genes_use2 <- unique(unlist(lapply(markers, function(x) rownames(x)[1:5])))
plotGroupedHeatmap(sfe, genes_use2, group = "cluster", colour = scales::viridis_pal()(100))
```

# Local spatial statistics of marker genes

First we plot those genes in space as a reference
```{r, fig.width=9, fig.height=9}
plotSpatialFeature(sfe, genes_use, colGeometryName = "centroids", ncol = 3,
                   pointsize = 0.3, scattermore = TRUE)
```

Global Moran's I of these marker genes is shown below:

```{r}
setNames(rowData(sfe)[genes_use, "moran_sample01"], genes_use)
```

All these marker genes have positive spatial autocorrelation, but some stronger than others.

Local Moran's I of these marker genes is shown below:
```{r}
sfe <- runUnivariate(sfe, "localmoran", features = genes_use, colGraphName = "knn5",
                     BPPARAM = MulticoreParam(2))
```

```{r, fig.width=9, fig.height=9}
plotLocalResult(sfe, "localmoran", features = genes_use, 
                colGeometryName = "centroids", ncol = 3, divergent = TRUE,
                diverge_center = 0, scattermore = TRUE, pointsize = 0.3)
```

It seems that some histological regions tend to be more spatially homogenous in gene expression than others. The epithelial region tends to be more homogenous. For some genes, regions with higher expression also have higher local Moran's I, such as FOXA1 and GATA3, while for some genes, this is not the case, such as FGL2 and LUM. 

Finally, we assess local spatial heteroscdasticity (LOSH) for these marker genes to find local heterogeneity:
```{r}
sfe <- runUnivariate(sfe, "LOSH", features = genes_use, colGraphName = "knn5",
                     BPPARAM = MulticoreParam(2))
```

```{r, fig.width=9, fig.height=9}
plotLocalResult(sfe, "LOSH", features = genes_use, 
                colGeometryName = "centroids", ncol = 3, scattermore = TRUE, 
                pointsize = 0.3)
```

Again, just like in the [CosMX dataset](https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#local-spatial-statistics-of-marker-genes), LOSH is higher where the gene is more highly expressed in some (e.g. CD3E, LUM, TENT5C) but not all cases (e.g. FOXA1, GATA3). This may be due to spatial distribution of different cell types.

# Session info
```{r}
sessionInfo()
```

# References
