---
title: "xenium"
author: "Lambda Moses"
date: "`r Sys.Date()`"
output: 
  html_document:
  toc: true
  toc_float:
    collapsed: false
  toc_depth: 3
bibliography: ref.bib
vignette: >
  %\VignetteIndexEntry{xenium}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

# Introduction

```{r setup}
library(Voyager)
library(SFEData)
library(SingleCellExperiment)
library(SpatialExperiment)
library(ggplot2)
library(stringr)
library(scater)
library(scuttle)
library(ggforce)
theme_set(theme_void())
```

```{r}
sfe <- JanesickBreastData(dataset = "rep2")
sfe
```

This is what the tissue, with the cell outlines, looks like

```{r}
ggplot(cellSeg(sfe)) + geom_sf(size = 0.1) +
    theme_bw() +
    scale_x_continuous(expand = expansion()) +
    scale_y_continuous(expand = expansion())
```

Plot cell density
```{r}
as.data.frame(spatialCoords(sfe)) |> 
    ggplot(aes(x_centroid, y_centroid)) +
    geom_hex(bins = 200) +
    scale_fill_distiller(palette = "Blues", direction = 1) +
    theme_bw() + coord_equal() +
    scale_x_continuous(expand = expansion()) +
    scale_y_continuous(expand = expansion())
```

# Quality control
Some QC metrics are precomputed and are stored in `colData`

```{r}
names(colData(sfe))
```

Since there're more cells, it would be better to plot the tissue larger, so we'll plot the histogram of QC metrics and the spatial plots separately, unlike in the CosMx vignette.

```{r}
n_panel <- 313
```

```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(nCounts/n_panel)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

Here we divided nCounts by the total number of genes probed, so this histogram is comparable to those from other smFISH-based datasets. 

```{r}
plotSpatialFeature(sfe, "nCounts", colGeometryName = "cellSeg")
```

There seems to be FOV artifacts. However, the cell ID and FOV information is unavailable. 

```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(nGenes/n_panel)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

I wonder why some numbers are less represented, at a regular interval. Probably because of the selection of marker genes for different cell types in the probe panel.

Compared to the FFPE CosMX non-small cell lung cancer dataset, more transcript per gene on average and a larger proportion of all genes are detected in this dataset, which is also FFPE. However, this should be interpreted with care, since these two datasets are from different tissues and have different gene panels, so this may or may not indicate that Xenium has better detection efficiency than CosMX. 

```{r}
plotSpatialFeature(sfe, "nGenes", colGeometryName = "cellSeg")
```


```{r}
# I think the function to plot density2d rather than points should go into the package
as.data.frame(colData(sfe)) |> 
    ggplot(aes(nCounts, nGenes)) +
    geom_hex(bins = 150) +
    scale_fill_viridis_c() +
    theme_bw()
```

There appear to be two branches. I wonder if they're related to the sparse vs. dense regions. They probably are. 

Cell area
```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(cell_area)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

I think that's in pixels. There's a very long tail.

```{r}
plotSpatialFeature(sfe, "cell_area", colGeometryName = "cellSeg")
```

Cells in the sparse region tend to be larger than those in the dense region

Nuclei area

```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(nucleus_area)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

Probably an outlier

```{r}
plotSpatialFeature(sfe, "nucleus_area", colGeometryName = "cellSeg")
```

Ah, that's the outlier, throwing off the dynamic range. Upon inspection of the H&E image, that outlier is a bit of debris that doesn't look like a cell. But it's still visible that cells in the dense, gland like region tend to have larger nuclei.

How does cell area relate to nuclei area?
```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(cell_area, nucleus_area)) +
    geom_hex(bins = 150) +
    scale_fill_viridis_c() +
    theme_bw()
```

Again, there are two branches, probably related to cell density. That nucleus outlier also has large cell area, though not as much an outlier in cell area.

How about proportion of cell in this z-plane taken up by the nucleus?
```{r}
colData(sfe)$prop_nuc <- sfe$nucleus_area / sfe$cell_area
```

```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(prop_nuc)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

Not sure if this distribution could have been generated from two peaks added up.

```{r}
plotSpatialFeature(sfe, "prop_nuc", colGeometryName = "cellSeg")
```

aRt! 

How does the proportion of cell area occupied by the nucleus relate to cell area?
```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(prop_nuc, cell_area)) +
    geom_hex(bins = 150) +
    scale_fill_viridis_c() +
    theme_bw()
```

Smaller cells tend to have higher proportion occupied by the nucleus. This can be related to cell type, or limitation to how small the nuclei can be in this tissue.

And nucleus area?
```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(prop_nuc, nucleus_area)) +
    geom_hex(bins = 150) +
    scale_fill_viridis_c() +
    theme_bw()
```

This is all over the place. There are more cells with both small nuclei and low proportion of area occupied by the nucleus.

## Negative controls
Find what the negative control probes are called
```{r}
head(rownames(sfe))
tail(rownames(sfe))
```

According to the Xenium paper, there are 3 types of controls:

> 1) probe controls to assess non-specific binding to RNA, 
2) decoding controls to assess misassigned genes, and 
3) genomic DNA (gDNA) controls to ensure the signal is from RNA.

```{r}
is_blank <- str_detect(rownames(sfe), "^BLANK_")
sum(is_blank)
```

I think this is number 1, the probe control
```{r}
is_neg <- str_detect(rownames(sfe), "^NegControlProbe")
sum(is_neg)
```

I think this is number 2, the decoding control
```{r}
is_neg2 <- str_detect(rownames(sfe), "^NegControlCodeword")
sum(is_neg2)
```

This must be number 3, gDNA control
```{r}
is_anti <- str_detect(rownames(sfe), "^antisense")
sum(is_anti)
```

```{r}
sfe <- addPerCellQCMetrics(sfe, subsets = list(blank = is_blank,
                                               negProbe = is_neg,
                                               negCodeword = is_neg2,
                                               anti = is_anti))
```

```{r}
names(colData(sfe))
```

```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(subsets_blank_percent)) +
    geom_histogram(bins = 100) +
    theme_bw()
```
Mostly very small, except for an outlier somewhere. Only plot the part >0
```{r}
as.data.frame(colData(sfe)[sfe$subsets_blank_percent > 0 & sfe$nCounts > 0,]) |> 
    ggplot(aes(subsets_blank_percent)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

It's interesting that this distribution has a peak somewhat away from 0.

```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(subsets_negProbe_percent)) +
    geom_histogram(bins = 100) +
    theme_bw()
```
Mostly very small, except for an outlier somewhere

```{r}
as.data.frame(colData(sfe)[sfe$subsets_negProbe_percent > 0 & sfe$nCounts > 0,]) |> 
    ggplot(aes(subsets_negProbe_percent)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(subsets_negCodeword_percent)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

```{r}
as.data.frame(colData(sfe)[sfe$subsets_negCodeword_percent > 0 & sfe$nCounts > 0,]) |> 
    ggplot(aes(subsets_negCodeword_percent)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

```{r}
as.data.frame(colData(sfe)) |> 
    ggplot(aes(subsets_anti_percent)) +
    geom_histogram(bins = 100) +
    theme_bw()
```

```{r}
as.data.frame(colData(sfe)[sfe$subsets_anti_percent > 0 & sfe$nCounts > 0,]) |> 
    ggplot(aes(subsets_anti_percent)) +
    geom_histogram(bins = 100) +
    theme_bw()
```


```{r}
plotSpatialFeature(sfe, "prop_anti", colGeometryName = "cellSeg")
```

The outlier is yet another cell. I sort of wonder what those outliers look like in H&E. So it would still be nice to be able to plot the H&E image, if only cropped bits when zooming in. I can try SpatialOmicsOverlay, though I kind of wonder if it works for non-Nanostring datasets. But here, upon inspection, the H&E image doesn't cover the top part the tissue here so I can't inspect these outliers.

The paper does not explain in detail how those control probes were designed, nor explain what the blank probes are. But the blank probes can be used as a negative control.

These proportions can be calculated with scuttle.

Now we have identified the outliers. We will remove the outliers and empty cells before proceeding to further analyses.


```{r}
ggplot(as.data.frame(colData(sfe))) +
    geom_hex(aes(x = .panel_x, y = .panel_y), bins = 100) +
    geom_autodensity() +
    scale_fill_viridis_c() +
    facet_matrix(vars(ends_with("")), layer.diag = 2) +
    theme_bw()
```


So scuttle says anything above 0 is an outlier. But I think it makes sense to allow a small proportion of negative controls. I think I'll use the distribution just for cells with >0 negative control counts to find outliers. This distribution has a very long tail. When plotting, it should be after removing outliers. 
