<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>How to build count matrices with `cellatlas` â€¢ Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="How to build count matrices with `cellatlas`">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.2.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-methods">Methods</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-methods">
    <h6 class="dropdown-header" data-toc-skip>Univariate global</h6>
    <a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a>
    <a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a>
    <a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a>
    <a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Univariate local</h6>
    <a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a>
    <a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a>
    <a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a>
    <a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bivariate</h6>
    <a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a>
    <a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Multivariate</h6>
    <a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a>
    <a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/sfemethod">Custom methods</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>How to build count matrices with `cellatlas`</h1>
                        <h4 data-toc-skip class="author">Kayla Jackson
and A. Sina Booeshaghi</h4>
            
            <h4 data-toc-skip class="date">2023-06-07</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/vignettes/create_mtx_ca.Rmd" class="external-link"><code>vignettes/create_mtx_ca.Rmd</code></a></small>
      <div class="d-none name"><code>create_mtx_ca.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="building-count-matrices-with-cellatlas">Building Count Matrices with <code>cellatlas</code><a class="anchor" aria-label="anchor" href="#building-count-matrices-with-cellatlas"></a>
</h2>
<p>A major challenge in uniformly preprocessing large amounts of
single-cell genomics data from a variety of different assays is
identifying and handling sequenced elements in a coherent and consistent
fashion. Cell barcodes in reads from RNAseq data from 10x Multiome, for
example, must be extracted and error corrected in the manner as cell
barcodes in reads from ATACseq data from 10x Multiome so that
barcode-barcode registration can occur. Uniform processing in this way
minimzes computational variability and enables cross-assay
comparisons.</p>
<p>In this notebook we demonstrate how single-cell genomics data can be
preprocessed to generate a cell by feature count matrix. This
requires:</p>
<ol style="list-style-type: decimal">
<li>FASTQ files</li>
<li>
<code>seqspec</code> specification for the FASTQ files</li>
<li>Genome Sequence FASTA</li>
<li>Genome Annotation GTF</li>
<li>(optional) Feature barcode list</li>
</ol>
</div>
<div class="section level2">
<h2 id="install-packages">Install Packages<a class="anchor" aria-label="anchor" href="#install-packages"></a>
</h2>
<p>The vignette makes use of two non-standard command line tools, <a href="https://jqlang.github.io/jq/" class="external-link"><code>jq</code></a> and <a href="https://mama.indstate.edu/users/ice/tree/" class="external-link"><code>tree</code></a>.
The code cell below installs these tools on a Linux operating system and
should be updated for Mac and Windows users.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install `tree` to view files</span></span>
<span><span class="co"># system("curl --quiet --show-progress ftp://mama.indstate.edu/linux/tree/tree-2.1.0.tgz")</span></span>
<span></span>
<span><span class="co"># Install `jq`, a command-line tool for extracting key value pairs from JSON files </span></span>
<span><span class="co"># system("wget --quiet --show-progress https://github.com/stedolan/jq/releases/download/jq-1.6/jq-osx-amd64")</span></span>
<span><span class="co"># system("chmod +x jq-linux64 &amp;&amp; mv jq-linux64 /usr/local/bin/jq")</span></span></code></pre></div>
<p>We will continue with other dependencies</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clone the cellatlas repo and install the package</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"git clone https://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@github.com/cellatlas/cellatlas.git"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"cd cellatlas &amp;&amp; pip install --quiet ."</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install dependencies</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"yes | pip uninstall --quiet seqspec"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"pip install --quiet git+https://github.com/IGVF/seqspec.git"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"pip install --quiet gget kb-python"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preprocessing-workflows">Preprocessing Workflows<a class="anchor" aria-label="anchor" href="#preprocessing-workflows"></a>
</h2>
<p>Below, we have generated technology specific pre-processing workflows
that utilize <code>seqspec</code> to format genomic library sequence and
structure and <code>cellatlas</code> to generate pipelines that process
raw data into a count matrix. Currently, example workflows exist for
data generated using <a href="#rna_10xsplitseq">SPLiT-seq</a>, <a href="#clicktag">ClickTags</a>, <a href="#rna_10xv3">Chromium V3</a>, <a href="#atac_10x">Chromium Single Cell ATAC-seq</a>, <a href="#atac_10xmulti">Chromium Single Cell ATAC Multiome ATAC + Gene
Expression</a>, <a href="#crispr_10x">Chromium Single Cell CRISPR
Screening</a>, and <a href="#nuclei">Chromium Nuclei Isolation</a>.</p>
<p>Currently, <a href="#rna_10xvisium">Visium</a> is the only spatial
technology with an available <code>seqspec</code>.</p>
<p>More information on how to generate a <code>seqspec</code> for
technologies not listed here is available on <a href="https://github.com/IGVF/seqspec" class="external-link">GitHub</a>. ##
<a id="rna_10xvisium"></a>Preprocessing for Visium</p>
<div class="section level4">
<h4 id="examine-the-spec">Examine the spec<a class="anchor" aria-label="anchor" href="#examine-the-spec"></a>
</h4>
<p>We first use <code>seqspec print</code> to check that the read
structure matches what we expect. This command prints out an ordered
tree representation of the sequenced elements contained in the FASTQ
files. Note that the names of the nodes in the <code>seqspec</code> must
match the names of the FASTQ files.</p>
</div>
<div class="section level4">
<h4 id="fetch-the-references">Fetch the references<a class="anchor" aria-label="anchor" href="#fetch-the-references"></a>
</h4>
<p>This step is only necessary if the modality that we are processing
uses a transcriptome reference-based alignment.</p>
</div>
<div class="section level4">
<h4 id="build-the-pipeline">Build the pipeline<a class="anchor" aria-label="anchor" href="#build-the-pipeline"></a>
</h4>
</div>
<div class="section level3">
<h3 id="preprocessing-for-split-seq">
<a id="rna_10xsplitseq"></a>Preprocessing for SPLiT-seq<a class="anchor" aria-label="anchor" href="#preprocessing-for-split-seq"></a>
</h3>
<div class="section level4">
<h4 id="examine-the-spec-1">Examine the spec<a class="anchor" aria-label="anchor" href="#examine-the-spec-1"></a>
</h4>
<p>We first use <code>seqspec print</code> to check that the read
structure matches what we expect. This command prints out an ordered
tree representation of the sequenced elements contained in the FASTQ
files. Note that the names of the nodes in the <code>seqspec</code> must
match the names of the FASTQ files.</p>
</div>
<div class="section level4">
<h4 id="fetch-the-references-1">Fetch the references<a class="anchor" aria-label="anchor" href="#fetch-the-references-1"></a>
</h4>
<p>This step is only necessary if the modality that we are processing
uses a transcriptome reference-based alignment.</p>
</div>
<div class="section level4">
<h4 id="build-the-pipeline-1">Build the pipeline<a class="anchor" aria-label="anchor" href="#build-the-pipeline-1"></a>
</h4>
</div>
</div>
<div class="section level3">
<h3 id="preprocessing-for-clicktags">
<a id="clicktag"></a>Preprocessing for ClickTags<a class="anchor" aria-label="anchor" href="#preprocessing-for-clicktags"></a>
</h3>
<div class="section level4">
<h4 id="examine-the-spec-2">Examine the spec<a class="anchor" aria-label="anchor" href="#examine-the-spec-2"></a>
</h4>
<p>We first use <code>seqspec print</code> to check that the read
structure matches what we expect. This command prints out an ordered
tree representation of the sequenced elements contained in the FASTQ
files. Note that the names of the nodes in the <code>seqspec</code> must
match the names of the FASTQ files.</p>
</div>
<div class="section level4">
<h4 id="fetch-the-references-2">Fetch the references<a class="anchor" aria-label="anchor" href="#fetch-the-references-2"></a>
</h4>
<p>This step is only necessary if the modality that we are processing
uses a transcriptome reference-based alignment.</p>
</div>
<div class="section level4">
<h4 id="build-the-pipeline-2">Build the pipeline<a class="anchor" aria-label="anchor" href="#build-the-pipeline-2"></a>
</h4>
</div>
</div>
<div class="section level3">
<h3 id="preprocessing-for-chromium-v3-chemistry">
<a id="rna_10xv3"></a>Preprocessing for Chromium V3 Chemistry<a class="anchor" aria-label="anchor" href="#preprocessing-for-chromium-v3-chemistry"></a>
</h3>
<div class="section level4">
<h4 id="examine-the-spec-3">Examine the spec<a class="anchor" aria-label="anchor" href="#examine-the-spec-3"></a>
</h4>
<p>We first use <code>seqspec print</code> to check that the read
structure matches what we expect. This command prints out an ordered
tree representation of the sequenced elements contained in the FASTQ
files. Note that the names of the nodes in the <code>seqspec</code> must
match the names of the FASTQ files.</p>
</div>
<div class="section level4">
<h4 id="fetch-the-references-3">Fetch the references<a class="anchor" aria-label="anchor" href="#fetch-the-references-3"></a>
</h4>
<p>This step is only necessary if the modality that we are processing
uses a transcriptome reference-based alignment.</p>
</div>
<div class="section level4">
<h4 id="build-the-pipeline-3">Build the pipeline<a class="anchor" aria-label="anchor" href="#build-the-pipeline-3"></a>
</h4>
</div>
</div>
<div class="section level3">
<h3 id="preprocessing-for-chromium-single-cell-atac-seq">
<a id="atac_10x"></a>Preprocessing for Chromium Single Cell
ATAC-seq<a class="anchor" aria-label="anchor" href="#preprocessing-for-chromium-single-cell-atac-seq"></a>
</h3>
<div class="section level4">
<h4 id="examine-the-spec-4">Examine the spec<a class="anchor" aria-label="anchor" href="#examine-the-spec-4"></a>
</h4>
<p>We first use <code>seqspec print</code> to check that the read
structure matches what we expect. This command prints out an ordered
tree representation of the sequenced elements contained in the FASTQ
files. Note that the names of the nodes in the <code>seqspec</code> must
match the names of the FASTQ files.</p>
</div>
<div class="section level4">
<h4 id="fetch-the-references-4">Fetch the references<a class="anchor" aria-label="anchor" href="#fetch-the-references-4"></a>
</h4>
<p>This step is only necessary if the modality that we are processing
uses a transcriptome reference-based alignment.</p>
</div>
<div class="section level4">
<h4 id="build-the-pipeline-4">Build the pipeline<a class="anchor" aria-label="anchor" href="#build-the-pipeline-4"></a>
</h4>
</div>
</div>
<div class="section level3">
<h3 id="preprocessing-for-chromium-single-cell-atac-multiome-atac">
<a id="atac_10xmulti"></a>Preprocessing for Chromium Single Cell
ATAC Multiome ATAC<a class="anchor" aria-label="anchor" href="#preprocessing-for-chromium-single-cell-atac-multiome-atac"></a>
</h3>
<div class="section level4">
<h4 id="examine-the-spec-5">Examine the spec<a class="anchor" aria-label="anchor" href="#examine-the-spec-5"></a>
</h4>
<p>We first use <code>seqspec print</code> to check that the read
structure matches what we expect. This command prints out an ordered
tree representation of the sequenced elements contained in the FASTQ
files. Note that the names of the nodes in the <code>seqspec</code> must
match the names of the FASTQ files.</p>
</div>
<div class="section level4">
<h4 id="fetch-the-references-5">Fetch the references<a class="anchor" aria-label="anchor" href="#fetch-the-references-5"></a>
</h4>
<p>This step is only necessary if the modality that we are processing
uses a transcriptome reference-based alignment.</p>
</div>
<div class="section level4">
<h4 id="build-the-pipeline-5">Build the pipeline<a class="anchor" aria-label="anchor" href="#build-the-pipeline-5"></a>
</h4>
</div>
</div>
<div class="section level3">
<h3 id="preprocessing-for-chromium-single-cell-crispr-screening">
<a id="crispr_10x"></a>Preprocessing for Chromium Single Cell CRISPR
Screening<a class="anchor" aria-label="anchor" href="#preprocessing-for-chromium-single-cell-crispr-screening"></a>
</h3>
<div class="section level4">
<h4 id="examine-the-spec-6">Examine the spec<a class="anchor" aria-label="anchor" href="#examine-the-spec-6"></a>
</h4>
<p>We first use <code>seqspec print</code> to check that the read
structure matches what we expect. This command prints out an ordered
tree representation of the sequenced elements contained in the FASTQ
files. Note that the names of the nodes in the <code>seqspec</code> must
match the names of the FASTQ files.</p>
</div>
<div class="section level4">
<h4 id="fetch-the-references-6">Fetch the references<a class="anchor" aria-label="anchor" href="#fetch-the-references-6"></a>
</h4>
<p>This step is only necessary if the modality that we are processing
uses a transcriptome reference-based alignment.</p>
</div>
<div class="section level4">
<h4 id="build-the-pipeline-6">Build the pipeline<a class="anchor" aria-label="anchor" href="#build-the-pipeline-6"></a>
</h4>
</div>
</div>
<div class="section level3">
<h3 id="preprocessing-for-chromium-nuclei-isolation">
<a id="#nuclei"></a>Preprocessing for Chromium Nuclei Isolation<a class="anchor" aria-label="anchor" href="#preprocessing-for-chromium-nuclei-isolation"></a>
</h3>
<div class="section level4">
<h4 id="examine-the-spec-7">Examine the spec<a class="anchor" aria-label="anchor" href="#examine-the-spec-7"></a>
</h4>
<p>We first use <code>seqspec print</code> to check that the read
structure matches what we expect. This command prints out an ordered
tree representation of the sequenced elements contained in the FASTQ
files. Note that the names of the nodes in the <code>seqspec</code> must
match the names of the FASTQ files.</p>
</div>
<div class="section level4">
<h4 id="fetch-the-references-7">Fetch the references<a class="anchor" aria-label="anchor" href="#fetch-the-references-7"></a>
</h4>
<p>This step is only necessary if the modality that we are processing
uses a transcriptome reference-based alignment.</p>
</div>
<div class="section level4">
<h4 id="build-the-pipeline-7">Build the pipeline<a class="anchor" aria-label="anchor" href="#build-the-pipeline-7"></a>
</h4>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Kayla Jackson, Laura Luebbert, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
