<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>Slide-Seq V2 Exploratory Data Analysis • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Slide-Seq V2 Exploratory Data Analysis">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.0.10</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Slide-Seq V2 Exploratory Data Analysis</h1>
                        <h4 data-toc-skip class="author">Kayla
Jackson</h4>
            
            <h4 data-toc-skip class="date">2023-03-19</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/vignettes/vig3_slideseq_v2.Rmd" class="external-link"><code>vignettes/vig3_slideseq_v2.Rmd</code></a></small>
      <div class="d-none name"><code>vig3_slideseq_v2.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p><a href="https://doi.org/10.1038/s41587-020-0739-1" class="external-link">Slide-seq V2</a>
is a spatial transcriptomic tool that measures genome-wide expression
using DNA-barcoded beads patterned on a slide in an non-regular array.
The beads used in the current protocol have a diameter of <span class="math inline">\(10 \mu m\)</span> and are thus larger than a
single cell, but the number of detected transcripts is an order of
magnitude higher compared to the previous iteration of the
technology.</p>
<p>In this vignette, we use <code>Voyager</code> to analyze a dataset
generated using the Slide-Seq V2 technology. The data is described in <a href="https://doi.org/10.1016/j.cell.2022.06.007" class="external-link">Dissecting the
treatment-naive ecosystem of human melanoma brain metastasis</a> <span class="citation">(Biermann et al. 2022)</span>. The raw counts and cell
metadata are publicly available from GEO. We will focus on one of the
human melanoma brain metastasis (MBM) samples that are provided in the
<code>SFEData</code> package as a
<code>SpatialFeatureExperiment</code>(SFE) object. The SFE object
contains raw counts, QC metrics such as number of UMIs and genes
detected per barcode, and centroid coordinates for each barcode as a
<code>sf</code> POINT geometry.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: scuttle</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-spatial/spdep/" class="external-link">spdep</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: spData</span></span>
<span><span class="co">#&gt; To access larger datasets in this package, install the spDataLarge</span></span>
<span><span class="co">#&gt; package with: `install.packages('spDataLarge',</span></span>
<span><span class="co">#&gt; repos='https://nowosad.github.io/drat/', type='source')`</span></span>
<span><span class="co">#&gt; Loading required package: sf</span></span>
<span><span class="co">#&gt; Linking to GEOS 3.10.2, GDAL 3.4.2, PROJ 8.2.1; sf_use_s2() is TRUE</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/BiermannMelaMetasData.html" class="external-link">BiermannMelaMetasData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"MBM05_rep1"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; snapshotDate(): 2022-10-31</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; downloading 1 resources</span></span>
<span><span class="co">#&gt; retrieving 1 resource</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; require("SpatialFeatureExperiment")</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 27566 29536 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(27566): A1BG A1BG-AS1 ... ZZZ3 snoZ196</span></span>
<span><span class="co">#&gt; rowData names(3): means vars cv2</span></span>
<span><span class="co">#&gt; colnames(29536): ACCACTCATTTCTC-1 GTTCANTCCACGTA-1 ... ACGCGCAATCGTAG-1</span></span>
<span><span class="co">#&gt;   TTGTTCCGTTCATA-1</span></span>
<span><span class="co">#&gt; colData names(4): sample_id nCounts nGenes prop_mito</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : xcoord ycoord</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>The SFE object in the <code>SFEData</code> package includes
information for 27,566 features and 29,536 beads/barcodes.</p>
</div>
<div class="section level2">
<h2 id="quality-control-qc">Quality control (QC)<a class="anchor" aria-label="anchor" href="#quality-control-qc"></a>
</h2>
<p>We begin by performing some exploratory data analysis on the barcodes
in the tissue. There are some pre-computed QC measures that are stored
in the object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "sample_id" "nCounts"   "nGenes"    "prop_mito"</span></span></code></pre></div>
<p>Total UMI counts (<code>nCounts</code>), number of genes detected per
spot (<code>nGenes</code>), and the proportion of mitochondrially
encoded counts (<code>prop_mito</code>). Below, we plot the total number
of UMI counts per barcode as a violin plot and in space. For the latter
task, we leverage the function <code><a href="../reference/plotSpatialFeature.html">plotSpatialFeature()</a></code> which
uses <code><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf()</a></code> to plot geometries where applicable. The
first few lines compute the average number of UMI counts per barcode and
this average is plotted as a red line in the vilin plot.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">avg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="op">-</span><span class="va">sample_id</span>, <span class="va">mean</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">violin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">nCounts</span><span class="op">)</span>, <span class="va">avg</span>, color<span class="op">=</span><span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, features <span class="op">=</span> <span class="st">"nCounts"</span>,</span>
<span>                               colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">violin</span> <span class="op">+</span> <span class="va">spatial</span></span></code></pre></div>
<p><img src="vig3_slideseq_v2_files/figure-html/unnamed-chunk-5-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Each barcode is represented as an <code>sf</code> POINT geometry In
the plot above, we note that many beads have quite low UMI counts, but
there are small regions throughout the tissue that appear to have high
counts. This is perhaps due to high cellular density of the melanoma
cells, but we can only speculate without an image of the tissue.
Interestingly, there are not any barcodes with zero counts. This is in
contrast to many scRNA-seq dataset were many cells have zero counts.</p>
<p>Given the density of points, we may choose to aggregate points into a
hexagonal grid to avoid overplotting. Each hexagon will be colored by
the total number of UMI counts in that space and each hexagon may
represent more than one barcode.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">xcoord</span>, <span class="va">ycoord</span>, z<span class="op">=</span><span class="va">nCounts</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/stat_summary_2d.html" class="external-link">stat_summary_hex</a></span><span class="op">(</span>fun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, bins<span class="op">=</span><span class="fl">100</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Blues"</span>, direction <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>fill<span class="op">=</span><span class="st">'nCounts'</span><span class="op">)</span>  <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_slideseq_v2_files/figure-html/unnamed-chunk-6-1.png" width="700" style="display: block; margin: auto;"></p>
<p>It is worthwhile to note that cell segmentation data were not
included with this dataset. Even though Slide-Seq V2 does not profile
gene expression at single cell resolution, cell segmentation data can be
flexibly stored as <code>annotGeometries</code> in the <code>SFE</code>
object. These geometries can be plotted with barcode-level data and can
be used with <code>sf</code> for operations like finding the number of
barcodes localized to a single cell.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">log_nCounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nCounts</span><span class="op">)</span></span>
<span></span>
<span><span class="va">avg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="op">-</span><span class="va">sample_id</span>, <span class="va">mean</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">violin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"log_nCounts"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">log_nCounts</span><span class="op">)</span>, <span class="va">avg</span>, color<span class="op">=</span><span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, features <span class="op">=</span> <span class="st">"log_nCounts"</span>,</span>
<span>                               colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>                              size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">violin</span> <span class="op">+</span> <span class="va">spatial</span></span></code></pre></div>
<p><img src="vig3_slideseq_v2_files/figure-html/unnamed-chunk-7-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The plot above visualizes the number of UMI counts per barcode on a
log scale. It appears that barcodes with higher counts are co-localized
in regions throughout the tissue, however, these regions are rather
small and may not suggest spatial autocorrelation.</p>
<p>Next we find number of genes detected per barcode. Again, this QC
feature is provided as <code>nGenes</code> in the <code>colData</code>
attribute for barcodes.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">violin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nGenes"</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">nGenes</span><span class="op">)</span>, <span class="va">avg</span>, color<span class="op">=</span><span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, features <span class="op">=</span> <span class="st">"nGenes"</span>,</span>
<span>                              colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>                              size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">violin</span> <span class="op">+</span> <span class="va">spatial</span></span></code></pre></div>
<p><img src="vig3_slideseq_v2_files/figure-html/unnamed-chunk-8-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Similar to the number of UMI counts per barcode, there seem to be
small regions with higher number of genes throughout the tissue. These
may correspond to regions of cellular diversity or high cellular
density, as might be expected in the context of melanoma.</p>
<p>We can compute the degree to which the number of UMI counts per
barcode depends on the spatial location of each measurement. This
relationship, spatial autocorrelation, can be quantified using Moran’s
index of spatial autocorrelation, or Moran’s <em>I</em>. The computation
of Moran’s <em>I</em> requires first a definition of what constitutes
objects being “near” to each other. Most simply, this is represented as
a spatial weights matrix. One possible representation is an adjacency
matrix. This matrix can be computed for polygonal data and the resulting
matrix can be binary, where entries are 1 if polygons share a border,
and 0 elsewhere (including the diagonal). These entries can be weighted
in different ways, including by the length of border shared between two
polygons.</p>
<p>This schema does not necessarily lend itself well to spatial
transcriptomic technologies, where the polygonal boundaries of cell
objects may not correspond to the measurements in the count matrix, or
individual spots or barcodes may themselves correspond to multiple
neighborhoods of cells. Certainly, the interpretation of spatial weights
matrix will change depending on the technology.</p>
<p>In any case, we can generate a putative spatial graph using the
k-nearest neighbors algorithm. This is implemented in the
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors()</a></code> function with the argument
<code>method = "knearneigh"</code> . We will store the result in the
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraphs()</a></code> slot of the SFE object.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"knn5"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>, method <span class="op">=</span> <span class="st">"knearneigh"</span>,</span>
<span>                                              dist_type <span class="op">=</span> <span class="st">"idw"</span>, k <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                                              style <span class="op">=</span> <span class="st">"W"</span><span class="op">)</span></span></code></pre></div>
<p>Now compute Moran's I for some barcode QC metrics using
<code><a href="../reference/calculateUnivariate.html">colDataMoransI()</a></code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">features_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataMoransI</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">features_use</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/colFeatureData.html">colFeatureData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">features_use</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">#&gt;         moran_sample01 K_sample01</span></span>
<span><span class="co">#&gt;              &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts      0.0965909    48.6328</span></span>
<span><span class="co">#&gt; nGenes       0.0957030    11.2037</span></span></code></pre></div>
<p>The results above do not substantiate the visual check for spatial
autocorrelation. We will continue with investigating other QC
metrics.</p>
<p>The proportion of UMIs mapping to mitochondrial genes is a useful
metric for assessing cell quality in scRNA-seq data. We will examine
this QC metric below by plotting it versus total number of UMI counts
for each barcode.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">violin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"prop_mito"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">prop_mito</span><span class="op">)</span>, <span class="va">avg</span>, color<span class="op">=</span><span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"prop_mito"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">violin</span> <span class="op">+</span> <span class="va">mito</span></span></code></pre></div>
<p><img src="vig3_slideseq_v2_files/figure-html/unnamed-chunk-12-1.png" width="700" style="display: block; margin: auto;"></p>
<p>In keeping with expectations, barcodes associated with fewer counts
appear to be associated with higher proportions of mitochondrial reads.
We will exclude barcodes containing more than &gt;10% mitochondrial
reads for subsequent analysis. The second line removes barcodes with
zero counts, but this is not necessary for this dataset as there are no
barcodes with zero counts. We keep it here just to demonstrate the
method.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Spatial neighborhood graph is reconstructed when subsetting columns</span></span>
<span><span class="co"># Use drop = TRUE to drop the graph without reconstruction, whose indices are </span></span>
<span><span class="co"># no longer valid</span></span>
<span><span class="va">sfe_filt</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">prop_mito</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">]</span></span>
<span><span class="va">sfe_filt</span> <span class="op">&lt;-</span> <span class="va">sfe_filt</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_filt</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-normalization">Data Normalization<a class="anchor" aria-label="anchor" href="#data-normalization"></a>
</h2>
<p>Normalization of spatial transcriptomics data is non-trivial and
requires thoughtful consideration. Similarly to scRNA-seq data analysis,
the goal of normalization is to remove the effects of technical
variation and derive a quantity that reflects biological variation.
However, several questions arise when considering best practices for
spatial data normalization. For example, spatial methods on average
detect fewer UMIs than their single-cell counterparts, which may
preclude the use normalization techniques such as log transformation as
shown <a href="https://doi.org/10.1093/bioinformatics/btab085" class="external-link">here</a>.
What’s more, it is not always evident whether spatial autocorrelation
between genes (or QC measures) is an artifact of the technology, and
thus, whether normalization methods should preserve the spatial
autocorrelation architecture. These questions provide avenues for active
research and development, but are currently unresolved. To this end, we
log-normalize the data in the cell below and identify variable genes for
subsequent analysis.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_filt</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sfe_filt</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dimension-reduction-and-clustering">Dimension Reduction and Clustering<a class="anchor" aria-label="anchor" href="#dimension-reduction-and-clustering"></a>
</h2>
<p>Much like in scRNA-seq analysis, we perform principal component
analysis (PCA) before clustering. We note that the method does not use
any spatial information.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">sfe_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe_filt</span>, ncomponents <span class="op">=</span> <span class="fl">30</span>, subset_row <span class="op">=</span> <span class="va">hvgs</span>,</span>
<span>                   scale <span class="op">=</span> <span class="cn">TRUE</span>, BSPARAM <span class="op">=</span> <span class="fu">BiocSingular</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html" class="external-link">IrlbaParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="co"># scale as in Seurat</span></span></code></pre></div>
<p>We can plot the variance explained by each PC.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe_filt</span>, ndims <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_slideseq_v2_files/figure-html/unnamed-chunk-16-1.png" width="700" style="display: block; margin: auto;"></p>
<p>We see that the first few components explain most of the variance in
the data. The principal components (PCs) can be plotted in space. Here
we notice that the PCs may show some spatial structure that correlates
to biological niches of cells.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe_filt</span>, <span class="st">"PCA"</span>, ncomponents <span class="op">=</span> <span class="fl">4</span>, </span>
<span>                  colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                  diverge_center <span class="op">=</span> <span class="fl">0</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>, pointsize <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_slideseq_v2_files/figure-html/unnamed-chunk-17-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Without the cellular overlays, we can only speculate about the
potential relevance of the barcodes that seem to be separated by each
PC, but each PC doe seem to separate distinct neighborhoods of
barcodes.</p>
<p>Now we can cluster the barcodes using a graph-based clustering
algorithm and plot them in space.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_filt</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_filt</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>,</span>
<span>                                           BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">SNNGraphParam</a></span><span class="op">(</span></span>
<span>                                               cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>,</span>
<span>                                               cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                                   resolution_parameter <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                                   objective_function <span class="op">=</span> <span class="st">"modularity"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The plot below is colored by cluster id. A naive interpretation of
the plot shows distinct niches of barcodes separated by more abundant,
intervening types. This may be indicative of the biological processes at
hand, namely melanoma metastasis, where ‘hotspots’ of melanoma
proliferation would be separated by unaffected normal tissue.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_filt</span>, <span class="st">"cluster"</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_slideseq_v2_files/figure-html/unnamed-chunk-19-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="section level3">
<h3 id="morans-i">Moran’s <em>I</em><a class="anchor" aria-label="anchor" href="#morans-i"></a>
</h3>
<p>One avenue for future analysis includes identifying genes that are
differentially expressed in each cluster, This can be interrogated with
<code><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers()</a></code> in a non-spatial context and with
<code><a href="../reference/calculateUnivariate.html">calculateMoransI()</a></code> in a spatial context. In the spatial
case, some consideration should be given to whether the differences seen
in across the tissue represent biological difference or artifacts from
field of view.</p>
<p>Here we run global Moran’s <em>I</em> on log normalized gene
expression.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runMoransI</a></span><span class="op">(</span><span class="va">sfe_filt</span>, features <span class="op">=</span> <span class="va">hvgs</span>, </span>
<span>                       BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now, we might ask: which genes display the most spatial
autocorrelation?</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe_filt</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_filt</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span>, </span>
<span>                                      decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_filt</span>, <span class="va">top_moran</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>                   scattermore <span class="op">=</span> <span class="cn">TRUE</span>, pointsize <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_slideseq_v2_files/figure-html/unnamed-chunk-21-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Spatial variability can also be investigated using differential
expression testing known anatomical regions complemented with spatial
location. One potential drawback to this approach is the variability
that is induced by the melanoma, rather than the native tissue
architecture, which may preclude identification of typical
structures.</p>
<p>Further analyses that can be done at this stage:</p>
<ol style="list-style-type: decimal">
<li>What gene expression patterns, if any, differentiate the
neighborhoods of melanoma cells?</li>
<li>What genes are differentially expressed in each cluster?</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.3 (2023-03-15)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Big Sur ... 10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] SpatialFeatureExperiment_1.0.3 BiocParallel_1.32.5           </span></span>
<span><span class="co">#&gt;  [3] spdep_1.2-8                    sf_1.0-11                     </span></span>
<span><span class="co">#&gt;  [5] spData_2.2.2                   patchwork_1.1.2               </span></span>
<span><span class="co">#&gt;  [7] bluster_1.8.0                  scran_1.26.2                  </span></span>
<span><span class="co">#&gt;  [9] scater_1.27.7                  ggplot2_3.4.1                 </span></span>
<span><span class="co">#&gt; [11] scuttle_1.8.4                  SpatialExperiment_1.8.1       </span></span>
<span><span class="co">#&gt; [13] SingleCellExperiment_1.20.0    SummarizedExperiment_1.28.0   </span></span>
<span><span class="co">#&gt; [15] Biobase_2.58.0                 GenomicRanges_1.50.2          </span></span>
<span><span class="co">#&gt; [17] GenomeInfoDb_1.34.9            IRanges_2.32.0                </span></span>
<span><span class="co">#&gt; [19] S4Vectors_0.36.2               BiocGenerics_0.44.0           </span></span>
<span><span class="co">#&gt; [21] MatrixGenerics_1.10.0          matrixStats_0.63.0            </span></span>
<span><span class="co">#&gt; [23] SFEData_1.0.2                  Voyager_1.0.10                </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] AnnotationHub_3.6.0           BiocFileCache_2.6.1          </span></span>
<span><span class="co">#&gt;   [3] systemfonts_1.0.4             igraph_1.4.1                 </span></span>
<span><span class="co">#&gt;   [5] sp_1.6-0                      scattermore_0.8              </span></span>
<span><span class="co">#&gt;   [7] digest_0.6.31                 htmltools_0.5.4              </span></span>
<span><span class="co">#&gt;   [9] viridis_0.6.2                 magick_2.7.4                 </span></span>
<span><span class="co">#&gt;  [11] fansi_1.0.4                   magrittr_2.0.3               </span></span>
<span><span class="co">#&gt;  [13] memoise_2.0.1                 ScaledMatrix_1.6.0           </span></span>
<span><span class="co">#&gt;  [15] cluster_2.1.4                 limma_3.54.2                 </span></span>
<span><span class="co">#&gt;  [17] Biostrings_2.66.0             R.utils_2.12.2               </span></span>
<span><span class="co">#&gt;  [19] pkgdown_2.0.7                 colorspace_2.1-0             </span></span>
<span><span class="co">#&gt;  [21] ggrepel_0.9.3                 blob_1.2.4                   </span></span>
<span><span class="co">#&gt;  [23] rappdirs_0.3.3                textshaping_0.3.6            </span></span>
<span><span class="co">#&gt;  [25] xfun_0.37                     dplyr_1.1.0                  </span></span>
<span><span class="co">#&gt;  [27] hexbin_1.28.2                 crayon_1.5.2                 </span></span>
<span><span class="co">#&gt;  [29] RCurl_1.98-1.10               jsonlite_1.8.4               </span></span>
<span><span class="co">#&gt;  [31] glue_1.6.2                    gtable_0.3.2                 </span></span>
<span><span class="co">#&gt;  [33] zlibbioc_1.44.0               XVector_0.38.0               </span></span>
<span><span class="co">#&gt;  [35] DelayedArray_0.24.0           scico_1.3.1                  </span></span>
<span><span class="co">#&gt;  [37] BiocSingular_1.14.0           DropletUtils_1.18.1          </span></span>
<span><span class="co">#&gt;  [39] Rhdf5lib_1.20.0               HDF5Array_1.26.0             </span></span>
<span><span class="co">#&gt;  [41] scales_1.2.1                  DBI_1.1.3                    </span></span>
<span><span class="co">#&gt;  [43] edgeR_3.40.2                  Rcpp_1.0.10                  </span></span>
<span><span class="co">#&gt;  [45] viridisLite_0.4.1             xtable_1.8-4                 </span></span>
<span><span class="co">#&gt;  [47] units_0.8-1                   dqrng_0.3.0                  </span></span>
<span><span class="co">#&gt;  [49] rsvd_1.0.5                    bit_4.0.5                    </span></span>
<span><span class="co">#&gt;  [51] proxy_0.4-27                  metapod_1.6.0                </span></span>
<span><span class="co">#&gt;  [53] httr_1.4.5                    RColorBrewer_1.1-3           </span></span>
<span><span class="co">#&gt;  [55] wk_0.7.2                      ellipsis_0.3.2               </span></span>
<span><span class="co">#&gt;  [57] farver_2.1.1                  pkgconfig_2.0.3              </span></span>
<span><span class="co">#&gt;  [59] R.methodsS3_1.8.2             sass_0.4.5                   </span></span>
<span><span class="co">#&gt;  [61] dbplyr_2.3.1                  deldir_1.0-6                 </span></span>
<span><span class="co">#&gt;  [63] locfit_1.5-9.7                utf8_1.2.3                   </span></span>
<span><span class="co">#&gt;  [65] labeling_0.4.2                tidyselect_1.2.0             </span></span>
<span><span class="co">#&gt;  [67] rlang_1.1.0                   later_1.3.0                  </span></span>
<span><span class="co">#&gt;  [69] AnnotationDbi_1.60.2          munsell_0.5.0                </span></span>
<span><span class="co">#&gt;  [71] BiocVersion_3.16.0            tools_4.2.3                  </span></span>
<span><span class="co">#&gt;  [73] cachem_1.0.7                  cli_3.6.0                    </span></span>
<span><span class="co">#&gt;  [75] dbscan_1.1-11                 ExperimentHub_2.6.0          </span></span>
<span><span class="co">#&gt;  [77] generics_0.1.3                RSQLite_2.3.0                </span></span>
<span><span class="co">#&gt;  [79] evaluate_0.20                 stringr_1.5.0                </span></span>
<span><span class="co">#&gt;  [81] fastmap_1.1.1                 yaml_2.3.7                   </span></span>
<span><span class="co">#&gt;  [83] ragg_1.2.5                    knitr_1.42                   </span></span>
<span><span class="co">#&gt;  [85] bit64_4.0.5                   fs_1.6.1                     </span></span>
<span><span class="co">#&gt;  [87] purrr_1.0.1                   s2_1.1.2                     </span></span>
<span><span class="co">#&gt;  [89] KEGGREST_1.38.0               sparseMatrixStats_1.10.0     </span></span>
<span><span class="co">#&gt;  [91] mime_0.12                     R.oo_1.25.0                  </span></span>
<span><span class="co">#&gt;  [93] compiler_4.2.3                beeswarm_0.4.0               </span></span>
<span><span class="co">#&gt;  [95] filelock_1.0.2                curl_5.0.0                   </span></span>
<span><span class="co">#&gt;  [97] png_0.1-8                     interactiveDisplayBase_1.36.0</span></span>
<span><span class="co">#&gt;  [99] e1071_1.7-13                  statmod_1.5.0                </span></span>
<span><span class="co">#&gt; [101] tibble_3.2.0                  bslib_0.4.2                  </span></span>
<span><span class="co">#&gt; [103] stringi_1.7.12                highr_0.10                   </span></span>
<span><span class="co">#&gt; [105] desc_1.4.2                    lattice_0.20-45              </span></span>
<span><span class="co">#&gt; [107] Matrix_1.5-3                  classInt_0.4-9               </span></span>
<span><span class="co">#&gt; [109] vctrs_0.6.0                   pillar_1.8.1                 </span></span>
<span><span class="co">#&gt; [111] lifecycle_1.0.3               rhdf5filters_1.10.0          </span></span>
<span><span class="co">#&gt; [113] BiocManager_1.30.20           jquerylib_0.1.4              </span></span>
<span><span class="co">#&gt; [115] BiocNeighbors_1.16.0          cowplot_1.1.1                </span></span>
<span><span class="co">#&gt; [117] irlba_2.3.5.1                 bitops_1.0-7                 </span></span>
<span><span class="co">#&gt; [119] httpuv_1.6.9                  R6_2.5.1                     </span></span>
<span><span class="co">#&gt; [121] promises_1.2.0.1              gridExtra_2.3                </span></span>
<span><span class="co">#&gt; [123] KernSmooth_2.23-20            vipor_0.4.5                  </span></span>
<span><span class="co">#&gt; [125] codetools_0.2-19              boot_1.3-28.1                </span></span>
<span><span class="co">#&gt; [127] rhdf5_2.42.0                  rprojroot_2.0.3              </span></span>
<span><span class="co">#&gt; [129] rjson_0.2.21                  withr_2.5.0                  </span></span>
<span><span class="co">#&gt; [131] GenomeInfoDbData_1.2.9        parallel_4.2.3               </span></span>
<span><span class="co">#&gt; [133] grid_4.2.3                    beachmat_2.14.0              </span></span>
<span><span class="co">#&gt; [135] class_7.3-21                  rmarkdown_2.20               </span></span>
<span><span class="co">#&gt; [137] DelayedMatrixStats_1.20.0     ggnewscale_0.4.8             </span></span>
<span><span class="co">#&gt; [139] shiny_1.7.4                   ggbeeswarm_0.7.1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Biermann2022-cu" class="csl-entry">
Biermann, Jana, Johannes C Melms, Amit Dipak Amin, Yiping Wang, Lindsay
A Caprio, Alcida Karz, Somnath Tagore, et al. 2022. <span>“Dissecting
the Treatment-Naive Ecosystem of Human Melanoma Brain
Metastasis.”</span> <em>Cell</em> 185 (14): 2591–2608.e30.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Kayla Jackson, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
