<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>Basic Visium exploratory data analysis • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Basic Visium exploratory data analysis">
<meta property="og:description" content="Voyager">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.7</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-vignettes">Vignettes</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-vignettes">
    <a class="dropdown-item" href="../articles/create_sfe.html">Create a SpatialFeatureExperiment object</a>
    <a class="dropdown-item" href="../articles/vig1_visium_basic.html">Basic Visium exploratory data analysis</a>
    <a class="dropdown-item" href="../articles/vig2_visium.html">Spatial Visium exploratory data analysis</a>
    <a class="dropdown-item" href="../articles/vig4_cosmx.html">CosMX non-small cell lung canger data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/Voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Basic Visium exploratory data analysis</h1>
                        <h4 data-toc-skip class="author">Lambda Moses</h4>
            
            <h4 data-toc-skip class="date">2022-10-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/Voyager/blob/HEAD/vignettes/vig1_visium_basic.Rmd" class="external-link"><code>vignettes/vig1_visium_basic.Rmd</code></a></small>
      <div class="d-none name"><code>vig1_visium_basic.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this introductory vignette for <a href="https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html" class="external-link"><code>SpatialFeatureExperiment</code></a> data representation and <a href="https://bioconductor.org/packages/devel/bioc/html/Voyager.html" class="external-link"><code>Voyager</code></a> anlaysis package, we demonstrate a basic exploratory data analysis (EDA) of spatial transcriptomics data. The vignette showcases the packages with a Visium spatial gene expression system dataset. The technology was chosen due to its popularity, and therefore the availability of numerous publicly available datasets for analysis <span class="citation">(Moses and Pachter 2022)</span>. <code>{r, echo = FALSE, fig.alt="Bar chart showing the number of institutions using each spatial transcriptomics data collection method. Only methods used by at least 3 different institutions are shown. The bars are colored by category of the methods. 10X Visium which is based on sequence barcoding is used by over 100 institutions. Following Visium are GeoMX DSP and GeoMX WTA which are "regions of interest" (ROI) methods, along with 2016 ST and some other ROI selection methods. knitr::include_graphics("https://pachterlab.github.io/LP_2021/04-current_files/figure-html/n-insts-1.png")</code></p>
<p>While Voyager was developed with the goal of facilitating the use of geospatial methods in spatial genomics, this introductory vignette is restricted to non-spatial scRNA-seq EDA with the Visium dataset. For a vignette illustrating univariate spatial analysis with the same dataset, see the more advanced <a href="https://pachterlab.github.io/voyager/articles/vig2_visium.htil">exploratory spatial data analyis vignette</a> with the same dataset.</p>
<p>Here we load the packages used in this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/Voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: scuttle</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/const-ae/sparseMatrixStats" class="external-link">sparseMatrixStats</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="mouse-skeletal-muscle-dataset">Mouse skeletal muscle dataset<a class="anchor" aria-label="anchor" href="#mouse-skeletal-muscle-dataset"></a>
</h2>
<p>The dataset used in this vignette comes from <a href="https://doi.org/10.1038/s42003-021-02810-x" class="external-link">Large-scale integration of single-cell transcriptomic data captures transitional progenitor states in mouse skeletal muscle regeneration</a> <span class="citation">(McKellar et al. 2021)</span>. Notexin was injected into the tibialis anterior muscle to induce injury, and the healing muscle was collected 2, 5, and 7 days post injury for Visium. The dataset here is from the 2 day timepoint. The dataset was originally not in the Space Ranger output format, but is already in a <code>SpatialFeatureExperiment</code> (SFE) object.</p>
<p>The gene count matrix was directly downloaded <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4904759" class="external-link">from GEO</a>. All 4992 spots, whether in tissue or not, are included. The H&amp;E image was used for nuclei and myofiber segmentation. A subset of nuclei from randomly selected regions from all 3 timepoints were manually annotated to train a StarDist model to segment the rest of the nuclei, and the myofibers were all manually segmented. The tissue boundary was found by thresholding in OpenCV, and small polygons were removed as they are likely to be debris. Spot polygons were constructed with the spot centroid coordinates and diameter in the Space Ranger output. The <code>in_tissue</code> column in <code>colData</code> indicates which spot polygons intersect the tissue polygons, and is based on <code><a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">st_intersects()</a></code>.</p>
<p>Tissue boundary, nuclei, myofiber, and Visium spot polygons are stored as <code>sf</code> data frames in the SFE object. See <a href="https://bioconductor.org/packages/devel/bioc/vignettes/SpatialFeatureExperiment/inst/doc/SFE.html" class="external-link">the vignette of <code>SpatialFeatureExperiment</code></a> for more details on the structure of the SFE object. The SFE object of this dataset is provided in the <code>SFEData</code> package. Here we download the data and load it into R.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="st">"full"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; snapshotDate(): 2022-10-17</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 4992 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(4992): AAACAACGAATAGTTC AAACAAGTATCTCCCA ... TTGTTTGTATTACACG</span></span>
<span><span class="co">#&gt;   TTGTTTGTGTAAATTC</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (POLYGON), myofiber_simplified (POLYGON), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A:</span></span></code></pre></div>
<p>The authors provided the full resolution hematoxylin and eosin (H&amp;E) image on GEO, which we downsized to more quickly show online. This is what it looks like: <img src="tissue_lowres_5a.jpeg" alt="A cross section of mouse muscle is slightly off center to the lower left. In the middle of the tissue is the notexin injury site with leukocyte infiltration and fewer myofibers. The rest of the tissue section is tightly packed with myofibers." width="100%"></p>
</div>
<div class="section level2">
<h2 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<div class="section level3">
<h3 id="spots">Spots<a class="anchor" aria-label="anchor" href="#spots"></a>
</h3>
<p>Here we plot the quality control (QC) metrics both as violin plots and in space. The QC metrics are pre-computed and stored in <code>colData</code> (for spots) and <code>rowData</code> of the SFE object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "barcode"   "col"       "row"       "x"         "y"         "dia"      </span></span>
<span><span class="co">#&gt;  [7] "tissue"    "sample_id" "nCounts"   "nGenes"    "prop_mito" "in_tissue"</span></span></code></pre></div>
<p>Here we plot the total UMI counts per spot. The commented out line of code shows how to compute total UMI counts.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># colData(sfe)$nCounts &lt;- colSums(counts(sfe))</span></span>
<span><span class="va">violin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, x <span class="op">=</span> <span class="st">"in_tissue"</span>, colour_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></span>
<span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                              annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">violin</span> <span class="op">+</span> <span class="va">spatial</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>Some spots in the injury site with leukocyte infiltration have high total counts. Spatial autocorrelation of the total counts is apparent, which will be discussed in a later section of this vignette.</p>
<p>Here we find number of genes detected per spot. The commented out line of code shows how to find the number of genes detected.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># colData(sfe)$nGenes &lt;- colSums(counts(sfe) &gt; 0)</span></span>
<span><span class="va">violin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nGenes"</span>, x <span class="op">=</span> <span class="st">"in_tissue"</span>, colour_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></span>
<span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nGenes"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                              annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">violin</span> <span class="op">+</span> <span class="va">spatial</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>As commonly done for scRNA-seq data, here we plot nCounts vs. nGenes</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"nGenes"</span>, colour_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>This plot has two branches for the spots in tissue, which turn out to be related to myofiber size. See <a href="https://pachterlab.github.io/voyager/articles/vig2_visium.htil">the exploratory spatial data analysis (ESDA) Visium vignette</a>.</p>
<p>As commonly done for scRNA-seq data, we plot the proportion of mitochondrially encoded counts. The commented out code shows how to find this proportion.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># mito_ind &lt;- str_detect(rowData(sfe)$symbol, "^Mt-")</span></span>
<span><span class="co"># colData(sfe)$prop_mito &lt;- colSums(counts(sfe)[mito_ind,]) / colData(sfe)$nCounts</span></span>
<span><span class="va">violin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"prop_mito"</span>, x <span class="op">=</span> <span class="st">"in_tissue"</span>, colour_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></span>
<span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"prop_mito"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                              annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">violin</span> <span class="op">+</span> <span class="va">spatial</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>As expected, spots outside tissue have higher proportion of mitochondrial counts, because when the tissue is lysed, mitochondrial transcripts are protected by a double membrane so are less likely to degrade than cytosolic transcripts. However, spots on myofibers also have high proportion of mitochondrial counts, because of the function of myofibers. The injury site with leukocyte infiltration has lower proportion of mitochondrial counts.</p>
<p>Here we plot the proportion of mitochondrial counts relate to total UMI counts, as commonly done for scRNA-seq to identify low quality cells with low UMI counts and high proportion of mitochondrial counts.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"prop_mito"</span>, colour_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>There are two clusters for the spots in tissue, which also turn out to be related to myofiber size. See <a href="https://pachterlab.github.io/voyager/articles/vig2_visium.htil">the ESDA Visium vignette</a>.</p>
<p>So far we haven’t seen spots that are obvious outliers in these QC metrics</p>
<p>The following analyses would only use spots in tissue.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">in_tissue</span><span class="op">]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe_tissue</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="genes">Genes<a class="anchor" aria-label="anchor" href="#genes"></a>
</h3>
<p>Like in scRNA-seq, gene expression in Visium is overdispersed compared to Poisson.</p>
<p>Here we compute the mean, variance, and coefficient of variance (CV2) for each gene among spots in tissue.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowVars.html" class="external-link">rowVars</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Coefficient of variance</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">cv2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">vars</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">means</span><span class="op">^</span><span class="fl">2</span></span></code></pre></div>
<p>To avoid overplotting and better show point density on the plot, we use a 2D histogram. The color of each bin indicates the number of points in that bin.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">means</span>, <span class="va">vars</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bin_2d.html" class="external-link">geom_bin2d</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, intercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Blues"</span>, direction <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>The red line, <span class="math inline">\(y = x\)</span>, is expected for Poisson distributed data, but here the variance is higher for more highly expressed genes than expected from Poisson. The CV2 gives the same picture.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">means</span>, <span class="va">cv2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bin_2d.html" class="external-link">geom_bin2d</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, intercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Blues"</span>, direction <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="normalize-data">Normalize data<a class="anchor" aria-label="anchor" href="#normalize-data"></a>
</h2>
<p>While <code>scran</code> <span class="citation">(Lun, Bach, and Marioni 2016)</span> is used for data normalization here for demonstration purposes and to make the data more normally distributed, we do not mean that it is the best practice in normalizing spatial transcriptomics data, as we don’t know what the best practice really should be. As seen in the <code>nCounts</code> plot in space above, spatial autocorrelation is evident. In Visium, reverse transcription occurs in situ on the spots, but PCR amplification occurs after the cDNA is dissociated from the spots. Then artifacts introduced from the amplification step would not be spatial. Spatial artifacts may arise from diffusion of transcripts and tissue permeablization. However, given how the total counts seem to correspond to histological regions, the total counts may have a biological component and hence should not be treated as a technical artifact to be normalized away as in scRNA-seq data normalization methods.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">quickCluster</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, clusters<span class="op">=</span><span class="va">clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in (function (x, sizes, min.mean = NULL, positive = FALSE, scaling =</span></span>
<span><span class="co">#&gt; NULL) : encountered non-positive size factor estimates</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe_tissue</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">sizeFactors</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<p>Find highly variable genes, and then use the highly variable genes for PCA</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dimension-reduction-and-clustering">Dimension reduction and clustering<a class="anchor" aria-label="anchor" href="#dimension-reduction-and-clustering"></a>
</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ncomponents <span class="op">=</span> <span class="fl">30</span>, subset_row <span class="op">=</span> <span class="va">hvgs</span>,</span>
<span>                     scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># scale as in Seurat</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ndims <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimLoadings.html">plotDimLoadings</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>The principal components (PCs) can be plotted in space. Due to spatial autocorrelation of many genes and spatial regions of different histological characters, even though spatial information is not used in PCA, the PCs may show spatial structure.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span>, ncomponents <span class="op">=</span> <span class="fl">4</span>, </span>
<span>                  colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                  diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-19-1.png" width="576"></p>
<p>PC1, which explains far more variance than PC2, separates the injury site myofibers close to the site from the visium myofibers. PC2 highlight some myofibers near the edge. PC3 highlights the muscle tendon junctions. PC4 does not seem to be informative; it might have picked up an outlier.</p>
<p>Also run UMAP based on PCA</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span>, n_dimred <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p>Do the clustering to show on the dimension reduction plots</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>,</span>
<span>                                           BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">SNNGraphParam</a></span><span class="op">(</span></span>
<span>                                               cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>,</span>
<span>                                               cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                                   resolution_parameter <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                                   objective_function <span class="op">=</span> <span class="st">"modularity"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ncomponents <span class="op">=</span> <span class="fl">3</span>, colour_by <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-22-1.png" width="672"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, colour_by <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<p>Where are the clusters in space?</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"cluster"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>While spatial information is not explicitly used, due to spatial autocorrelation of gene expression and the histological regions, some of these clusters are somewhat spatially contiguous. There are many methods to find spatially informed clusters, such as <a href="https://bioconductor.org/packages/release/bioc/html/BayesSpace.html" class="external-link"><code>BayesSpace</code></a> <span class="citation">(Zhao et al. 2021)</span>, which is on Bioconductor.</p>
<p>Remark on spatial regions: In geographical space, there usually is no one single way to define spatial regions. For example, influenced by both sociology and geology, LA county can be partitioned into regions such as Eastside, Westside, San Fernado Valley, San Gabriel Valley, Pomona Valley, Gateway Cities, South Bay, and etc., each containing multiple smaller cities or parts of LA City, each of which can be further divided into many neighborhoods, such as downtown LA, Koreatown, Highland Park, and etc. Definitions of some of these regions are subject to dispute. Meanwhile, LA county can also be partitioned into watersheds of the LA River, San Gabriel River, Ballona Creek, and etc., as well as different rock formations. Which kind of spatial region at which resolution is relevant depends on the question asked.</p>
<p>In contrast, in spatial transcriptomics, methods identifying spatial regions generally only aim to give one result, while multiple results at different resolutions depending on the question asked can be relevant. Also, just like how the definition of Eastside is subject to a lot of dispute, it would be great if spatial region methods for spatial -omics can give uncertainty in assignment of cells or Visium spots to spatial regions.</p>
</div>
<div class="section level2">
<h2 id="non-spatial-differential-expression">Non-spatial differential expression<a class="anchor" aria-label="anchor" href="#non-spatial-differential-expression"></a>
</h2>
<p>Cluster marker genes are found with Wilcoxon rank sum test as commonly done for scRNA-seq.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span>,</span>
<span>                       test.type <span class="op">=</span> <span class="st">"wilcox"</span>, pval.type <span class="op">=</span> <span class="st">"all"</span>, direction <span class="op">=</span> <span class="st">"up"</span><span class="op">)</span></span></code></pre></div>
<p>It’s already sorted by p-values</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 15043 rows and 8 columns</span></span>
<span><span class="co">#&gt;                        p.value         FDR summary.AUC     AUC.1     AUC.2</span></span>
<span><span class="co">#&gt;                      &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000098178 5.05056e-11 7.59756e-07    0.734338  0.734338  0.875668</span></span>
<span><span class="co">#&gt; ENSMUSG00000028031 5.91274e-02 1.00000e+00    0.523428  0.539526  0.545455</span></span>
<span><span class="co">#&gt; ENSMUSG00000044349 1.43199e-01 1.00000e+00    0.505682  0.505682  0.505682</span></span>
<span><span class="co">#&gt; ENSMUSG00000095676 1.43199e-01 1.00000e+00    0.505682  0.505682  0.505682</span></span>
<span><span class="co">#&gt; ENSMUSG00000050640 1.43199e-01 1.00000e+00    0.505682  0.505682  0.505682</span></span>
<span><span class="co">#&gt; ...                        ...         ...         ...       ...       ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000087095           1           1         0.5       0.5  0.495098</span></span>
<span><span class="co">#&gt; ENSMUSG00000043969           1           1         0.5       0.5  0.500000</span></span>
<span><span class="co">#&gt; ENSMUSG00000091378           1           1         0.5       0.5  0.500000</span></span>
<span><span class="co">#&gt; ENSMUSG00000072437           1           1         0.5       0.5  0.500000</span></span>
<span><span class="co">#&gt; ENSMUSG00000094649           1           1         0.5       0.5  0.495098</span></span>
<span><span class="co">#&gt;                        AUC.3     AUC.4     AUC.5</span></span>
<span><span class="co">#&gt;                    &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000098178  0.910395  0.789497  0.875780</span></span>
<span><span class="co">#&gt; ENSMUSG00000028031  0.538003  0.523428  0.540998</span></span>
<span><span class="co">#&gt; ENSMUSG00000044349  0.505682  0.505682  0.505682</span></span>
<span><span class="co">#&gt; ENSMUSG00000095676  0.505682  0.505682  0.505682</span></span>
<span><span class="co">#&gt; ENSMUSG00000050640  0.505682  0.505682  0.505682</span></span>
<span><span class="co">#&gt; ...                      ...       ...       ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000087095  0.497268       0.5  0.495098</span></span>
<span><span class="co">#&gt; ENSMUSG00000043969  0.497268       0.5  0.500000</span></span>
<span><span class="co">#&gt; ENSMUSG00000091378  0.497268       0.5  0.500000</span></span>
<span><span class="co">#&gt; ENSMUSG00000072437  0.494536       0.5  0.500000</span></span>
<span><span class="co">#&gt; ENSMUSG00000094649  0.500000       0.5  0.500000</span></span></code></pre></div>
<p>Get the the significant marker for each cluster to plot</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, FUN.VALUE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="va">genes_use</span>, <span class="st">"symbol"</span><span class="op">]</span>, x <span class="op">=</span> <span class="st">"cluster"</span>,</span>
<span>               colour_by <span class="op">=</span> <span class="st">"cluster"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>Plot those genes in space</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">genes_use</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="morans-i">Moran’s I<a class="anchor" aria-label="anchor" href="#morans-i"></a>
</h2>
<p>Tobler’s first law of geography:</p>
<blockquote>
<p>Everything is related to everything else. But near things are more related than distant things.</p>
</blockquote>
<p>This is what spatial autocorrelation basically is. In positive spatial autocorrelation, nearby things tend to be more similar, such as that weather in Pasadena and downtoan Los Angeles tend to be more similar than weather in Pasadena and San Francisco. There can also be negative spatial autocorrelation, where nearby things tend to be more dissimilar, like a checkerboard. Spatial autocorrelation can be caused by an intrinsic process such as diffusion or communication by physical contact, or by another covariate that has such an intrinsic process.</p>
<p>The most commonly used metric of spatial autocorrelation is Moran’s I, defined as</p>
<p><span class="math display">\[
I = \frac{n}{\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n \sum_{j=1}^n w_{ij} (x_i - \bar{x})(x_j - \bar{x})}{\sum_{i=1}^n (x_i - \bar{x})^2},
\]</span></p>
<p>where <span class="math inline">\(n\)</span> is the number of spots or locations, <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> are different locations, or spots in the Visium context, <span class="math inline">\(x\)</span> is a variable with values at each location, and <span class="math inline">\(w_{ij}\)</span> is a spatial weight, which can be inversely proportional to distance between spots or an indicator of whether two spots are neighbors, subject to various definitions of neighborhood and whether to normalize the number of neighbors. The <a href="https://r-spatial.github.io/spdep/index.html" class="external-link"><code>spdep</code></a> package uses the neighborhood.</p>
<p>Moran’s I looks kind of like Pearson correlation between the value at each location and the average value at its neighbors. Just like Pearson correlation, Moran’s I is generally bound between -1 and 1, where positive value indicates positive spatial autocorrelation and negative value indicates negative spatial autocorrelation.</p>
<p>Spatial dependence analysis in <code>spdep</code> requires a spatial neighborhood graph. The graph for adjacent Visium spot can be found with</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<p>We mentioned that spatial autocorrelation is apparent in total UMI counts. Here’s what Moran’s I has to say:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/calculateUnivariate.html">calculateMoransI</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                 listw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">#&gt;             moran         K</span></span>
<span><span class="co">#&gt;         &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts  0.528705   3.00082</span></span>
<span><span class="co">#&gt; nGenes   0.384028   3.88036</span></span></code></pre></div>
<p>K means kurtosis. The positive values of Moran’s I indicate positive spatial autocorrelation.</p>
<div class="section level3">
<h3 id="spatially-variable-genes">Spatially variable genes<a class="anchor" aria-label="anchor" href="#spatially-variable-genes"></a>
</h3>
<p>A spatially variable gene is a gene whose expression depends on spatial locations, rather than being scatter around like salt and pepper. Then spatial variable genes would have spatial autocorrelation, and sometimes Moran’s I is used to compare and validate the spatially variable genes identified with different methods. Seurat ranks Moran’s I to identify spatially variable genes. This can also be done with <code>Voyager</code>. Here <code>BPPARAM</code> is used to parallelize the computation of Moran’s I for 2000 highly variable genes, and 2 cores are used with the SNOW backend.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runMoransI</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="va">hvgs</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>,</span>
<span>                         BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: &lt;anonymous&gt;: ... may be used in an incorrect context: 'fun(x[i, ], listw, ...)'</span></span></code></pre></div>
<p>The results are stored in <code>rowData</code></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 15043 rows and 8 columns</span></span>
<span><span class="co">#&gt;                               Ensembl      symbol            type      means</span></span>
<span><span class="co">#&gt;                           &lt;character&gt; &lt;character&gt;     &lt;character&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 ENSMUSG00000025902       Sox17 Gene Expression 0.03969957</span></span>
<span><span class="co">#&gt; ENSMUSG00000096126 ENSMUSG00000096126     Gm22307 Gene Expression 0.00107296</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 ENSMUSG00000033845      Mrpl15 Gene Expression 0.38197425</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 ENSMUSG00000025903      Lypla1 Gene Expression 0.28755365</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 ENSMUSG00000033813       Tcea1 Gene Expression 0.26502146</span></span>
<span><span class="co">#&gt; ...                               ...         ...             ...        ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000064360 ENSMUSG00000064360      mt-Nd3 Gene Expression  56.445279</span></span>
<span><span class="co">#&gt; ENSMUSG00000064363 ENSMUSG00000064363      mt-Nd4 Gene Expression 123.991416</span></span>
<span><span class="co">#&gt; ENSMUSG00000064367 ENSMUSG00000064367      mt-Nd5 Gene Expression  14.645923</span></span>
<span><span class="co">#&gt; ENSMUSG00000064368 ENSMUSG00000064368      mt-Nd6 Gene Expression   0.109442</span></span>
<span><span class="co">#&gt; ENSMUSG00000064370 ENSMUSG00000064370     mt-Cytb Gene Expression 121.273605</span></span>
<span><span class="co">#&gt;                           vars       cv2 moran_Vis5A   K_Vis5A</span></span>
<span><span class="co">#&gt;                      &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000025902  0.04460915  28.30429          NA        NA</span></span>
<span><span class="co">#&gt; ENSMUSG00000096126  0.00107296 932.00000          NA        NA</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845  0.47048031   3.22458          NA        NA</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903  0.34686963   4.19497          NA        NA</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813  0.32388797   4.61140   0.0197543   29.6881</span></span>
<span><span class="co">#&gt; ...                        ...       ...         ...       ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000064360 2.47976e+03  0.778314    0.605694   4.70819</span></span>
<span><span class="co">#&gt; ENSMUSG00000064363 1.45282e+04  0.944991    0.680026   4.79491</span></span>
<span><span class="co">#&gt; ENSMUSG00000064367 2.34858e+02  1.094895    0.635612   2.74170</span></span>
<span><span class="co">#&gt; ENSMUSG00000064368 1.31941e-01 11.015664          NA        NA</span></span>
<span><span class="co">#&gt; ENSMUSG00000064370 1.48225e+04  1.007833    0.722350   3.95620</span></span></code></pre></div>
<p>The <code>NA</code>’s are for genes that are not highly variable and Moran’s I was not computed for those genes. Now we rank the genes by Moran’s I and plot them in space.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="va">hvgs</span>,<span class="op">]</span></span>
<span><span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">moran_Vis5A</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">df</span><span class="op">[</span><span class="va">ord</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"symbol"</span>, <span class="st">"moran_Vis5A"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 2000 rows and 2 columns</span></span>
<span><span class="co">#&gt;                         symbol moran_Vis5A</span></span>
<span><span class="co">#&gt;                    &lt;character&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000064351      mt-Co1    0.814980</span></span>
<span><span class="co">#&gt; ENSMUSG00000064345      mt-Nd2    0.784339</span></span>
<span><span class="co">#&gt; ENSMUSG00000018893          Mb    0.761522</span></span>
<span><span class="co">#&gt; ENSMUSG00000056328        Myh1    0.727282</span></span>
<span><span class="co">#&gt; ENSMUSG00000064370     mt-Cytb    0.722350</span></span>
<span><span class="co">#&gt; ...                        ...         ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000055436      Srsf11  -0.0397929</span></span>
<span><span class="co">#&gt; ENSMUSG00000067713      Prkag1  -0.0404949</span></span>
<span><span class="co">#&gt; ENSMUSG00000090262       Mpv17  -0.0423579</span></span>
<span><span class="co">#&gt; ENSMUSG00000030061        Uba3  -0.0424449</span></span>
<span><span class="co">#&gt; ENSMUSG00000020964       Sel1l  -0.0507095</span></span></code></pre></div>
<p>Here we see some genes that have strong positive spatial autocorrelation, but don’t observe strong negative spatial autocorrelation. Plot the genes with the strongest positive spatial autocorrelation in space</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
<p>These genes do indeed look like spatially variable. However, such spatial variability can simply be due to the histological regions in space, or in other words, spatial distribution of different cell types. There are many methods to identify spatially variable genes, often involving Gaussian process modeling, which are far more complex than Moran’s I, such as <a href="https://bioconductor.org/packages/release/bioc/html/spatialDE.html" class="external-link"><code>SpatialDE</code></a> <span class="citation">(Svensson, Teichmann, and Stegle 2018)</span>. However, such methods usually don’t account for the histological regions, except for <code>C-SIDE</code> <span class="citation">(Cable et al. 2022)</span>, which identifies spatially variable genes within cell types. But then what do we really mean by “cell type”? It remains to see whether spatial methods made specifically for identifying spatially variable genes compare with methods that don’t explicitly use spatial information but simply performs DE between cell types which often are in spatially defined histological regions.</p>
<p>Further considerations on Moran’s I: strength of spatial autocorrelation can vary in space. What if a gene has strong spatial autocorrelation in one region, but is more like salt and pepper in another? Should the different histological regions be analyzed separately in some cases?</p>
<p>There are ways to see whether Moran’s I is statistically significant, and many other methods to explore spatial autocorrelation. These are discussed in <a href="https://pachterlab.github.io/voyager/articles/vig2_visium.htil">the ESDA Visium vignette</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 20.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] sparseMatrixStats_1.8.0          stringr_1.4.1                   </span></span>
<span><span class="co">#&gt;  [3] BiocParallel_1.30.4              SFEData_0.99.5                  </span></span>
<span><span class="co">#&gt;  [5] bluster_1.6.0                    patchwork_1.1.2                 </span></span>
<span><span class="co">#&gt;  [7] scran_1.24.1                     scater_1.24.0                   </span></span>
<span><span class="co">#&gt;  [9] ggplot2_3.3.6                    scuttle_1.7.4                   </span></span>
<span><span class="co">#&gt; [11] SpatialExperiment_1.6.1          SingleCellExperiment_1.18.1     </span></span>
<span><span class="co">#&gt; [13] SummarizedExperiment_1.26.1      Biobase_2.57.1                  </span></span>
<span><span class="co">#&gt; [15] GenomicRanges_1.48.0             GenomeInfoDb_1.33.14            </span></span>
<span><span class="co">#&gt; [17] IRanges_2.31.2                   S4Vectors_0.35.4                </span></span>
<span><span class="co">#&gt; [19] BiocGenerics_0.43.4              MatrixGenerics_1.9.1            </span></span>
<span><span class="co">#&gt; [21] matrixStats_0.62.0               SpatialFeatureExperiment_0.99.10</span></span>
<span><span class="co">#&gt; [23] Voyager_0.99.7                  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] snow_0.4-4                    AnnotationHub_3.5.2          </span></span>
<span><span class="co">#&gt;   [3] BiocFileCache_2.4.0           systemfonts_1.0.4            </span></span>
<span><span class="co">#&gt;   [5] igraph_1.3.5                  sp_1.5-0                     </span></span>
<span><span class="co">#&gt;   [7] digest_0.6.30                 htmltools_0.5.3              </span></span>
<span><span class="co">#&gt;   [9] viridis_0.6.2                 magick_2.7.3                 </span></span>
<span><span class="co">#&gt;  [11] fansi_1.0.3                   magrittr_2.0.3               </span></span>
<span><span class="co">#&gt;  [13] memoise_2.0.1                 ScaledMatrix_1.5.1           </span></span>
<span><span class="co">#&gt;  [15] cluster_2.1.3                 limma_3.53.10                </span></span>
<span><span class="co">#&gt;  [17] Biostrings_2.65.6             R.utils_2.12.0               </span></span>
<span><span class="co">#&gt;  [19] pkgdown_2.0.6                 colorspace_2.0-3             </span></span>
<span><span class="co">#&gt;  [21] rappdirs_0.3.3                blob_1.2.3                   </span></span>
<span><span class="co">#&gt;  [23] ggrepel_0.9.1                 textshaping_0.3.6            </span></span>
<span><span class="co">#&gt;  [25] xfun_0.34                     dplyr_1.0.10                 </span></span>
<span><span class="co">#&gt;  [27] crayon_1.5.2                  RCurl_1.98-1.9               </span></span>
<span><span class="co">#&gt;  [29] jsonlite_1.8.2                glue_1.6.2                   </span></span>
<span><span class="co">#&gt;  [31] gtable_0.3.1                  zlibbioc_1.43.0              </span></span>
<span><span class="co">#&gt;  [33] XVector_0.37.1                DelayedArray_0.23.2          </span></span>
<span><span class="co">#&gt;  [35] scico_1.3.1                   BiocSingular_1.13.1          </span></span>
<span><span class="co">#&gt;  [37] DropletUtils_1.17.3           Rhdf5lib_1.19.2              </span></span>
<span><span class="co">#&gt;  [39] HDF5Array_1.25.2              scales_1.2.1                 </span></span>
<span><span class="co">#&gt;  [41] DBI_1.1.3                     edgeR_3.39.6                 </span></span>
<span><span class="co">#&gt;  [43] Rcpp_1.0.9                    xtable_1.8-4                 </span></span>
<span><span class="co">#&gt;  [45] viridisLite_0.4.1             spData_2.2.0                 </span></span>
<span><span class="co">#&gt;  [47] units_0.8-0                   dqrng_0.3.0                  </span></span>
<span><span class="co">#&gt;  [49] bit_4.0.4                     spdep_1.2-7                  </span></span>
<span><span class="co">#&gt;  [51] rsvd_1.0.5                    proxy_0.4-27                 </span></span>
<span><span class="co">#&gt;  [53] httr_1.4.4                    metapod_1.5.0                </span></span>
<span><span class="co">#&gt;  [55] FNN_1.1.3.1                   RColorBrewer_1.1-3           </span></span>
<span><span class="co">#&gt;  [57] ellipsis_0.3.2                wk_0.7.0                     </span></span>
<span><span class="co">#&gt;  [59] farver_2.1.1                  pkgconfig_2.0.3              </span></span>
<span><span class="co">#&gt;  [61] R.methodsS3_1.8.2             uwot_0.1.14                  </span></span>
<span><span class="co">#&gt;  [63] dbplyr_2.2.1                  sass_0.4.2                   </span></span>
<span><span class="co">#&gt;  [65] deldir_1.0-6                  locfit_1.5-9.6               </span></span>
<span><span class="co">#&gt;  [67] utf8_1.2.2                    labeling_0.4.2               </span></span>
<span><span class="co">#&gt;  [69] AnnotationDbi_1.59.1          later_1.3.0                  </span></span>
<span><span class="co">#&gt;  [71] tidyselect_1.2.0              rlang_1.0.6                  </span></span>
<span><span class="co">#&gt;  [73] munsell_0.5.0                 BiocVersion_3.16.0           </span></span>
<span><span class="co">#&gt;  [75] tools_4.2.1                   cachem_1.0.6                 </span></span>
<span><span class="co">#&gt;  [77] cli_3.4.1                     ExperimentHub_2.5.0          </span></span>
<span><span class="co">#&gt;  [79] generics_0.1.3                RSQLite_2.2.18               </span></span>
<span><span class="co">#&gt;  [81] evaluate_0.17                 fastmap_1.1.0                </span></span>
<span><span class="co">#&gt;  [83] yaml_2.3.6                    ragg_1.2.3                   </span></span>
<span><span class="co">#&gt;  [85] knitr_1.40                    bit64_4.0.5                  </span></span>
<span><span class="co">#&gt;  [87] fs_1.5.2                      purrr_0.3.5                  </span></span>
<span><span class="co">#&gt;  [89] s2_1.1.0                      KEGGREST_1.36.3              </span></span>
<span><span class="co">#&gt;  [91] mime_0.12                     R.oo_1.25.0                  </span></span>
<span><span class="co">#&gt;  [93] compiler_4.2.1                png_0.1-7                    </span></span>
<span><span class="co">#&gt;  [95] interactiveDisplayBase_1.34.0 filelock_1.0.2               </span></span>
<span><span class="co">#&gt;  [97] curl_4.3.3                    beeswarm_0.4.0               </span></span>
<span><span class="co">#&gt;  [99] e1071_1.7-11                  tibble_3.1.8                 </span></span>
<span><span class="co">#&gt; [101] statmod_1.4.37                bslib_0.4.0                  </span></span>
<span><span class="co">#&gt; [103] stringi_1.7.8                 highr_0.9                    </span></span>
<span><span class="co">#&gt; [105] desc_1.4.2                    lattice_0.20-45              </span></span>
<span><span class="co">#&gt; [107] Matrix_1.5-1                  classInt_0.4-8               </span></span>
<span><span class="co">#&gt; [109] vctrs_0.4.2                   pillar_1.8.1                 </span></span>
<span><span class="co">#&gt; [111] lifecycle_1.0.3               rhdf5filters_1.9.0           </span></span>
<span><span class="co">#&gt; [113] BiocManager_1.30.18           jquerylib_0.1.4              </span></span>
<span><span class="co">#&gt; [115] BiocNeighbors_1.14.0          cowplot_1.1.1                </span></span>
<span><span class="co">#&gt; [117] bitops_1.0-7                  irlba_2.3.5.1                </span></span>
<span><span class="co">#&gt; [119] httpuv_1.6.6                  R6_2.5.1                     </span></span>
<span><span class="co">#&gt; [121] promises_1.2.0.1              KernSmooth_2.23-20           </span></span>
<span><span class="co">#&gt; [123] gridExtra_2.3                 vipor_0.4.5                  </span></span>
<span><span class="co">#&gt; [125] codetools_0.2-18              boot_1.3-28                  </span></span>
<span><span class="co">#&gt; [127] assertthat_0.2.1              rhdf5_2.40.0                 </span></span>
<span><span class="co">#&gt; [129] rprojroot_2.0.3               rjson_0.2.21                 </span></span>
<span><span class="co">#&gt; [131] withr_2.5.0                   GenomeInfoDbData_1.2.8       </span></span>
<span><span class="co">#&gt; [133] parallel_4.2.1                grid_4.2.1                   </span></span>
<span><span class="co">#&gt; [135] beachmat_2.13.4               class_7.3-20                 </span></span>
<span><span class="co">#&gt; [137] rmarkdown_2.17                DelayedMatrixStats_1.18.2    </span></span>
<span><span class="co">#&gt; [139] ggnewscale_0.4.8              sf_1.0-8                     </span></span>
<span><span class="co">#&gt; [141] shiny_1.7.2                   ggbeeswarm_0.6.0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Cable2022-ma" class="csl-entry">
Cable, Dylan M, Evan Murray, Vignesh Shanmugam, Simon Zhang, Luli S Zou, Michael Diao, Haiqi Chen, Evan Z Macosko, Rafael A Irizarry, and Fei Chen. 2022. <span>“Cell Type-Specific Inference of Differential Expression in Spatial Transcriptomics.”</span> <em>Nat. Methods</em> 19 (9): 1076–87.
</div>
<div id="ref-Lun2016-yq" class="csl-entry">
Lun, Aaron T L, Karsten Bach, and John C Marioni. 2016. <span>“Pooling Across Cells to Normalize Single-Cell <span>RNA</span> Sequencing Data with Many Zero Counts.”</span> <em>Genome Biol.</em> 17 (April): 75.
</div>
<div id="ref-McKellar2021-ek" class="csl-entry">
McKellar, David W, Lauren D Walter, Leo T Song, Madhav Mantri, Michael F Z Wang, Iwijn De Vlaminck, and Benjamin D Cosgrove. 2021. <span>“Large-Scale Integration of Single-Cell Transcriptomic Data Captures Transitional Progenitor States in Mouse Skeletal Muscle Regeneration.”</span> <em>Commun Biol</em> 4 (1): 1280.
</div>
<div id="ref-Moses2022-xz" class="csl-entry">
Moses, Lambda, and Lior Pachter. 2022. <span>“Publisher Correction: Museum of Spatial Transcriptomics.”</span> <em>Nat. Methods</em> 19 (5): 628.
</div>
<div id="ref-Svensson2018-sx" class="csl-entry">
Svensson, Valentine, Sarah A Teichmann, and Oliver Stegle. 2018. <span>“<span>SpatialDE</span>: Identification of Spatially Variable Genes.”</span> <em>Nat. Methods</em> 15 (5): 343–46.
</div>
<div id="ref-Zhao2021-vb" class="csl-entry">
Zhao, Edward, Matthew R Stone, Xing Ren, Jamie Guenthoer, Kimberly S Smythe, Thomas Pulliam, Stephen R Williams, et al. 2021. <span>“Spatial Transcriptomics at Subspot Resolution with <span>BayesSpace</span>.”</span> <em>Nat. Biotechnol.</em> 39 (11): 1375–84.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
